// @schedulesync/frontend/src/components/grid/TaskRow.tsx
/**
 * TaskRow — a single row in the task grid.
 *
 * Supports inline editing for Name, Duration, Start, Finish, % Complete.
 * Shows indentation via outlineLevel and summary/milestone icons.
 */

import React, { useState, useCallback, memo } from 'react';
import { TableRow, TableCell, Box, IconButton, Checkbox } from '@mui/material';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import ChevronRightIcon from '@mui/icons-material/ChevronRight';
import DiamondIcon from '@mui/icons-material/Diamond';

import InlineEditor from './InlineEditor';
import { useProjectStore, type TaskRow as TaskRowType } from '../../stores';
import { useUIStore } from '../../stores';
import {
  shortDate,
  isoDate,
  durationDays,
  parseDuration,
  pctLabel,
} from '../../utils/format';

interface TaskRowProps {
  task: TaskRowType;
  index: number;
  isSelected: boolean;
  isExpanded: boolean;
  hasChildren: boolean;
  onToggleExpand: (taskId: string) => void;
}

type EditingField = 'name' | 'duration' | 'start' | 'finish' | 'pct' | null;

const TaskRowComponent: React.FC<TaskRowProps> = ({
  task,
  index,
  isSelected,
  isExpanded,
  hasChildren,
  onToggleExpand,
}) => {
  const updateTask = useProjectStore((s) => s.updateTask);
  const selectTask = useProjectStore((s) => s.selectTask);
  const showCriticalPath = useUIStore((s) => s.showCriticalPath);
  const [editing, setEditing] = useState<EditingField>(null);

  const isSummary = task.type === 2;
  const isMilestone = task.type === 1 || task.durationMinutes === 0;

  const handleDoubleClick = useCallback(
    (field: EditingField) => {
      if (isSummary && field !== 'name') return; // don't edit computed summary fields
      setEditing(field);
    },
    [isSummary],
  );

  const commitEdit = useCallback(
    async (field: string, rawValue: string) => {
      setEditing(null);
      const data: Record<string, unknown> = {};

      switch (field) {
        case 'name':
          if (rawValue.trim() && rawValue !== task.name) {
            data.name = rawValue.trim();
          }
          break;
        case 'duration': {
          const mins = parseDuration(rawValue);
          if (mins !== null && mins !== task.durationMinutes) {
            data.durationMinutes = mins;
          }
          break;
        }
        case 'start':
          if (rawValue && rawValue !== isoDate(task.start)) {
            data.start = new Date(rawValue).toISOString();
          }
          break;
        case 'finish':
          if (rawValue && rawValue !== isoDate(task.finish)) {
            data.finish = new Date(rawValue).toISOString();
          }
          break;
        case 'pct': {
          const pct = parseFloat(rawValue);
          if (!isNaN(pct) && pct >= 0 && pct <= 100 && pct !== task.percentComplete) {
            data.percentComplete = pct;
          }
          break;
        }
      }

      if (Object.keys(data).length > 0) {
        await updateTask(task.id, data);
      }
    },
    [task, updateTask],
  );

  const indentPx = task.outlineLevel * 20;
  const criticalColor =
    showCriticalPath && task.isCritical ? '#D32F2F' : undefined;

  return (
    <TableRow
      hover
      selected={isSelected}
      onClick={(e) => selectTask(task.id, e.ctrlKey || e.metaKey)}
      sx={{
        cursor: 'pointer',
        '& td': {
          color: criticalColor,
          fontWeight: isSummary ? 700 : 400,
        },
      }}
    >
      {/* Row number */}
      <TableCell align="center" sx={{ width: 40, minWidth: 40 }}>
        {index + 1}
      </TableCell>

      {/* Task name with indent */}
      <TableCell
        onDoubleClick={() => handleDoubleClick('name')}
        sx={{ minWidth: 240 }}
      >
        {editing === 'name' ? (
          <InlineEditor
            value={task.name}
            onCommit={(v) => commitEdit('name', v)}
            onCancel={() => setEditing(null)}
          />
        ) : (
          <Box sx={{ display: 'flex', alignItems: 'center', pl: `${indentPx}px` }}>
            {hasChildren && (
              <IconButton
                size="small"
                onClick={(e) => {
                  e.stopPropagation();
                  onToggleExpand(task.id);
                }}
                sx={{ p: 0, mr: 0.5 }}
              >
                {isExpanded ? (
                  <ExpandMoreIcon fontSize="small" />
                ) : (
                  <ChevronRightIcon fontSize="small" />
                )}
              </IconButton>
            )}
            {isMilestone && !isSummary && (
              <DiamondIcon sx={{ fontSize: 14, mr: 0.5, color: '#333' }} />
            )}
            <span>{task.name}</span>
          </Box>
        )}
      </TableCell>

      {/* Duration */}
      <TableCell
        align="right"
        onDoubleClick={() => handleDoubleClick('duration')}
        sx={{ width: 80 }}
      >
        {editing === 'duration' ? (
          <InlineEditor
            value={durationDays(task.durationMinutes)}
            onCommit={(v) => commitEdit('duration', v)}
            onCancel={() => setEditing(null)}
            type="duration"
          />
        ) : (
          durationDays(task.durationMinutes)
        )}
      </TableCell>

      {/* Start */}
      <TableCell
        onDoubleClick={() => handleDoubleClick('start')}
        sx={{ width: 100 }}
      >
        {editing === 'start' ? (
          <InlineEditor
            value={isoDate(task.start)}
            onCommit={(v) => commitEdit('start', v)}
            onCancel={() => setEditing(null)}
            type="date"
          />
        ) : (
          shortDate(task.start)
        )}
      </TableCell>

      {/* Finish */}
      <TableCell
        onDoubleClick={() => handleDoubleClick('finish')}
        sx={{ width: 100 }}
      >
        {editing === 'finish' ? (
          <InlineEditor
            value={isoDate(task.finish)}
            onCommit={(v) => commitEdit('finish', v)}
            onCancel={() => setEditing(null)}
            type="date"
          />
        ) : (
          shortDate(task.finish)
        )}
      </TableCell>

      {/* % Complete */}
      <TableCell
        align="right"
        onDoubleClick={() => handleDoubleClick('pct')}
        sx={{ width: 60 }}
      >
        {editing === 'pct' ? (
          <InlineEditor
            value={String(task.percentComplete)}
            onCommit={(v) => commitEdit('pct', v)}
            onCancel={() => setEditing(null)}
            type="number"
          />
        ) : (
          pctLabel(task.percentComplete)
        )}
      </TableCell>
    </TableRow>
  );
};

export default memo(TaskRowComponent);
