// @schedulesync/engine/dist/scheduler/constraints.js
/**
 * Constraint handling for the CPM scheduler.
 * Mirrors MS Project's PjConstraint behaviour.
 */
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc.js';
import { ConstraintType, ValidationSeverity } from '../types.js';
dayjs.extend(utc);
/**
 * Apply constraint to the computed early start/finish during forward pass.
 * Some constraints (ALAP) are only meaningful after the backward pass and
 * are handled separately by applyALAPConstraint.
 */
export function applyConstraintForwardPass(taskId, constraintType, constraintDate, earlyStart, earlyFinish, durationMinutes, addWorkingMinutesFn, subtractWorkingMinutesFn) {
    const warnings = [];
    if (constraintType === ConstraintType.ASAP ||
        constraintType === ConstraintType.ALAP ||
        !constraintDate) {
        return { earlyStart, earlyFinish, warnings };
    }
    const cd = dayjs.utc(constraintDate);
    switch (constraintType) {
        case ConstraintType.MustStartOn:
            return {
                earlyStart: cd,
                earlyFinish: addWorkingMinutesFn(cd, durationMinutes),
                warnings,
            };
        case ConstraintType.MustFinishOn:
            return {
                earlyStart: subtractWorkingMinutesFn(cd, durationMinutes),
                earlyFinish: cd,
                warnings,
            };
        case ConstraintType.StartNoEarlierThan:
            if (earlyStart.isBefore(cd)) {
                const newStart = cd;
                return {
                    earlyStart: newStart,
                    earlyFinish: addWorkingMinutesFn(newStart, durationMinutes),
                    warnings,
                };
            }
            return { earlyStart, earlyFinish, warnings };
        case ConstraintType.StartNoLaterThan:
            if (earlyStart.isAfter(cd)) {
                warnings.push({
                    taskId,
                    message: `Task cannot start by constraint date ${constraintDate} due to dependencies. Computed start: ${earlyStart.toISOString()}.`,
                    severity: ValidationSeverity.Warning,
                });
            }
            return { earlyStart, earlyFinish, warnings };
        case ConstraintType.FinishNoEarlierThan:
            if (earlyFinish.isBefore(cd)) {
                const newFinish = cd;
                const newStart = subtractWorkingMinutesFn(newFinish, durationMinutes);
                return {
                    earlyStart: newStart,
                    earlyFinish: newFinish,
                    warnings,
                };
            }
            return { earlyStart, earlyFinish, warnings };
        case ConstraintType.FinishNoLaterThan:
            if (earlyFinish.isAfter(cd)) {
                warnings.push({
                    taskId,
                    message: `Task cannot finish by constraint date ${constraintDate} due to dependencies. Computed finish: ${earlyFinish.toISOString()}.`,
                    severity: ValidationSeverity.Warning,
                });
            }
            return { earlyStart, earlyFinish, warnings };
        default:
            return { earlyStart, earlyFinish, warnings };
    }
}
/**
 * Apply constraint during backward pass for late dates.
 */
export function applyConstraintBackwardPass(constraintType, constraintDate, lateStart, lateFinish, durationMinutes, addWorkingMinutesFn, subtractWorkingMinutesFn) {
    if (constraintType === ConstraintType.ASAP ||
        constraintType === ConstraintType.ALAP ||
        !constraintDate) {
        return { lateStart, lateFinish };
    }
    const cd = dayjs.utc(constraintDate);
    switch (constraintType) {
        case ConstraintType.MustStartOn:
            return {
                lateStart: cd,
                lateFinish: addWorkingMinutesFn(cd, durationMinutes),
            };
        case ConstraintType.MustFinishOn:
            return {
                lateStart: subtractWorkingMinutesFn(cd, durationMinutes),
                lateFinish: cd,
            };
        case ConstraintType.FinishNoLaterThan:
            if (lateFinish.isAfter(cd)) {
                return {
                    lateFinish: cd,
                    lateStart: subtractWorkingMinutesFn(cd, durationMinutes),
                };
            }
            return { lateStart, lateFinish };
        case ConstraintType.StartNoLaterThan:
            if (lateStart.isAfter(cd)) {
                return {
                    lateStart: cd,
                    lateFinish: addWorkingMinutesFn(cd, durationMinutes),
                };
            }
            return { lateStart, lateFinish };
        default:
            return { lateStart, lateFinish };
    }
}
