// @schedulesync/engine/dist/fields/formulaEvaluator.js
/**
 * Formula Evaluator — safe expression parser and evaluator for custom fields.
 *
 * Supports: arithmetic (+, -, *, /), comparison (>, <, >=, <=, ==, !=),
 * logical (&&, ||, !), ternary (condition ? a : b), field references [FieldName],
 * and built-in functions (IIf, Switch, Abs, Int, Round, Now, ProjDateAdd, etc.).
 *
 * Uses a recursive-descent parser — NO eval() or Function() constructor.
 */
function tokenize(expr) {
    const tokens = [];
    let i = 0;
    while (i < expr.length) {
        const ch = expr[i];
        if (/\s/.test(ch)) {
            i++;
            continue;
        }
        // Number literal
        if (/\d/.test(ch) || (ch === '.' && i + 1 < expr.length && /\d/.test(expr[i + 1]))) {
            let num = '';
            while (i < expr.length && /[\d.]/.test(expr[i]))
                num += expr[i++];
            tokens.push({ type: 'number', value: num });
            continue;
        }
        // String literal
        if (ch === '"' || ch === "'") {
            const quote = ch;
            let str = '';
            i++;
            while (i < expr.length && expr[i] !== quote) {
                if (expr[i] === '\\') {
                    i++;
                    str += expr[i] ?? '';
                }
                else {
                    str += expr[i];
                }
                i++;
            }
            i++; // skip closing quote
            tokens.push({ type: 'string', value: str });
            continue;
        }
        // Field reference [FieldName]
        if (ch === '[') {
            let field = '';
            i++;
            while (i < expr.length && expr[i] !== ']')
                field += expr[i++];
            i++; // skip ]
            tokens.push({ type: 'field', value: field });
            continue;
        }
        // Two-char operators
        if (i + 1 < expr.length) {
            const two = expr.slice(i, i + 2);
            if (['>=', '<=', '==', '!=', '&&', '||'].includes(two)) {
                tokens.push({ type: 'op', value: two });
                i += 2;
                continue;
            }
        }
        // Single-char operators and punctuation
        if ('+-*/><='.includes(ch)) {
            tokens.push({ type: 'op', value: ch });
            i++;
            continue;
        }
        if (ch === '!') {
            tokens.push({ type: 'not', value: '!' });
            i++;
            continue;
        }
        if (ch === '(') {
            tokens.push({ type: 'lparen', value: '(' });
            i++;
            continue;
        }
        if (ch === ')') {
            tokens.push({ type: 'rparen', value: ')' });
            i++;
            continue;
        }
        if (ch === ',') {
            tokens.push({ type: 'comma', value: ',' });
            i++;
            continue;
        }
        if (ch === '?') {
            tokens.push({ type: 'question', value: '?' });
            i++;
            continue;
        }
        if (ch === ':') {
            tokens.push({ type: 'colon', value: ':' });
            i++;
            continue;
        }
        // Identifier (function name or boolean literal)
        if (/[a-zA-Z_]/.test(ch)) {
            let ident = '';
            while (i < expr.length && /[a-zA-Z0-9_]/.test(expr[i]))
                ident += expr[i++];
            tokens.push({ type: 'ident', value: ident });
            continue;
        }
        // Skip unknown characters
        i++;
    }
    tokens.push({ type: 'eof', value: '' });
    return tokens;
}
// ---------------------------------------------------------------------------
// Recursive-descent parser → FormulaNode AST
// ---------------------------------------------------------------------------
class Parser {
    tokens;
    pos = 0;
    constructor(tokens) { this.tokens = tokens; }
    peek() { return this.tokens[this.pos] ?? { type: 'eof', value: '' }; }
    advance() { return this.tokens[this.pos++] ?? { type: 'eof', value: '' }; }
    expect(type) {
        const t = this.advance();
        if (t.type !== type)
            throw new Error(`Expected ${type}, got ${t.type} "${t.value}"`);
        return t;
    }
    parse() {
        const node = this.parseTernary();
        if (this.peek().type !== 'eof')
            throw new Error(`Unexpected token: ${this.peek().value}`);
        return node;
    }
    parseTernary() {
        let node = this.parseOr();
        if (this.peek().type === 'question') {
            this.advance();
            const whenTrue = this.parseTernary();
            this.expect('colon');
            const whenFalse = this.parseTernary();
            node = { type: 'ternary', condition: node, whenTrue, whenFalse };
        }
        return node;
    }
    parseOr() {
        let node = this.parseAnd();
        while (this.peek().value === '||') {
            this.advance();
            node = { type: 'binary', op: '||', left: node, right: this.parseAnd() };
        }
        return node;
    }
    parseAnd() {
        let node = this.parseComparison();
        while (this.peek().value === '&&') {
            this.advance();
            node = { type: 'binary', op: '&&', left: node, right: this.parseComparison() };
        }
        return node;
    }
    parseComparison() {
        let node = this.parseAddSub();
        const ops = ['>', '<', '>=', '<=', '==', '!='];
        while (ops.includes(this.peek().value)) {
            const op = this.advance().value;
            node = { type: 'binary', op, left: node, right: this.parseAddSub() };
        }
        return node;
    }
    parseAddSub() {
        let node = this.parseMulDiv();
        while (this.peek().value === '+' || this.peek().value === '-') {
            const op = this.advance().value;
            node = { type: 'binary', op, left: node, right: this.parseMulDiv() };
        }
        return node;
    }
    parseMulDiv() {
        let node = this.parseUnary();
        while (this.peek().value === '*' || this.peek().value === '/') {
            const op = this.advance().value;
            node = { type: 'binary', op, left: node, right: this.parseUnary() };
        }
        return node;
    }
    parseUnary() {
        if (this.peek().value === '-') {
            this.advance();
            return { type: 'unary', op: '-', operand: this.parseUnary() };
        }
        if (this.peek().type === 'not') {
            this.advance();
            return { type: 'unary', op: '!', operand: this.parseUnary() };
        }
        return this.parsePrimary();
    }
    parsePrimary() {
        const t = this.peek();
        if (t.type === 'number') {
            this.advance();
            return { type: 'literal', value: parseFloat(t.value) };
        }
        if (t.type === 'string') {
            this.advance();
            return { type: 'literal', value: t.value };
        }
        if (t.type === 'field') {
            this.advance();
            return { type: 'field', name: t.value };
        }
        if (t.type === 'ident') {
            const lower = t.value.toLowerCase();
            if (lower === 'true') {
                this.advance();
                return { type: 'literal', value: true };
            }
            if (lower === 'false') {
                this.advance();
                return { type: 'literal', value: false };
            }
            // Function call
            this.advance();
            if (this.peek().type === 'lparen') {
                this.advance();
                const args = [];
                if (this.peek().type !== 'rparen') {
                    args.push(this.parseTernary());
                    while (this.peek().type === 'comma') {
                        this.advance();
                        args.push(this.parseTernary());
                    }
                }
                this.expect('rparen');
                return { type: 'func', name: t.value, args };
            }
            // Bare identifier treated as field reference
            return { type: 'field', name: t.value };
        }
        if (t.type === 'lparen') {
            this.advance();
            const node = this.parseTernary();
            this.expect('rparen');
            return node;
        }
        throw new Error(`Unexpected token: ${t.type} "${t.value}"`);
    }
}
// ---------------------------------------------------------------------------
// Compile (parse string → AST)
// ---------------------------------------------------------------------------
export function compileFormula(expression) {
    const tokens = tokenize(expression);
    return new Parser(tokens).parse();
}
// ---------------------------------------------------------------------------
// Evaluate AST against a task record
// ---------------------------------------------------------------------------
/** Known field mapping from display name to task property. */
const FIELD_MAP = {
    'Duration': 'durationMinutes',
    'Start': 'start',
    'Finish': 'finish',
    '% Complete': 'percentComplete',
    'Percent Complete': 'percentComplete',
    'Cost': 'cost',
    'Work': 'work',
    'Actual Cost': 'actualCost',
    'Actual Work': 'actualWork',
    'Actual Start': 'actualStart',
    'Actual Finish': 'actualFinish',
    'Remaining Cost': 'remainingCost',
    'Remaining Work': 'remainingWork',
    'Fixed Cost': 'fixedCost',
    'BCWS': 'bcws',
    'BCWP': 'bcwp',
    'ACWP': 'acwp',
    'SV': 'sv',
    'CV': 'cv',
    'SPI': 'spi',
    'CPI': 'cpi',
    'EAC': 'eac',
    'VAC': 'vac',
    'Total Slack': 'totalSlackMinutes',
    'Free Slack': 'freeSlackMinutes',
    'Name': 'name',
    'Notes': 'notes',
    'Critical': 'isCritical',
    'Is Critical': 'isCritical',
};
function resolveField(task, fieldName) {
    const mapped = FIELD_MAP[fieldName] ?? fieldName;
    // Try exact property name, then camelCase version
    if (mapped in task)
        return task[mapped];
    const camel = mapped.charAt(0).toLowerCase() + mapped.slice(1);
    if (camel in task)
        return task[camel];
    return 0;
}
function toNumber(v) {
    if (typeof v === 'number')
        return v;
    if (typeof v === 'boolean')
        return v ? 1 : 0;
    if (typeof v === 'string') {
        const n = parseFloat(v);
        return isNaN(n) ? 0 : n;
    }
    return 0;
}
function toBool(v) {
    if (typeof v === 'boolean')
        return v;
    if (typeof v === 'number')
        return v !== 0;
    if (typeof v === 'string')
        return v.length > 0;
    return false;
}
/** Built-in function implementations. */
const BUILTINS = {
    iif: ([cond, t, f]) => toBool(cond) ? t : f,
    abs: ([v]) => Math.abs(toNumber(v)),
    int: ([v]) => Math.floor(toNumber(v)),
    round: ([v, d]) => {
        const digits = d !== undefined ? toNumber(d) : 0;
        const factor = Math.pow(10, digits);
        return Math.round(toNumber(v) * factor) / factor;
    },
    max: (args) => Math.max(...args.map(toNumber)),
    min: (args) => Math.min(...args.map(toNumber)),
    now: () => new Date().toISOString(),
    len: ([v]) => typeof v === 'string' ? v.length : 0,
    mid: ([s, start, len]) => typeof s === 'string' ? s.substring(toNumber(start), toNumber(start) + toNumber(len)) : '',
    left: ([s, n]) => typeof s === 'string' ? s.substring(0, toNumber(n)) : '',
    right: ([s, n]) => typeof s === 'string' ? s.slice(-toNumber(n)) : '',
    upper: ([s]) => typeof s === 'string' ? s.toUpperCase() : String(s).toUpperCase(),
    lower: ([s]) => typeof s === 'string' ? s.toLowerCase() : String(s).toLowerCase(),
    trim: ([s]) => typeof s === 'string' ? s.trim() : '',
    val: ([s]) => typeof s === 'string' ? parseFloat(s) || 0 : toNumber(s),
    str: ([v]) => String(v),
    switch: (args) => {
        // Switch(expr, val1, result1, val2, result2, ..., defaultResult)
        if (args.length < 3)
            return '';
        const expr = args[0];
        for (let i = 1; i < args.length - 1; i += 2) {
            if (expr === args[i])
                return args[i + 1];
        }
        // If odd number of remaining args, last is default
        return args.length % 2 === 0 ? args[args.length - 1] : '';
    },
};
export function evaluateFormula(node, task) {
    switch (node.type) {
        case 'literal':
            return node.value;
        case 'field':
            return resolveField(task, node.name);
        case 'unary': {
            const operand = evaluateFormula(node.operand, task);
            if (node.op === '-')
                return -toNumber(operand);
            if (node.op === '!')
                return !toBool(operand);
            return operand;
        }
        case 'binary': {
            const left = evaluateFormula(node.left, task);
            const right = evaluateFormula(node.right, task);
            switch (node.op) {
                case '+': return typeof left === 'string' || typeof right === 'string'
                    ? String(left) + String(right) : toNumber(left) + toNumber(right);
                case '-': return toNumber(left) - toNumber(right);
                case '*': return toNumber(left) * toNumber(right);
                case '/': {
                    const d = toNumber(right);
                    return d === 0 ? 0 : toNumber(left) / d;
                }
                case '>': return toNumber(left) > toNumber(right);
                case '<': return toNumber(left) < toNumber(right);
                case '>=': return toNumber(left) >= toNumber(right);
                case '<=': return toNumber(left) <= toNumber(right);
                case '==': return left === right;
                case '!=': return left !== right;
                case '&&': return toBool(left) && toBool(right);
                case '||': return toBool(left) || toBool(right);
            }
            return 0;
        }
        case 'ternary': {
            const cond = evaluateFormula(node.condition, task);
            return toBool(cond)
                ? evaluateFormula(node.whenTrue, task)
                : evaluateFormula(node.whenFalse, task);
        }
        case 'func': {
            const fn = BUILTINS[node.name.toLowerCase()];
            if (!fn)
                throw new Error(`Unknown function: ${node.name}`);
            const args = node.args.map((a) => evaluateFormula(a, task));
            return fn(args);
        }
    }
}
