// @schedulesync/engine/dist/parsers/jsonUpdateSource.js
/**
 * JSON update source — ported from C# JsonUpdateSource.
 * Expects a JSON array of update objects. Case-insensitive property names.
 */
import { createParseResult } from '../types.js';
function makeEmptyUpdate() {
    return {
        externalKey: null,
        uniqueId: null,
        name: null,
        newStart: null,
        newFinish: null,
        newDurationMinutes: null,
        newPercentComplete: null,
        newConstraintType: null,
        newConstraintDate: null,
        newDeadline: null,
        notesAppend: null,
        allowConstraintOverride: false,
        isNew: false,
        metadata: {},
    };
}
/** Normalize all top-level keys to lower case. */
function normalizeKeys(obj) {
    const result = {};
    for (const key of Object.keys(obj)) {
        result[key.toLowerCase()] = obj[key];
    }
    return result;
}
function asString(val) {
    if (val == null)
        return null;
    if (typeof val === 'string')
        return val;
    return String(val);
}
function asInt(val) {
    if (val == null)
        return null;
    if (typeof val === 'number')
        return Math.round(val);
    if (typeof val === 'string') {
        const n = parseInt(val, 10);
        return isNaN(n) ? null : n;
    }
    return null;
}
function asFloat(val) {
    if (val == null)
        return null;
    if (typeof val === 'number')
        return val;
    if (typeof val === 'string') {
        const n = parseFloat(val);
        return isNaN(n) ? null : n;
    }
    return null;
}
function asDate(val) {
    if (val == null)
        return null;
    if (typeof val === 'string') {
        const d = new Date(val);
        return isNaN(d.getTime()) ? null : d.toISOString();
    }
    return null;
}
function asBool(val) {
    if (val == null)
        return false;
    if (typeof val === 'boolean')
        return val;
    if (typeof val === 'string')
        return val.toLowerCase() === 'true' || val === '1';
    if (typeof val === 'number')
        return val !== 0;
    return false;
}
export class JsonUpdateSource {
    parse(content) {
        const updates = [];
        const errors = [];
        if (!content || content.trim() === '') {
            errors.push({ rowNumber: null, fieldName: null, message: 'Content is empty.' });
            return createParseResult(updates, errors);
        }
        let parsed;
        try {
            parsed = JSON.parse(content);
        }
        catch {
            errors.push({
                rowNumber: null,
                fieldName: null,
                message: 'Invalid JSON format.',
            });
            return createParseResult(updates, errors);
        }
        if (!Array.isArray(parsed)) {
            errors.push({
                rowNumber: null,
                fieldName: null,
                message: 'JSON must be an array of update objects.',
            });
            return createParseResult(updates, errors);
        }
        for (let i = 0; i < parsed.length; i++) {
            const item = parsed[i];
            const rowNum = i + 1;
            if (typeof item !== 'object' || item === null || Array.isArray(item)) {
                errors.push({
                    rowNumber: rowNum,
                    fieldName: null,
                    message: 'Array element must be an object.',
                });
                continue;
            }
            const obj = normalizeKeys(item);
            const update = makeEmptyUpdate();
            let hasError = false;
            // UniqueId
            if ('uniqueid' in obj) {
                const v = asInt(obj['uniqueid']);
                if (v === null && obj['uniqueid'] != null) {
                    errors.push({ rowNumber: rowNum, fieldName: 'uniqueId', message: `Invalid integer: ${obj['uniqueid']}` });
                    hasError = true;
                }
                else {
                    update.uniqueId = v;
                }
            }
            // ExternalKey
            if ('externalkey' in obj)
                update.externalKey = asString(obj['externalkey']);
            // Name
            if ('name' in obj)
                update.name = asString(obj['name']);
            // NewStart
            if ('newstart' in obj) {
                const v = asDate(obj['newstart']);
                if (v === null && obj['newstart'] != null) {
                    errors.push({ rowNumber: rowNum, fieldName: 'newStart', message: `Invalid date: ${obj['newstart']}` });
                    hasError = true;
                }
                else {
                    update.newStart = v;
                }
            }
            // NewFinish
            if ('newfinish' in obj) {
                const v = asDate(obj['newfinish']);
                if (v === null && obj['newfinish'] != null) {
                    errors.push({ rowNumber: rowNum, fieldName: 'newFinish', message: `Invalid date: ${obj['newfinish']}` });
                    hasError = true;
                }
                else {
                    update.newFinish = v;
                }
            }
            // NewDurationMinutes
            if ('newdurationminutes' in obj) {
                const v = asFloat(obj['newdurationminutes']);
                if (v === null && obj['newdurationminutes'] != null) {
                    errors.push({ rowNumber: rowNum, fieldName: 'newDurationMinutes', message: `Invalid number: ${obj['newdurationminutes']}` });
                    hasError = true;
                }
                else {
                    update.newDurationMinutes = v;
                }
            }
            // NewPercentComplete
            if ('newpercentcomplete' in obj) {
                const v = asInt(obj['newpercentcomplete']);
                if (v === null && obj['newpercentcomplete'] != null) {
                    errors.push({ rowNumber: rowNum, fieldName: 'newPercentComplete', message: `Invalid integer: ${obj['newpercentcomplete']}` });
                    hasError = true;
                }
                else {
                    update.newPercentComplete = v;
                }
            }
            // NewConstraintType
            if ('newconstrainttype' in obj) {
                const v = asInt(obj['newconstrainttype']);
                if (v === null && obj['newconstrainttype'] != null) {
                    errors.push({ rowNumber: rowNum, fieldName: 'newConstraintType', message: `Invalid integer: ${obj['newconstrainttype']}` });
                    hasError = true;
                }
                else {
                    update.newConstraintType = v;
                }
            }
            // NewConstraintDate
            if ('newconstraintdate' in obj) {
                const v = asDate(obj['newconstraintdate']);
                if (v === null && obj['newconstraintdate'] != null) {
                    errors.push({ rowNumber: rowNum, fieldName: 'newConstraintDate', message: `Invalid date: ${obj['newconstraintdate']}` });
                    hasError = true;
                }
                else {
                    update.newConstraintDate = v;
                }
            }
            // NewDeadline
            if ('newdeadline' in obj) {
                const v = asDate(obj['newdeadline']);
                if (v === null && obj['newdeadline'] != null) {
                    errors.push({ rowNumber: rowNum, fieldName: 'newDeadline', message: `Invalid date: ${obj['newdeadline']}` });
                    hasError = true;
                }
                else {
                    update.newDeadline = v;
                }
            }
            // NotesAppend
            if ('notesappend' in obj)
                update.notesAppend = asString(obj['notesappend']);
            // AllowConstraintOverride
            if ('allowconstraintoverride' in obj) {
                update.allowConstraintOverride = asBool(obj['allowconstraintoverride']);
            }
            if (!hasError) {
                updates.push(update);
            }
        }
        return createParseResult(updates, errors);
    }
}
