// @schedulesync/engine/dist/cost/effortDriven.js
/**
 * Effort-Driven Scheduling — recalculates work/duration/units
 * based on task mode when assignments change.
 *
 * MS Project task equation: Work = Duration × Units
 *
 * When a task is effort-driven and assignments change:
 * - FixedWork: duration adjusts (work stays same, units change)
 * - FixedDuration: units adjust (duration stays, work redistributed)
 * - FixedUnits: duration adjusts (units stay, work redistributed)
 */
import { TaskMode } from '../types.js';
/**
 * Recalculate duration and assignment work/units after an assignment change
 * for an effort-driven task.
 *
 * @param task - The task being modified
 * @param assignments - Current assignments (including any newly added)
 * @param previousTotalUnits - Sum of units before the change
 */
export function recalculateEffortDriven(task, assignments, previousTotalUnits) {
    const mode = task.taskMode ?? TaskMode.FixedUnits;
    const totalWork = task.work ?? task.durationMinutes * (previousTotalUnits || 1);
    const newTotalUnits = assignments.reduce((s, a) => s + a.units, 0);
    if (assignments.length === 0 || newTotalUnits === 0) {
        return {
            durationMinutes: task.durationMinutes,
            assignments: [],
        };
    }
    switch (mode) {
        case 'fixedWork': {
            // Work stays the same. Duration = Work / TotalUnits.
            const newDuration = totalWork / newTotalUnits;
            return {
                durationMinutes: newDuration,
                assignments: assignments.map((a) => ({
                    id: a.id,
                    units: a.units,
                    workMinutes: totalWork * (a.units / newTotalUnits),
                })),
            };
        }
        case 'fixedDuration': {
            // Duration stays. Work redistributed based on units proportion.
            const newWork = task.durationMinutes * newTotalUnits;
            return {
                durationMinutes: task.durationMinutes,
                assignments: assignments.map((a) => ({
                    id: a.id,
                    units: a.units,
                    workMinutes: newWork * (a.units / newTotalUnits),
                })),
            };
        }
        case 'fixedUnits':
        default: {
            // Units stay the same. Duration = Work / TotalUnits.
            const newDuration = totalWork / newTotalUnits;
            return {
                durationMinutes: newDuration,
                assignments: assignments.map((a) => ({
                    id: a.id,
                    units: a.units,
                    workMinutes: totalWork * (a.units / newTotalUnits),
                })),
            };
        }
    }
}
/**
 * Calculate remaining work and duration for a task based on actuals.
 */
export function calculateRemaining(task) {
    const pctComplete = (task.percentComplete ?? 0) / 100;
    const totalDuration = task.durationMinutes;
    const totalWork = task.work ?? totalDuration;
    return {
        remainingDuration: totalDuration * (1 - pctComplete),
        remainingWork: totalWork * (1 - pctComplete),
    };
}
