// @schedulesync/engine/dist/cost/costEngine.js
/**
 * Cost Engine — calculates task/assignment costs based on resource rates,
 * accrual methods, and rate tables.
 *
 * Implements MS Project's cost calculation model:
 * - Work resources: cost = (work hours) × (rate) + costPerUse
 * - Material resources: cost = units × rate
 * - Cost resources: cost = assignment cost (no rate)
 * - Fixed costs accrue based on accrual method (start/end/prorated)
 */
// ---------------------------------------------------------------------------
// Rate lookup
// ---------------------------------------------------------------------------
/**
 * Get the applicable rate entry from a cost rate table at a given date.
 * Returns the entry whose effectiveDate is on or before the given date.
 */
export function getApplicableRate(rateTables, tableIndex, date) {
    const defaultEntry = {
        effectiveDate: '1984-01-01T00:00:00Z',
        standardRate: 0,
        overtimeRate: 0,
        costPerUse: 0,
    };
    if (!rateTables || rateTables.length === 0)
        return defaultEntry;
    const tableName = String.fromCharCode(65 + (tableIndex || 0)); // 0→A, 1→B, ...
    const table = rateTables.find((t) => t.name === tableName);
    if (!table || table.entries.length === 0)
        return defaultEntry;
    const d = new Date(date).getTime();
    let best = table.entries[0];
    for (const entry of table.entries) {
        if (new Date(entry.effectiveDate).getTime() <= d)
            best = entry;
    }
    return best;
}
export function calculateAssignmentCost(assignment, resource) {
    const rateEntry = getApplicableRate(resource.costRateTable, assignment.costRateTableIndex ?? 0, assignment.start ?? new Date().toISOString());
    if (resource.type === 'cost') {
        // Cost resource — the assignment's actual/remaining cost is user-entered
        return {
            cost: (assignment.actualCost ?? 0) + (assignment.remainingCost ?? 0),
            actualCost: assignment.actualCost ?? 0,
            remainingCost: assignment.remainingCost ?? 0,
        };
    }
    if (resource.type === 'material') {
        const unitCost = (resource.standardRate ?? 0) * assignment.units;
        const perUse = resource.costPerUse ?? 0;
        return {
            cost: unitCost + perUse,
            actualCost: unitCost * ((assignment.actualWork ?? 0) / Math.max(assignment.workMinutes, 1)),
            remainingCost: unitCost * ((assignment.remainingWork ?? 0) / Math.max(assignment.workMinutes, 1)),
        };
    }
    // Work resource
    const workHours = assignment.workMinutes / 60;
    const actualWorkHours = (assignment.actualWork ?? 0) / 60;
    const remainingWorkHours = (assignment.remainingWork ?? 0) / 60;
    const cost = workHours * rateEntry.standardRate + rateEntry.costPerUse;
    const actualCost = actualWorkHours * rateEntry.standardRate +
        (actualWorkHours > 0 ? rateEntry.costPerUse : 0);
    const remainingCost = remainingWorkHours * rateEntry.standardRate;
    return { cost, actualCost, remainingCost };
}
/**
 * Calculate total cost for a task based on its assignments and fixed cost.
 */
export function calculateTaskCost(task, assignments, resources) {
    const resourceMap = new Map(resources.map((r) => [r.id, r]));
    let assignmentCost = 0;
    let assignmentActualCost = 0;
    let assignmentRemainingCost = 0;
    let totalWork = 0;
    let totalActualWork = 0;
    let totalRemainingWork = 0;
    for (const a of assignments) {
        const resource = resourceMap.get(a.resourceId);
        if (!resource)
            continue;
        const costResult = calculateAssignmentCost(a, resource);
        assignmentCost += costResult.cost;
        assignmentActualCost += costResult.actualCost;
        assignmentRemainingCost += costResult.remainingCost;
        totalWork += a.workMinutes;
        totalActualWork += a.actualWork ?? 0;
        totalRemainingWork += a.remainingWork ?? 0;
    }
    const fixedCost = task.fixedCost ?? 0;
    return {
        cost: assignmentCost + fixedCost,
        actualCost: assignmentActualCost + applyFixedCostAccrual(fixedCost, task),
        remainingCost: assignmentRemainingCost + applyRemainingFixedCost(fixedCost, task),
        work: totalWork,
        actualWork: totalActualWork,
        remainingWork: totalRemainingWork,
    };
}
function applyFixedCostAccrual(fixedCost, task) {
    const accrual = task.fixedCostAccrual ?? 'prorated';
    const pct = (task.percentComplete ?? 0) / 100;
    switch (accrual) {
        case 'start': return fixedCost; // fully accrued at start
        case 'end': return task.percentComplete === 100 ? fixedCost : 0;
        case 'prorated': return fixedCost * pct;
        default: return fixedCost * pct;
    }
}
function applyRemainingFixedCost(fixedCost, task) {
    return fixedCost - applyFixedCostAccrual(fixedCost, task);
}
// ---------------------------------------------------------------------------
// Project-wide cost rollup
// ---------------------------------------------------------------------------
/**
 * Calculate costs for all tasks in a project.
 * Returns updated tasks with cost fields populated.
 */
export function rollupProjectCosts(project) {
    const assignmentsByTask = new Map();
    for (const a of project.assignments) {
        const list = assignmentsByTask.get(a.taskId) ?? [];
        list.push(a);
        assignmentsByTask.set(a.taskId, list);
    }
    // Leaf tasks first
    for (const task of project.tasks) {
        if (task.type === 'summary')
            continue;
        const taskAssignments = assignmentsByTask.get(task.id) ?? [];
        const result = calculateTaskCost(task, taskAssignments, project.resources);
        task.cost = result.cost;
        task.actualCost = result.actualCost;
        task.remainingCost = result.remainingCost;
        task.work = result.work;
        task.actualWork = result.actualWork;
        task.remainingWork = result.remainingWork;
    }
    // Summary rollup (bottom-up)
    const childMap = new Map();
    for (const task of project.tasks) {
        const list = childMap.get(task.parentId ?? null) ?? [];
        list.push(task);
        childMap.set(task.parentId ?? null, list);
    }
    function rollupSummary(parentId) {
        const children = childMap.get(parentId);
        if (!children)
            return;
        for (const child of children) {
            if (child.type === 'summary') {
                rollupSummary(child.id);
                const descendants = childMap.get(child.id) ?? [];
                child.cost = descendants.reduce((s, c) => s + (c.cost ?? 0), 0);
                child.actualCost = descendants.reduce((s, c) => s + (c.actualCost ?? 0), 0);
                child.remainingCost = descendants.reduce((s, c) => s + (c.remainingCost ?? 0), 0);
                child.work = descendants.reduce((s, c) => s + (c.work ?? 0), 0);
                child.actualWork = descendants.reduce((s, c) => s + (c.actualWork ?? 0), 0);
                child.remainingWork = descendants.reduce((s, c) => s + (c.remainingWork ?? 0), 0);
            }
        }
    }
    rollupSummary(null);
    return project.tasks;
}
