// @schedulesync/frontend/src/components/dialogs/PrintPreviewDialog.tsx
/**
 * PrintPreviewDialog — print preview with page setup options.
 * Mirrors MS Project's File > Print Preview.
 */

import React, { useState, useCallback, useRef } from 'react';
import {
  Dialog, DialogTitle, DialogContent, DialogActions, Button,
  Box, Typography, FormControl, InputLabel, Select, MenuItem,
  TextField, FormControlLabel, Checkbox, Divider,
} from '@mui/material';
import { useUIStore, useProjectStore } from '../../stores';

type Orientation = 'portrait' | 'landscape';
type PageSize = 'letter' | 'legal' | 'a4' | 'tabloid';

const PrintPreviewDialog: React.FC = () => {
  const openDialog = useUIStore((s) => s.openDialog);
  const closeDialog = useUIStore((s) => s.closeDialog);
  const open = openDialog === 'printPreview';
  const tasks = useProjectStore((s) => s.tasks);
  const activeView = useUIStore((s) => s.activeView);

  const [orientation, setOrientation] = useState<Orientation>('landscape');
  const [pageSize, setPageSize] = useState<PageSize>('letter');
  const [fitToPage, setFitToPage] = useState(true);
  const [includeHeader, setIncludeHeader] = useState(true);
  const [includeFooter, setIncludeFooter] = useState(true);
  const [headerText, setHeaderText] = useState('');
  const [footerText, setFooterText] = useState('Page &P of &N');

  const handlePrint = useCallback(() => {
    const style = document.createElement('style');
    style.setAttribute('media', 'print');
    style.textContent = `
      @page {
        size: ${pageSize} ${orientation};
        margin: 0.5in;
      }
      body * { visibility: hidden; }
      .print-area, .print-area * { visibility: visible; }
      .print-area { position: absolute; left: 0; top: 0; width: 100%; }
      ${fitToPage ? '.print-area { transform: scale(0.8); transform-origin: top left; }' : ''}
    `;
    document.head.appendChild(style);
    window.print();
    document.head.removeChild(style);
    closeDialog();
  }, [orientation, pageSize, fitToPage, closeDialog]);

  if (!open) return null;

  return (
    <Dialog open={open} onClose={closeDialog} maxWidth="md" fullWidth>
      <DialogTitle>Print Preview</DialogTitle>
      <DialogContent>
        <Box sx={{ display: 'flex', gap: 3 }}>
          {/* Settings panel */}
          <Box sx={{ width: 250, flexShrink: 0, display: 'flex', flexDirection: 'column', gap: 2, mt: 1 }}>
            <Typography variant="subtitle2">Page Setup</Typography>

            <FormControl size="small" fullWidth>
              <InputLabel>Orientation</InputLabel>
              <Select value={orientation} label="Orientation" onChange={(e) => setOrientation(e.target.value as Orientation)}>
                <MenuItem value="portrait">Portrait</MenuItem>
                <MenuItem value="landscape">Landscape</MenuItem>
              </Select>
            </FormControl>

            <FormControl size="small" fullWidth>
              <InputLabel>Page Size</InputLabel>
              <Select value={pageSize} label="Page Size" onChange={(e) => setPageSize(e.target.value as PageSize)}>
                <MenuItem value="letter">Letter (8.5 x 11)</MenuItem>
                <MenuItem value="legal">Legal (8.5 x 14)</MenuItem>
                <MenuItem value="a4">A4</MenuItem>
                <MenuItem value="tabloid">Tabloid (11 x 17)</MenuItem>
              </Select>
            </FormControl>

            <FormControlLabel
              control={<Checkbox size="small" checked={fitToPage} onChange={(e) => setFitToPage(e.target.checked)} />}
              label={<Typography variant="body2">Fit to one page</Typography>}
            />

            <Divider />
            <Typography variant="subtitle2">Header / Footer</Typography>

            <FormControlLabel
              control={<Checkbox size="small" checked={includeHeader} onChange={(e) => setIncludeHeader(e.target.checked)} />}
              label={<Typography variant="body2">Include header</Typography>}
            />
            {includeHeader && (
              <TextField size="small" label="Header text" value={headerText} onChange={(e) => setHeaderText(e.target.value)} fullWidth />
            )}

            <FormControlLabel
              control={<Checkbox size="small" checked={includeFooter} onChange={(e) => setIncludeFooter(e.target.checked)} />}
              label={<Typography variant="body2">Include footer</Typography>}
            />
            {includeFooter && (
              <TextField size="small" label="Footer text" value={footerText} onChange={(e) => setFooterText(e.target.value)} fullWidth
                helperText="&P = page, &N = total pages, &D = date"
              />
            )}
          </Box>

          {/* Preview panel */}
          <Box
            sx={{
              flex: 1,
              border: '1px solid #CCC',
              borderRadius: 1,
              minHeight: 400,
              p: 2,
              backgroundColor: '#FAFAFA',
              overflow: 'auto',
              position: 'relative',
            }}
          >
            {/* Simulate a page preview */}
            <Box
              sx={{
                width: orientation === 'landscape' ? 680 : 500,
                minHeight: orientation === 'landscape' ? 440 : 640,
                mx: 'auto',
                backgroundColor: 'white',
                boxShadow: '0 2px 8px rgba(0,0,0,0.12)',
                p: 2,
                fontSize: '0.65rem',
              }}
            >
              {includeHeader && (
                <Box sx={{ borderBottom: '1px solid #999', pb: 0.5, mb: 1, fontSize: '0.6rem', color: '#666' }}>
                  {headerText || 'Project Schedule'}
                </Box>
              )}

              <Typography variant="caption" sx={{ fontWeight: 600, display: 'block', mb: 1 }}>
                {activeView.charAt(0).toUpperCase() + activeView.slice(1)} View — {tasks.length} tasks
              </Typography>

              {/* Mini task table */}
              <Box component="table" sx={{ width: '100%', borderCollapse: 'collapse', fontSize: '0.55rem' }}>
                <thead>
                  <tr>
                    <th style={{ textAlign: 'left', borderBottom: '1px solid #CCC', padding: '2px 4px' }}>#</th>
                    <th style={{ textAlign: 'left', borderBottom: '1px solid #CCC', padding: '2px 4px' }}>Task Name</th>
                    <th style={{ textAlign: 'right', borderBottom: '1px solid #CCC', padding: '2px 4px' }}>Duration</th>
                    <th style={{ textAlign: 'left', borderBottom: '1px solid #CCC', padding: '2px 4px' }}>Start</th>
                    <th style={{ textAlign: 'left', borderBottom: '1px solid #CCC', padding: '2px 4px' }}>Finish</th>
                  </tr>
                </thead>
                <tbody>
                  {tasks.slice(0, 20).map((t, i) => (
                    <tr key={t.id}>
                      <td style={{ padding: '1px 4px' }}>{i + 1}</td>
                      <td style={{ padding: '1px 4px', paddingLeft: (t.outlineLevel || 0) * 12 + 4 }}>{t.name}</td>
                      <td style={{ padding: '1px 4px', textAlign: 'right' }}>
                        {t.durationMinutes != null ? `${Math.round(t.durationMinutes / 480)}d` : ''}
                      </td>
                      <td style={{ padding: '1px 4px' }}>{t.start ? new Date(t.start).toLocaleDateString() : ''}</td>
                      <td style={{ padding: '1px 4px' }}>{t.finish ? new Date(t.finish).toLocaleDateString() : ''}</td>
                    </tr>
                  ))}
                  {tasks.length > 20 && (
                    <tr>
                      <td colSpan={5} style={{ padding: '2px 4px', color: '#999', fontStyle: 'italic' }}>
                        ...and {tasks.length - 20} more tasks
                      </td>
                    </tr>
                  )}
                </tbody>
              </Box>

              {includeFooter && (
                <Box sx={{ borderTop: '1px solid #999', pt: 0.5, mt: 2, fontSize: '0.55rem', color: '#666', textAlign: 'center' }}>
                  {footerText.replace('&P', '1').replace('&N', '1').replace('&D', new Date().toLocaleDateString())}
                </Box>
              )}
            </Box>
          </Box>
        </Box>
      </DialogContent>
      <DialogActions>
        <Button onClick={closeDialog}>Close</Button>
        <Button variant="contained" onClick={handlePrint}>Print</Button>
      </DialogActions>
    </Dialog>
  );
};

export default PrintPreviewDialog;
