// @schedulesync/backend/src/services/costService.ts
/**
 * Cost & Earned Value service — backend wrapper around engine cost/EV calculations.
 */

import { prisma } from '../db.js';
import type { ProjectData, Task, Assignment, Resource, Baseline, Calendar, Dependency, ProjectSettings } from '@schedulesync/engine';

async function loadProjectData(projectId: string): Promise<ProjectData> {
  const project = await prisma.project.findUniqueOrThrow({
    where: { id: projectId },
    include: {
      tasks: { orderBy: { sortOrder: 'asc' } },
      dependencies: true,
      calendars: { include: { exceptions: true } },
      resources: true,
    },
  });

  const assignments = await prisma.assignment.findMany({
    where: { task: { projectId } },
  });

  const baselines = await prisma.baseline.findMany({
    where: { task: { projectId } },
  });

  const calendars: Calendar[] = project.calendars.map((c) => ({
    id: c.id,
    name: c.name,
    workingDaysOfWeek: JSON.parse(c.workingDaysOfWeek),
    defaultWorkingHours: JSON.parse(c.defaultWorkingHours),
    exceptions: c.exceptions.map((e) => ({
      startDate: e.startDate,
      endDate: e.endDate,
      isWorking: e.isWorking,
      workingHours: e.workingHours ? JSON.parse(e.workingHours) : null,
    })),
  }));

  const settings: ProjectSettings = {
    id: project.id,
    name: project.name,
    startDate: project.startDate.toISOString(),
    finishDate: project.finishDate?.toISOString() ?? null,
    defaultCalendarId: project.defaultCalendarId,
    scheduleFrom: project.scheduleFrom as any,
    statusDate: project.statusDate?.toISOString() ?? null,
    currencySymbol: project.currencySymbol,
    minutesPerDay: project.minutesPerDay,
    minutesPerWeek: project.minutesPerWeek,
    daysPerMonth: project.daysPerMonth,
    defaultTaskType: project.defaultTaskType as any,
    defaultFixedCostAccrual: project.defaultFixedCostAccrual as any,
    honorConstraints: project.honorConstraints,
    newTasksEffortDriven: project.newTasksEffortDriven,
    autolink: project.autolink,
    criticalSlackLimit: project.criticalSlackLimit,
  };

  return {
    settings,
    tasks: project.tasks.map((t) => ({
      id: t.id,
      wbsCode: t.wbsCode,
      outlineLevel: t.outlineLevel,
      parentId: t.parentId,
      name: t.name,
      type: t.type as any,
      durationMinutes: t.durationMinutes,
      start: t.start.toISOString(),
      finish: t.finish.toISOString(),
      constraintType: t.constraintType,
      constraintDate: t.constraintDate?.toISOString() ?? null,
      calendarId: t.calendarId,
      percentComplete: t.percentComplete,
      isManuallyScheduled: t.isManuallyScheduled,
      isCritical: t.isCritical,
      totalSlackMinutes: t.totalSlackMinutes,
      freeSlackMinutes: t.freeSlackMinutes,
      earlyStart: t.earlyStart?.toISOString() ?? null,
      earlyFinish: t.earlyFinish?.toISOString() ?? null,
      lateStart: t.lateStart?.toISOString() ?? null,
      lateFinish: t.lateFinish?.toISOString() ?? null,
      deadline: t.deadline?.toISOString() ?? null,
      notes: t.notes,
      externalKey: t.externalKey,
      sortOrder: t.sortOrder,
      actualStart: t.actualStart?.toISOString() ?? null,
      actualFinish: t.actualFinish?.toISOString() ?? null,
      actualDurationMinutes: t.actualDurationMinutes,
      actualWork: t.actualWork,
      actualCost: t.actualCost,
      remainingDuration: t.remainingDuration,
      remainingWork: t.remainingWork,
      remainingCost: t.remainingCost,
      fixedCost: t.fixedCost,
      fixedCostAccrual: t.fixedCostAccrual as any,
      cost: t.cost,
      work: t.work,
      taskMode: t.taskMode as any,
      isEffortDriven: t.isEffortDriven,
      isActive: t.isActive,
      bcws: t.bcws,
      bcwp: t.bcwp,
      acwp: t.acwp,
      sv: t.sv,
      cv: t.cv,
      spi: t.spi,
      cpi: t.cpi,
      eac: t.eac,
      vac: t.vac,
      physicalPercentComplete: t.physicalPercentComplete,
      isSplit: t.isSplit,
    })) as Task[],
    dependencies: project.dependencies.map((d) => ({
      id: d.id,
      fromTaskId: d.fromTaskId,
      toTaskId: d.toTaskId,
      type: d.type as any,
      lagMinutes: d.lagMinutes,
    })) as Dependency[],
    calendars,
    resources: project.resources.map((r) => ({
      id: r.id,
      name: r.name,
      type: r.type as any,
      maxUnits: r.maxUnits,
      calendarId: r.calendarId,
      initials: r.initials,
      group: r.group,
      emailAddress: r.emailAddress,
      standardRate: r.standardRate,
      overtimeRate: r.overtimeRate,
      costPerUse: r.costPerUse,
      costRateTable: JSON.parse(r.costRateTable),
      accrueAt: r.accrueAt as any,
      budgetCost: r.budgetCost,
      budgetWork: r.budgetWork,
      isBudget: r.isBudget,
      isGeneric: r.isGeneric,
    })) as Resource[],
    assignments: assignments.map((a) => ({
      id: a.id,
      taskId: a.taskId,
      resourceId: a.resourceId,
      units: a.units,
      workMinutes: a.workMinutes,
      actualWork: a.actualWork,
      actualCost: a.actualCost,
      remainingWork: a.remainingWork,
      remainingCost: a.remainingCost,
      start: a.start?.toISOString() ?? null,
      finish: a.finish?.toISOString() ?? null,
      delay: a.delay,
      costRateTableIndex: a.costRateTableIndex,
      contour: a.contour as any,
    })) as Assignment[],
    baselines: baselines.map((b) => ({
      taskId: b.taskId,
      baselineIndex: b.baselineIndex,
      baselineStart: b.baselineStart.toISOString(),
      baselineFinish: b.baselineFinish.toISOString(),
      baselineDurationMinutes: b.baselineDurationMinutes,
      baselineWork: b.baselineWork,
      baselineCost: b.baselineCost,
    })) as Baseline[],
  };
}

export { loadProjectData };

import { rollupProjectCosts } from '@schedulesync/engine';

export async function computeCosts(projectId: string) {
  const projectData = await loadProjectData(projectId);
  rollupProjectCosts(projectData);

  // Persist computed costs back to DB
  for (const task of projectData.tasks) {
    await prisma.task.update({
      where: { id: task.id },
      data: {
        cost: task.cost ?? 0,
        actualCost: task.actualCost ?? 0,
        remainingCost: task.remainingCost ?? 0,
        work: task.work ?? 0,
        actualWork: task.actualWork ?? 0,
        remainingWork: task.remainingWork ?? 0,
      },
    });
  }

  return projectData.tasks;
}

import { computeProjectEV } from '@schedulesync/engine';

export async function computeEarnedValue(projectId: string) {
  const projectData = await loadProjectData(projectId);
  const evSummary = computeProjectEV(projectData);

  // Persist EV fields back to tasks
  for (const result of evSummary.taskResults) {
    await prisma.task.update({
      where: { id: result.taskId },
      data: {
        bcws: result.bcws,
        bcwp: result.bcwp,
        acwp: result.acwp,
        sv: result.sv,
        cv: result.cv,
        spi: result.spi,
        cpi: result.cpi,
        eac: result.eac,
        vac: result.vac,
      },
    });
  }

  return evSummary;
}
