// @schedulesync/frontend/src/stores/useAiStore.ts
/**
 * AI store — manages chat state, suggestions, and Ollama connection status.
 */

import { create } from 'zustand';
import { aiApi, type AiChatResponse } from '../api/client.js';

export interface ChatMessage {
  role: 'user' | 'assistant' | 'system';
  content: string;
}

interface AiState {
  // Panel visibility
  panelOpen: boolean;
  togglePanel: () => void;
  openPanel: () => void;
  closePanel: () => void;

  // Connection
  ollamaAvailable: boolean;
  availableModels: string[];
  checkHealth: () => Promise<void>;

  // Chat
  conversationId: string | null;
  messages: ChatMessage[];
  isLoading: boolean;
  error: string | null;

  sendMessage: (projectId: string, message: string) => Promise<void>;
  newConversation: () => void;
  loadConversation: (conversationId: string) => Promise<void>;

  // Quick suggestions
  suggestion: string | null;
  isSuggesting: boolean;
  requestSuggestion: (
    projectId: string,
    type: 'duration' | 'dependency' | 'name' | 'resource' | 'general',
    context: Record<string, unknown>,
  ) => Promise<void>;
  clearSuggestion: () => void;
}

export const useAiStore = create<AiState>((set, get) => ({
  panelOpen: false,
  togglePanel: () => set((s) => ({ panelOpen: !s.panelOpen })),
  openPanel: () => set({ panelOpen: true }),
  closePanel: () => set({ panelOpen: false }),

  ollamaAvailable: false,
  availableModels: [],
  checkHealth: async () => {
    try {
      const health = await aiApi.health();
      set({
        ollamaAvailable: health.ollamaAvailable,
        availableModels: health.models,
      });
    } catch {
      set({ ollamaAvailable: false, availableModels: [] });
    }
  },

  conversationId: null,
  messages: [],
  isLoading: false,
  error: null,

  sendMessage: async (projectId, message) => {
    const { conversationId, messages } = get();

    // Add user message immediately
    const userMsg: ChatMessage = { role: 'user', content: message };
    set({ messages: [...messages, userMsg], isLoading: true, error: null });

    try {
      const result: AiChatResponse = await aiApi.chat(
        projectId,
        message,
        conversationId ?? undefined,
      );

      const assistantMsg: ChatMessage = {
        role: 'assistant',
        content: result.response,
      };

      set((s) => ({
        conversationId: result.conversationId,
        messages: [...s.messages, assistantMsg],
        isLoading: false,
      }));
    } catch (err: unknown) {
      const errMsg =
        err instanceof Error ? err.message : 'Failed to get AI response';
      set((s) => ({
        isLoading: false,
        error: errMsg,
        // Remove the optimistic user message on failure
        messages: s.messages,
      }));
    }
  },

  newConversation: () =>
    set({ conversationId: null, messages: [], error: null }),

  loadConversation: async (conversationId) => {
    try {
      const conv = await aiApi.getConversation(conversationId);
      set({
        conversationId: conv.id,
        messages: conv.messages as ChatMessage[],
        error: null,
      });
    } catch (err: unknown) {
      const errMsg =
        err instanceof Error ? err.message : 'Failed to load conversation';
      set({ error: errMsg });
    }
  },

  suggestion: null,
  isSuggesting: false,
  requestSuggestion: async (projectId, type, context) => {
    set({ isSuggesting: true, suggestion: null });
    try {
      const result = await aiApi.suggest(projectId, type, context);
      set({ suggestion: result.suggestion, isSuggesting: false });
    } catch {
      set({ isSuggesting: false });
    }
  },
  clearSuggestion: () => set({ suggestion: null }),
}));
