// @schedulesync/frontend/src/stores/useUIStore.ts
/**
 * UI store — manages view state (zoom, scroll, dialogs, sidebar, etc.)
 */

import { create } from 'zustand';

export type GanttZoom = 'day' | 'week' | 'month' | 'quarter' | 'year';

export type ViewType =
  | 'gantt'
  | 'networkDiagram'
  | 'resourceSheet'
  | 'resourceUsage'
  | 'taskUsage'
  | 'calendar'
  | 'trackingGantt';

export type RibbonTab = 'task' | 'resource' | 'project' | 'view' | 'format';

export type DialogType =
  | 'none'
  | 'projectInfo'
  | 'taskInfo'
  | 'calendar'
  | 'resource'
  | 'baselineCapture'
  | 'importPreview'
  | 'importMspdi'
  | 'filter'
  | 'newProject';

export interface FilterCriteria {
  field: string;
  operator: 'eq' | 'ne' | 'gt' | 'lt' | 'contains' | 'between';
  value: unknown;
  value2?: unknown; // for 'between'
}

export interface GroupByOption {
  field: string;
  direction: 'asc' | 'desc';
}

interface UIState {
  // View
  activeView: ViewType;
  setActiveView: (view: ViewType) => void;

  // Ribbon
  activeRibbonTab: RibbonTab;
  setActiveRibbonTab: (tab: RibbonTab) => void;

  // Sidebar
  sidebarOpen: boolean;
  toggleSidebar: () => void;

  // Gantt
  ganttZoom: GanttZoom;
  setGanttZoom: (zoom: GanttZoom) => void;
  ganttScrollDate: string | null;
  scrollGanttTo: (date: string) => void;
  showBaseline: number | null; // baseline index to overlay
  setShowBaseline: (index: number | null) => void;

  // Grid
  gridSplitPercent: number;
  setGridSplitPercent: (pct: number) => void;
  showCriticalPath: boolean;
  toggleCriticalPath: () => void;

  // Dialogs
  openDialog: DialogType;
  dialogPayload: unknown;
  openDialogWith: (type: DialogType, payload?: unknown) => void;
  closeDialog: () => void;

  // Filtering & grouping
  filters: FilterCriteria[];
  setFilters: (filters: FilterCriteria[]) => void;
  groupBy: GroupByOption | null;
  setGroupBy: (group: GroupByOption | null) => void;

  // Snackbar
  snackbar: { message: string; severity: 'success' | 'info' | 'warning' | 'error' } | null;
  showSnackbar: (
    message: string,
    severity?: 'success' | 'info' | 'warning' | 'error',
  ) => void;
  clearSnackbar: () => void;
}

export const useUIStore = create<UIState>((set) => ({
  activeView: 'gantt',
  setActiveView: (view) => set({ activeView: view }),

  activeRibbonTab: 'task',
  setActiveRibbonTab: (tab) => set({ activeRibbonTab: tab }),

  sidebarOpen: true,
  toggleSidebar: () => set((s) => ({ sidebarOpen: !s.sidebarOpen })),

  ganttZoom: 'week',
  setGanttZoom: (zoom) => set({ ganttZoom: zoom }),
  ganttScrollDate: null,
  scrollGanttTo: (date) => set({ ganttScrollDate: date }),
  showBaseline: null,
  setShowBaseline: (index) => set({ showBaseline: index }),

  gridSplitPercent: 50,
  setGridSplitPercent: (pct) => set({ gridSplitPercent: pct }),
  showCriticalPath: false,
  toggleCriticalPath: () =>
    set((s) => ({ showCriticalPath: !s.showCriticalPath })),

  openDialog: 'none',
  dialogPayload: null,
  openDialogWith: (type, payload) =>
    set({ openDialog: type, dialogPayload: payload ?? null }),
  closeDialog: () => set({ openDialog: 'none', dialogPayload: null }),

  filters: [],
  setFilters: (filters) => set({ filters }),
  groupBy: null,
  setGroupBy: (group) => set({ groupBy: group }),

  snackbar: null,
  showSnackbar: (message, severity = 'info') =>
    set({ snackbar: { message, severity } }),
  clearSnackbar: () => set({ snackbar: null }),
}));
