// @schedulesync/frontend/src/api/client.ts
/**
 * API client — thin wrapper around fetch for the backend REST API.
 */

const BASE = '/api';

async function request<T>(
  path: string,
  init: RequestInit = {},
): Promise<T> {
  const res = await fetch(`${BASE}${path}`, {
    headers: {
      'Content-Type': 'application/json',
      ...(init.headers as Record<string, string>),
    },
    ...init,
  });

  if (res.status === 204) return undefined as unknown as T;

  if (!res.ok) {
    const body = await res.json().catch(() => ({ error: res.statusText }));
    throw new Error(body.error ?? `Request failed: ${res.status}`);
  }

  return res.json();
}

// ---------- Projects ----------

export const projectsApi = {
  list: () => request<any[]>('/projects'),
  get: (id: string) => request<any>(`/projects/${id}`),
  create: (data: { name: string; startDate: string }) =>
    request<any>('/projects', { method: 'POST', body: JSON.stringify(data) }),
  update: (id: string, data: Record<string, unknown>) =>
    request<any>(`/projects/${id}`, {
      method: 'PATCH',
      body: JSON.stringify(data),
    }),
  delete: (id: string) =>
    request<void>(`/projects/${id}`, { method: 'DELETE' }),
};

// ---------- Tasks ----------

export const tasksApi = {
  list: (projectId: string) =>
    request<any[]>(`/projects/${projectId}/tasks`),
  get: (projectId: string, taskId: string) =>
    request<any>(`/projects/${projectId}/tasks/${taskId}`),
  create: (projectId: string, data: Record<string, unknown>) =>
    request<any>(`/projects/${projectId}/tasks`, {
      method: 'POST',
      body: JSON.stringify(data),
    }),
  update: (projectId: string, taskId: string, data: Record<string, unknown>) =>
    request<any>(`/projects/${projectId}/tasks/${taskId}`, {
      method: 'PATCH',
      body: JSON.stringify(data),
    }),
  batchUpdate: (
    projectId: string,
    updates: { id: string; data: Record<string, unknown> }[],
    recalc = true,
  ) =>
    request<{ updated: number }>(`/projects/${projectId}/tasks/batch`, {
      method: 'POST',
      body: JSON.stringify({ updates, recalculate: recalc }),
    }),
  delete: (projectId: string, taskId: string) =>
    request<void>(`/projects/${projectId}/tasks/${taskId}`, {
      method: 'DELETE',
    }),
  recalculate: (projectId: string) =>
    request<{ ok: boolean }>(`/projects/${projectId}/tasks/recalculate`, {
      method: 'POST',
    }),
};

// ---------- Dependencies ----------

export const dependenciesApi = {
  list: (projectId: string) =>
    request<any[]>(`/projects/${projectId}/dependencies`),
  create: (projectId: string, data: Record<string, unknown>) =>
    request<any>(`/projects/${projectId}/dependencies`, {
      method: 'POST',
      body: JSON.stringify(data),
    }),
  update: (projectId: string, depId: string, data: Record<string, unknown>) =>
    request<any>(`/projects/${projectId}/dependencies/${depId}`, {
      method: 'PATCH',
      body: JSON.stringify(data),
    }),
  delete: (projectId: string, depId: string) =>
    request<void>(`/projects/${projectId}/dependencies/${depId}`, {
      method: 'DELETE',
    }),
};

// ---------- Calendars ----------

export const calendarsApi = {
  list: (projectId: string) =>
    request<any[]>(`/projects/${projectId}/calendars`),
  get: (projectId: string, calId: string) =>
    request<any>(`/projects/${projectId}/calendars/${calId}`),
  create: (projectId: string, data: Record<string, unknown>) =>
    request<any>(`/projects/${projectId}/calendars`, {
      method: 'POST',
      body: JSON.stringify(data),
    }),
  update: (projectId: string, calId: string, data: Record<string, unknown>) =>
    request<any>(`/projects/${projectId}/calendars/${calId}`, {
      method: 'PATCH',
      body: JSON.stringify(data),
    }),
  delete: (projectId: string, calId: string) =>
    request<void>(`/projects/${projectId}/calendars/${calId}`, {
      method: 'DELETE',
    }),
  addException: (
    projectId: string,
    calId: string,
    data: Record<string, unknown>,
  ) =>
    request<any>(`/projects/${projectId}/calendars/${calId}/exceptions`, {
      method: 'POST',
      body: JSON.stringify(data),
    }),
  deleteException: (projectId: string, calId: string, excId: string) =>
    request<void>(
      `/projects/${projectId}/calendars/${calId}/exceptions/${excId}`,
      { method: 'DELETE' },
    ),
};

// ---------- Resources ----------

export const resourcesApi = {
  list: (projectId: string) =>
    request<any[]>(`/projects/${projectId}/resources`),
  create: (projectId: string, data: Record<string, unknown>) =>
    request<any>(`/projects/${projectId}/resources`, {
      method: 'POST',
      body: JSON.stringify(data),
    }),
  update: (
    projectId: string,
    resId: string,
    data: Record<string, unknown>,
  ) =>
    request<any>(`/projects/${projectId}/resources/${resId}`, {
      method: 'PATCH',
      body: JSON.stringify(data),
    }),
  delete: (projectId: string, resId: string) =>
    request<void>(`/projects/${projectId}/resources/${resId}`, {
      method: 'DELETE',
    }),
};

// ---------- Assignments ----------

export const assignmentsApi = {
  list: (projectId: string) =>
    request<any[]>(`/projects/${projectId}/assignments`),
  create: (projectId: string, data: Record<string, unknown>) =>
    request<any>(`/projects/${projectId}/assignments`, {
      method: 'POST',
      body: JSON.stringify(data),
    }),
  update: (
    projectId: string,
    assignId: string,
    data: Record<string, unknown>,
  ) =>
    request<any>(`/projects/${projectId}/assignments/${assignId}`, {
      method: 'PATCH',
      body: JSON.stringify(data),
    }),
  delete: (projectId: string, assignId: string) =>
    request<void>(`/projects/${projectId}/assignments/${assignId}`, {
      method: 'DELETE',
    }),
};

// ---------- Baselines ----------

export const baselinesApi = {
  list: (projectId: string) =>
    request<Record<number, any[]>>(`/projects/${projectId}/baselines`),
  capture: (projectId: string, baselineIndex: number) =>
    request<{ baselineIndex: number; taskCount: number }>(
      `/projects/${projectId}/baselines`,
      { method: 'POST', body: JSON.stringify({ baselineIndex }) },
    ),
  clear: (projectId: string, index: number) =>
    request<void>(`/projects/${projectId}/baselines/${index}`, {
      method: 'DELETE',
    }),
};

// ---------- Import / Export ----------

export const importExportApi = {
  importMspdi: async (projectId: string, file: File) => {
    const form = new FormData();
    form.append('file', file);
    const res = await fetch(
      `${BASE}/projects/${projectId}/import/mspdi`,
      { method: 'POST', body: form },
    );
    if (!res.ok) throw new Error('Import failed');
    return res.json();
  },

  exportMspdi: async (projectId: string) => {
    const res = await fetch(
      `${BASE}/projects/${projectId}/export/mspdi`,
    );
    if (!res.ok) throw new Error('Export failed');
    return res.blob();
  },

  previewUpdates: async (projectId: string, file: File) => {
    const form = new FormData();
    form.append('file', file);
    const res = await fetch(
      `${BASE}/projects/${projectId}/import/preview`,
      { method: 'POST', body: form },
    );
    if (!res.ok) throw new Error('Preview failed');
    return res.json();
  },

  applyUpdates: async (
    projectId: string,
    file: File,
    options?: Record<string, unknown>,
  ) => {
    const form = new FormData();
    form.append('file', file);
    if (options) form.append('options', JSON.stringify(options));
    const res = await fetch(
      `${BASE}/projects/${projectId}/import/apply`,
      { method: 'POST', body: form },
    );
    if (!res.ok) throw new Error('Apply failed');
    return res.json();
  },

  undo: (projectId: string) =>
    request<{ success: boolean }>(`/projects/${projectId}/undo`, {
      method: 'POST',
    }),
  redo: (projectId: string) =>
    request<{ success: boolean }>(`/projects/${projectId}/redo`, {
      method: 'POST',
    }),
};

// ---------- AI ----------

export interface AiChatResponse {
  conversationId: string;
  response: string;
}

export interface AiSuggestionResponse {
  suggestion: string;
}

export interface AiHealthResponse {
  ollamaAvailable: boolean;
  models: string[];
}

export interface AiConversationSummary {
  id: string;
  createdAt: string;
  updatedAt: string;
}

export interface AiPatternEntry {
  id: string;
  patternType: string;
  pattern: string;
  confidence: number;
  occurrenceCount: number;
}

export const aiApi = {
  health: () => request<AiHealthResponse>('/ai/health'),

  chat: (projectId: string, message: string, conversationId?: string) =>
    request<AiChatResponse>('/ai/chat', {
      method: 'POST',
      body: JSON.stringify({ projectId, message, conversationId }),
    }),

  suggest: (
    projectId: string,
    type: 'duration' | 'dependency' | 'name' | 'resource' | 'general',
    context: Record<string, unknown>,
  ) =>
    request<AiSuggestionResponse>('/ai/suggest', {
      method: 'POST',
      body: JSON.stringify({ projectId, type, context }),
    }),

  feedback: (
    projectId: string,
    suggestionType: string,
    suggestion: Record<string, unknown>,
    accepted: boolean,
    correctedValue?: string,
  ) =>
    request<void>('/ai/feedback', {
      method: 'POST',
      body: JSON.stringify({
        projectId,
        suggestionType,
        suggestion,
        accepted,
        correctedValue,
      }),
    }),

  listConversations: (projectId: string) =>
    request<AiConversationSummary[]>(
      `/ai/conversations?projectId=${encodeURIComponent(projectId)}`,
    ),

  getConversation: (id: string) =>
    request<{
      id: string;
      projectId: string;
      messages: { role: string; content: string }[];
    }>(`/ai/conversations/${id}`),

  deleteConversation: (id: string) =>
    request<void>(`/ai/conversations/${id}`, { method: 'DELETE' }),

  getPatterns: () => request<AiPatternEntry[]>('/ai/patterns'),

  getFeedbackStats: () =>
    request<{ total: number; accepted: number; acceptanceRate: number }>(
      '/ai/feedback/stats',
    ),
};

// ---------- Leveling ----------

export interface LevelingResult {
  delayedTasks: { taskId: string; taskName: string; delayMinutes: number }[];
  overAllocatedResources: string[];
}

export const levelingApi = {
  level: (projectId: string) =>
    request<LevelingResult>(`/projects/${projectId}/leveling/level`, {
      method: 'POST',
    }),
  clear: (projectId: string) =>
    request<{ ok: boolean }>(`/projects/${projectId}/leveling/clear`, {
      method: 'POST',
    }),
};
