// @schedulesync/frontend/src/components/ribbon/FormatRibbon.tsx
/**
 * FormatRibbon — ribbon content for the Format tab.
 * Contains: Text styles, Bar styles, column settings, Gantt chart style.
 */

import React from 'react';
import { Box, IconButton, Tooltip, Typography, FormControlLabel, Switch } from '@mui/material';
import FormatBoldIcon from '@mui/icons-material/FormatBold';
import FormatColorFillIcon from '@mui/icons-material/FormatColorFill';
import GridOnIcon from '@mui/icons-material/GridOn';
import VisibilityIcon from '@mui/icons-material/Visibility';
import VisibilityOffIcon from '@mui/icons-material/VisibilityOff';

import { useUIStore } from '../../stores';

const RibbonGroup: React.FC<{ label: string; children: React.ReactNode }> = ({
  label,
  children,
}) => (
  <Box
    sx={{
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
      borderRight: '1px solid #E0E0E0',
      px: 1,
      py: 0.25,
    }}
  >
    <Box sx={{ display: 'flex', alignItems: 'center', gap: 0.25 }}>{children}</Box>
    <Typography
      variant="caption"
      sx={{ fontSize: '0.6rem', color: 'text.secondary', lineHeight: 1, mt: 0.25 }}
    >
      {label}
    </Typography>
  </Box>
);

const FormatRibbon: React.FC = () => {
  const showBaseline = useUIStore((s) => s.showBaseline);
  const setShowBaseline = useUIStore((s) => s.setShowBaseline);
  const showCriticalPath = useUIStore((s) => s.showCriticalPath);
  const toggleCriticalPath = useUIStore((s) => s.toggleCriticalPath);
  const openDialogWith = useUIStore((s) => s.openDialogWith);

  return (
    <Box sx={{ display: 'flex', alignItems: 'stretch' }}>
      <RibbonGroup label="Bar Styles">
        <Tooltip title="Format options (coming soon)">
          <span>
            <IconButton size="small" disabled>
              <FormatColorFillIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
      </RibbonGroup>

      <RibbonGroup label="Show/Hide">
        <FormControlLabel
          control={
            <Switch
              size="small"
              checked={showCriticalPath}
              onChange={toggleCriticalPath}
            />
          }
          label={<Typography sx={{ fontSize: '0.7rem' }}>Critical Path</Typography>}
          sx={{ mx: 0 }}
        />
        <FormControlLabel
          control={
            <Switch
              size="small"
              checked={showBaseline !== null}
              onChange={() =>
                setShowBaseline(showBaseline !== null ? null : 0)
              }
            />
          }
          label={<Typography sx={{ fontSize: '0.7rem' }}>Baseline</Typography>}
          sx={{ mx: 0 }}
        />
      </RibbonGroup>

      <RibbonGroup label="Columns">
        <Tooltip title="Show/Hide Columns">
          <IconButton size="small" onClick={() => openDialogWith('columnChooser')}>
            <GridOnIcon fontSize="small" />
          </IconButton>
        </Tooltip>
      </RibbonGroup>
    </Box>
  );
};

export default FormatRibbon;
