// @schedulesync/frontend/src/components/ribbon/ResourceRibbon.tsx
/**
 * ResourceRibbon — ribbon content for the Resource tab.
 * Contains: Add Resource, Assign Resources, Resource Pool, Leveling.
 */

import React, { useCallback } from 'react';
import { Box, IconButton, Tooltip, Typography } from '@mui/material';
import PersonAddIcon from '@mui/icons-material/PersonAdd';
import BalanceIcon from '@mui/icons-material/Balance';
import ClearIcon from '@mui/icons-material/Clear';
import AssignmentIndIcon from '@mui/icons-material/AssignmentInd';
import InfoIcon from '@mui/icons-material/Info';
import TuneIcon from '@mui/icons-material/Tune';

import { useProjectStore, useUIStore } from '../../stores';
import { levelingApi } from '../../api';

const RibbonGroup: React.FC<{ label: string; children: React.ReactNode }> = ({
  label,
  children,
}) => (
  <Box
    sx={{
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
      borderRight: '1px solid #E0E0E0',
      px: 1,
      py: 0.25,
    }}
  >
    <Box sx={{ display: 'flex', alignItems: 'center', gap: 0.25 }}>{children}</Box>
    <Typography
      variant="caption"
      sx={{ fontSize: '0.6rem', color: 'text.secondary', lineHeight: 1, mt: 0.25 }}
    >
      {label}
    </Typography>
  </Box>
);

const ResourceRibbon: React.FC = () => {
  const activeProjectId = useProjectStore((s) => s.activeProjectId);
  const createResource = useProjectStore((s) => s.createResource);
  const showSnackbar = useUIStore((s) => s.showSnackbar);
  const openDialogWith = useUIStore((s) => s.openDialogWith);
  const disabled = !activeProjectId;

  const handleAddResource = useCallback(async () => {
    if (!activeProjectId) return;
    try {
      await createResource({ name: 'New Resource', type: 'work', maxUnits: 100 });
      showSnackbar('Resource added', 'success');
    } catch (e: any) {
      showSnackbar(e.message, 'error');
    }
  }, [activeProjectId, createResource, showSnackbar]);

  const handleLevelResources = useCallback(async () => {
    if (!activeProjectId) return;
    try {
      const result = await levelingApi.level(activeProjectId);
      if (result.delayedTasks.length === 0) {
        showSnackbar('No over-allocations found', 'info');
      } else {
        showSnackbar(
          `Leveled ${result.delayedTasks.length} task(s) across ${result.overAllocatedResources.length} resource(s)`,
          'success',
        );
        await useProjectStore.getState().setActiveProject(activeProjectId);
      }
    } catch (e: any) {
      showSnackbar(e.message, 'error');
    }
  }, [activeProjectId, showSnackbar]);

  const handleClearLeveling = useCallback(async () => {
    if (!activeProjectId) return;
    try {
      await levelingApi.clear(activeProjectId);
      showSnackbar('Leveling delays cleared', 'success');
      await useProjectStore.getState().setActiveProject(activeProjectId);
    } catch (e: any) {
      showSnackbar(e.message, 'error');
    }
  }, [activeProjectId, showSnackbar]);

  return (
    <Box sx={{ display: 'flex', alignItems: 'stretch' }}>
      <RibbonGroup label="Insert">
        <Tooltip title="Add Resource">
          <span>
            <IconButton size="small" onClick={handleAddResource} disabled={disabled}>
              <PersonAddIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
      </RibbonGroup>

      <RibbonGroup label="Assignments">
        <Tooltip title="Assign Resources">
          <span>
            <IconButton
              size="small"
              onClick={() => openDialogWith('resource')}
              disabled={disabled}
            >
              <AssignmentIndIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
        <Tooltip title="Resource Information">
          <span>
            <IconButton
              size="small"
              onClick={() => {
                const res = useProjectStore.getState().resources;
                if (res.length > 0) openDialogWith('resourceInfo', res[0]);
              }}
              disabled={disabled}
            >
              <InfoIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
      </RibbonGroup>

      <RibbonGroup label="Level">
        <Tooltip title="Level All Resources">
          <span>
            <IconButton size="small" onClick={handleLevelResources} disabled={disabled}>
              <BalanceIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
        <Tooltip title="Clear Leveling">
          <span>
            <IconButton size="small" onClick={handleClearLeveling} disabled={disabled}>
              <ClearIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
        <Tooltip title="Leveling Options">
          <span>
            <IconButton size="small" onClick={() => openDialogWith('leveling')} disabled={disabled}>
              <TuneIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
      </RibbonGroup>
    </Box>
  );
};

export default ResourceRibbon;
