// @schedulesync/frontend/src/components/ribbon/ResourceRibbon.tsx
/**
 * ResourceRibbon — ribbon content for the Resource tab.
 * Contains: Add Resource, Assign Resources, Resource Pool, Leveling.
 */

import React, { useCallback } from 'react';
import { Box, IconButton, Tooltip, Typography } from '@mui/material';
import PersonAddIcon from '@mui/icons-material/PersonAdd';
import GroupIcon from '@mui/icons-material/Group';
import BalanceIcon from '@mui/icons-material/Balance';
import AssignmentIndIcon from '@mui/icons-material/AssignmentInd';

import { useProjectStore, useUIStore } from '../../stores';

const RibbonGroup: React.FC<{ label: string; children: React.ReactNode }> = ({
  label,
  children,
}) => (
  <Box
    sx={{
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
      borderRight: '1px solid #E0E0E0',
      px: 1,
      py: 0.25,
    }}
  >
    <Box sx={{ display: 'flex', alignItems: 'center', gap: 0.25 }}>{children}</Box>
    <Typography
      variant="caption"
      sx={{ fontSize: '0.6rem', color: 'text.secondary', lineHeight: 1, mt: 0.25 }}
    >
      {label}
    </Typography>
  </Box>
);

const ResourceRibbon: React.FC = () => {
  const activeProjectId = useProjectStore((s) => s.activeProjectId);
  const createResource = useProjectStore((s) => s.createResource);
  const showSnackbar = useUIStore((s) => s.showSnackbar);
  const openDialogWith = useUIStore((s) => s.openDialogWith);
  const disabled = !activeProjectId;

  const handleAddResource = useCallback(async () => {
    if (!activeProjectId) return;
    try {
      await createResource({ name: 'New Resource', type: 'work', maxUnits: 100 });
      showSnackbar('Resource added', 'success');
    } catch (e: any) {
      showSnackbar(e.message, 'error');
    }
  }, [activeProjectId, createResource, showSnackbar]);

  return (
    <Box sx={{ display: 'flex', alignItems: 'stretch' }}>
      <RibbonGroup label="Insert">
        <Tooltip title="Add Resource">
          <span>
            <IconButton size="small" onClick={handleAddResource} disabled={disabled}>
              <PersonAddIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
      </RibbonGroup>

      <RibbonGroup label="Assignments">
        <Tooltip title="Assign Resources">
          <span>
            <IconButton
              size="small"
              onClick={() => openDialogWith('resource')}
              disabled={disabled}
            >
              <AssignmentIndIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
      </RibbonGroup>

      <RibbonGroup label="Level">
        <Tooltip title="Resource Leveling (coming soon)">
          <span>
            <IconButton size="small" disabled>
              <BalanceIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
      </RibbonGroup>
    </Box>
  );
};

export default ResourceRibbon;
