// @schedulesync/frontend/src/components/ribbon/TaskRibbon.tsx
/**
 * TaskRibbon — ribbon content for the Task tab.
 * Contains: Add/Delete, Indent/Outdent, Link/Unlink, Undo/Redo,
 * Import/Export, Baseline, Critical Path, Schedule info.
 */

import React, { useCallback, useRef } from 'react';
import {
  Box,
  IconButton,
  Button,
  Divider,
  Tooltip,
  Typography,
} from '@mui/material';
import AddIcon from '@mui/icons-material/Add';
import DeleteIcon from '@mui/icons-material/Delete';
import FormatIndentIncreaseIcon from '@mui/icons-material/FormatIndentIncrease';
import FormatIndentDecreaseIcon from '@mui/icons-material/FormatIndentDecrease';
import LinkIcon from '@mui/icons-material/Link';
import LinkOffIcon from '@mui/icons-material/LinkOff';
import UndoIcon from '@mui/icons-material/Undo';
import RedoIcon from '@mui/icons-material/Redo';
import InfoIcon from '@mui/icons-material/Info';
import FileUploadIcon from '@mui/icons-material/FileUpload';
import FileDownloadIcon from '@mui/icons-material/FileDownload';
import SaveAltIcon from '@mui/icons-material/SaveAlt';
import FlagIcon from '@mui/icons-material/Flag';
import TimelineIcon from '@mui/icons-material/Timeline';
import CheckCircleOutlineIcon from '@mui/icons-material/CheckCircleOutline';
import EventNoteIcon from '@mui/icons-material/EventNote';
import SmartToyIcon from '@mui/icons-material/SmartToy';
import HistoryIcon from '@mui/icons-material/History';

import { useProjectStore, useUIStore } from '../../stores';
import { useAiStore } from '../../stores/useAiStore.js';
import { importExportApi } from '../../api';

/** Labelled group with a title below */
const RibbonGroup: React.FC<{
  label: string;
  children: React.ReactNode;
}> = ({ label, children }) => (
  <Box
    sx={{
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
      borderRight: '1px solid #E0E0E0',
      px: 1,
      py: 0.25,
      minWidth: 0,
    }}
  >
    <Box sx={{ display: 'flex', alignItems: 'center', gap: 0.25, flexWrap: 'nowrap' }}>
      {children}
    </Box>
    <Typography
      variant="caption"
      sx={{ fontSize: '0.6rem', color: 'text.secondary', lineHeight: 1, mt: 0.25 }}
    >
      {label}
    </Typography>
  </Box>
);

const TaskRibbon: React.FC = () => {
  const activeProjectId = useProjectStore((s) => s.activeProjectId);
  const selectedTaskIds = useProjectStore((s) => s.selectedTaskIds);
  const createTask = useProjectStore((s) => s.createTask);
  const deleteTask = useProjectStore((s) => s.deleteTask);
  const batchUpdateTasks = useProjectStore((s) => s.batchUpdateTasks);
  const tasks = useProjectStore((s) => s.tasks);
  const dependencies = useProjectStore((s) => s.dependencies);
  const createDependency = useProjectStore((s) => s.createDependency);
  const deleteDependency = useProjectStore((s) => s.deleteDependency);
  const showCriticalPath = useUIStore((s) => s.showCriticalPath);
  const toggleCriticalPath = useUIStore((s) => s.toggleCriticalPath);
  const openDialogWith = useUIStore((s) => s.openDialogWith);
  const showSnackbar = useUIStore((s) => s.showSnackbar);

  const mspdiInputRef = useRef<HTMLInputElement>(null);
  const updateInputRef = useRef<HTMLInputElement>(null);
  const disabled = !activeProjectId;

  const handleAddTask = useCallback(async () => {
    if (!activeProjectId) return;
    try {
      await createTask({ name: 'New Task' });
    } catch (e: any) {
      showSnackbar(e.message, 'error');
    }
  }, [activeProjectId, createTask, showSnackbar]);

  const handleAddMilestone = useCallback(async () => {
    if (!activeProjectId) return;
    try {
      await createTask({ name: 'New Milestone', durationMinutes: 0 });
    } catch (e: any) {
      showSnackbar(e.message, 'error');
    }
  }, [activeProjectId, createTask, showSnackbar]);

  const handleDeleteTask = useCallback(async () => {
    const ids = Array.from(selectedTaskIds);
    for (const id of ids) {
      await deleteTask(id);
    }
  }, [selectedTaskIds, deleteTask]);

  const handleIndent = useCallback(async () => {
    const ids = Array.from(selectedTaskIds);
    if (ids.length === 0) return;
    const updates = ids
      .map((id) => {
        const task = tasks.find((t) => t.id === id);
        if (!task) return null;
        const idx = tasks.indexOf(task);
        const prevSibling = tasks
          .slice(0, idx)
          .reverse()
          .find(
            (t) =>
              t.parentId === task.parentId && t.outlineLevel === task.outlineLevel,
          );
        if (!prevSibling) return null;
        return { id, data: { parentId: prevSibling.id } };
      })
      .filter(Boolean) as { id: string; data: Record<string, unknown> }[];
    if (updates.length > 0) await batchUpdateTasks(updates);
  }, [selectedTaskIds, tasks, batchUpdateTasks]);

  const handleOutdent = useCallback(async () => {
    const ids = Array.from(selectedTaskIds);
    if (ids.length === 0) return;
    const updates = ids
      .map((id) => {
        const task = tasks.find((t) => t.id === id);
        if (!task || !task.parentId) return null;
        const parent = tasks.find((t) => t.id === task.parentId);
        return { id, data: { parentId: parent?.parentId ?? null } };
      })
      .filter(Boolean) as { id: string; data: Record<string, unknown> }[];
    if (updates.length > 0) await batchUpdateTasks(updates);
  }, [selectedTaskIds, tasks, batchUpdateTasks]);

  const handleLink = useCallback(async () => {
    const ids = Array.from(selectedTaskIds);
    if (ids.length < 2) {
      showSnackbar('Select at least 2 tasks to link', 'warning');
      return;
    }
    for (let i = 0; i < ids.length - 1; i++) {
      await createDependency({
        fromTaskId: ids[i],
        toTaskId: ids[i + 1],
        type: 'FS',
        lagMinutes: 0,
      });
    }
  }, [selectedTaskIds, createDependency, showSnackbar]);

  const handleUnlink = useCallback(async () => {
    const ids = new Set(selectedTaskIds);
    const toRemove = dependencies.filter(
      (d) => ids.has(d.fromTaskId) || ids.has(d.toTaskId),
    );
    for (const dep of toRemove) {
      await deleteDependency(dep.id);
    }
  }, [selectedTaskIds, dependencies, deleteDependency]);

  const handleUndo = useCallback(async () => {
    if (!activeProjectId) return;
    const result = await importExportApi.undo(activeProjectId);
    if (!result.success) showSnackbar('Nothing to undo', 'info');
    else {
      useProjectStore.getState().fetchTasks();
      useProjectStore.getState().fetchDependencies();
    }
  }, [activeProjectId, showSnackbar]);

  const handleRedo = useCallback(async () => {
    if (!activeProjectId) return;
    const result = await importExportApi.redo(activeProjectId);
    if (!result.success) showSnackbar('Nothing to redo', 'info');
    else {
      useProjectStore.getState().fetchTasks();
      useProjectStore.getState().fetchDependencies();
    }
  }, [activeProjectId, showSnackbar]);

  const handleImportMspdi = useCallback(
    async (e: React.ChangeEvent<HTMLInputElement>) => {
      const file = e.target.files?.[0];
      if (!file || !activeProjectId) return;
      try {
        const result = await importExportApi.importMspdi(activeProjectId, file);
        showSnackbar(
          `Imported ${result.imported.tasks} tasks, ${result.imported.dependencies} dependencies`,
          'success',
        );
        await useProjectStore.getState().setActiveProject(activeProjectId);
      } catch (err: any) {
        showSnackbar(err.message, 'error');
      }
      e.target.value = '';
    },
    [activeProjectId, showSnackbar],
  );

  const handleExportMspdi = useCallback(async () => {
    if (!activeProjectId) return;
    try {
      const blob = await importExportApi.exportMspdi(activeProjectId);
      const url = URL.createObjectURL(blob);
      const a = document.createElement('a');
      a.href = url;
      a.download = 'project.xml';
      a.click();
      URL.revokeObjectURL(url);
    } catch (err: any) {
      showSnackbar(err.message, 'error');
    }
  }, [activeProjectId, showSnackbar]);

  const handleImportUpdates = useCallback(
    async (e: React.ChangeEvent<HTMLInputElement>) => {
      const file = e.target.files?.[0];
      if (!file || !activeProjectId) return;
      openDialogWith('importPreview', file);
      e.target.value = '';
    },
    [activeProjectId, openDialogWith],
  );

  return (
    <Box sx={{ display: 'flex', alignItems: 'stretch', gap: 0 }}>
      {/* Insert group */}
      <RibbonGroup label="Insert">
        <Tooltip title="Add Task">
          <span>
            <IconButton size="small" onClick={handleAddTask} disabled={disabled}>
              <AddIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
        <Tooltip title="Add Milestone">
          <span>
            <IconButton size="small" onClick={handleAddMilestone} disabled={disabled}>
              <EventNoteIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
        <Tooltip title="Delete Task">
          <span>
            <IconButton
              size="small"
              onClick={handleDeleteTask}
              disabled={disabled || selectedTaskIds.size === 0}
            >
              <DeleteIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
      </RibbonGroup>

      {/* Schedule group */}
      <RibbonGroup label="Schedule">
        <Tooltip title="Indent">
          <span>
            <IconButton size="small" onClick={handleIndent} disabled={disabled}>
              <FormatIndentIncreaseIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
        <Tooltip title="Outdent">
          <span>
            <IconButton size="small" onClick={handleOutdent} disabled={disabled}>
              <FormatIndentDecreaseIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
        <Tooltip title="Link Tasks (FS)">
          <span>
            <IconButton size="small" onClick={handleLink} disabled={disabled}>
              <LinkIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
        <Tooltip title="Unlink Tasks">
          <span>
            <IconButton size="small" onClick={handleUnlink} disabled={disabled}>
              <LinkOffIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
      </RibbonGroup>

      {/* Clipboard group */}
      <RibbonGroup label="Editing">
        <Tooltip title="Undo">
          <span>
            <IconButton size="small" onClick={handleUndo} disabled={disabled}>
              <UndoIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
        <Tooltip title="Redo">
          <span>
            <IconButton size="small" onClick={handleRedo} disabled={disabled}>
              <RedoIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
        <Tooltip title="Undo History">
          <span>
            <IconButton size="small" onClick={() => openDialogWith('undoHistory')} disabled={disabled}>
              <HistoryIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
      </RibbonGroup>

      {/* Properties group */}
      <RibbonGroup label="Properties">
        <Tooltip title="Task Information">
          <span>
            <IconButton
              size="small"
              onClick={() => {
                const sel = Array.from(selectedTaskIds);
                if (sel.length === 1) {
                  const task = tasks.find((t) => t.id === sel[0]);
                  if (task) openDialogWith('taskInfo', task);
                }
              }}
              disabled={disabled || selectedTaskIds.size !== 1}
            >
              <InfoIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
        <Tooltip title="Mark on Track (100%)">
          <span>
            <IconButton
              size="small"
              onClick={async () => {
                const ids = Array.from(selectedTaskIds);
                if (ids.length > 0) {
                  await batchUpdateTasks(
                    ids.map((id) => ({ id, data: { percentComplete: 100 } })),
                  );
                }
              }}
              disabled={disabled || selectedTaskIds.size === 0}
            >
              <CheckCircleOutlineIcon fontSize="small" />
            </IconButton>
          </span>
        </Tooltip>
      </RibbonGroup>

      <Box sx={{ flex: 1 }} />

      {/* AI toggle */}
      <Box sx={{ display: 'flex', alignItems: 'center', px: 1 }}>
        <Tooltip title="AI Assistant">
          <IconButton
            size="small"
            onClick={useAiStore.getState().togglePanel}
            color={useAiStore.getState().panelOpen ? 'primary' : 'default'}
          >
            <SmartToyIcon fontSize="small" />
          </IconButton>
        </Tooltip>
      </Box>

      {/* Hidden file inputs */}
      <input type="file" accept=".xml" hidden ref={mspdiInputRef} onChange={handleImportMspdi} />
      <input type="file" accept=".csv,.json" hidden ref={updateInputRef} onChange={handleImportUpdates} />
    </Box>
  );
};

export default TaskRibbon;
