// @schedulesync/engine/dist/validation/diffEngine.js
/**
 * Diff engine — ported from C# ScheduleSync.Core.Diff.DiffEngine.
 *
 * Computes the set of changes between a task snapshot and a proposed update.
 * Uses the same change detection logic: value equality, 0.001 tolerance for doubles.
 */
import { ChangeFlags, ValidationSeverity, } from '../types.js';
import { validate, validateNew } from '../validation/taskValidator.js';
const DURATION_TOLERANCE = 0.001;
/**
 * Compute a diff for a single existing task.
 */
export function computeDiff(snapshot, update) {
    let changes = ChangeFlags.None;
    // Compare each field for actual change
    if (update.newStart != null && update.newStart !== snapshot.start) {
        changes |= ChangeFlags.Start;
    }
    if (update.newFinish != null && update.newFinish !== snapshot.finish) {
        changes |= ChangeFlags.Finish;
    }
    if (update.newDurationMinutes != null &&
        Math.abs(update.newDurationMinutes - snapshot.durationMinutes) >
            DURATION_TOLERANCE) {
        changes |= ChangeFlags.Duration;
    }
    if (update.newPercentComplete != null &&
        update.newPercentComplete !== snapshot.percentComplete) {
        changes |= ChangeFlags.PercentComplete;
    }
    if (update.newConstraintType != null &&
        update.newConstraintType !== snapshot.constraintType) {
        changes |= ChangeFlags.ConstraintType;
    }
    if (update.newConstraintDate != null &&
        update.newConstraintDate !== (snapshot.constraintDate ?? null)) {
        changes |= ChangeFlags.ConstraintDate;
    }
    if (update.notesAppend != null && update.notesAppend.trim().length > 0) {
        changes |= ChangeFlags.Notes;
    }
    if (update.newDeadline != null &&
        update.newDeadline !== (snapshot.deadline ?? null)) {
        changes |= ChangeFlags.Deadline;
    }
    // Run validation
    const warnings = validate(snapshot, update);
    // Blocked if any error-severity warning
    const isBlocked = warnings.some((w) => w.severity === ValidationSeverity.Error);
    return {
        uniqueId: snapshot.uniqueId,
        taskName: snapshot.name,
        before: snapshot,
        update,
        changes,
        warnings,
        isBlocked,
        isNewTask: false,
    };
}
/**
 * Compute a diff for a new task (no existing snapshot).
 */
export function computeNewTaskDiff(update) {
    let changes = ChangeFlags.None;
    if (update.newStart != null)
        changes |= ChangeFlags.Start;
    if (update.newFinish != null)
        changes |= ChangeFlags.Finish;
    if (update.newDurationMinutes != null)
        changes |= ChangeFlags.Duration;
    if (update.newPercentComplete != null)
        changes |= ChangeFlags.PercentComplete;
    if (update.newConstraintType != null)
        changes |= ChangeFlags.ConstraintType;
    if (update.newConstraintDate != null)
        changes |= ChangeFlags.ConstraintDate;
    if (update.notesAppend != null && update.notesAppend.trim().length > 0)
        changes |= ChangeFlags.Notes;
    if (update.newDeadline != null)
        changes |= ChangeFlags.Deadline;
    if (update.name != null)
        changes |= ChangeFlags.Name;
    const warnings = validateNew(update);
    const isBlocked = warnings.some((w) => w.severity === ValidationSeverity.Error);
    return {
        uniqueId: update.uniqueId ?? 0,
        taskName: update.name ?? '(new task)',
        before: null,
        update,
        changes,
        warnings,
        isBlocked,
        isNewTask: true,
    };
}
/**
 * Batch computation: resolve each update to a snapshot via the provided
 * resolver function, then compute the diff.
 *
 * @param updates Array of proposed updates.
 * @param taskResolver Function that resolves a TaskUpdate to its current TaskSnapshot (or null if unmatched).
 * @param allowCreation If true, unmatched tasks produce a new-task diff; if false, they're blocked.
 */
export function computeDiffs(updates, taskResolver, allowCreation = false) {
    const diffs = [];
    for (const update of updates) {
        const snapshot = taskResolver(update);
        if (snapshot) {
            diffs.push(computeDiff(snapshot, update));
        }
        else if (allowCreation || update.isNew) {
            diffs.push(computeNewTaskDiff(update));
        }
        else {
            // Unmatched and creation not allowed — blocked
            const diff = {
                uniqueId: update.uniqueId ?? 0,
                taskName: update.name ?? `(unmatched: ${update.externalKey ?? update.uniqueId})`,
                before: null,
                update,
                changes: ChangeFlags.None,
                warnings: [
                    {
                        severity: ValidationSeverity.Error,
                        message: `No matching task found for UniqueId=${update.uniqueId}, ExternalKey=${update.externalKey}.`,
                    },
                ],
                isBlocked: true,
                isNewTask: false,
            };
            diffs.push(diff);
        }
    }
    return diffs;
}
