// @schedulesync/engine/dist/validation/importOrchestrator.js
/**
 * Import orchestrator — ported from C# ScheduleSync.Core.ImportOrchestrator.
 *
 * Central workflow coordinator:
 * - Resolve task updates to snapshots
 * - Compute diffs
 * - Apply changes (three-phase: hierarchy creation → update existing → record blocked)
 * - Build hierarchy plan for preview
 */
import { ChangeFlags, TaskApplyStatus, DEFAULT_APPLY_OPTIONS, } from '../types.js';
import { computeDiffs } from '../validation/diffEngine.js';
export class ImportOrchestrator {
    store;
    options;
    constructor(store, options) {
        this.store = store;
        this.options = { ...DEFAULT_APPLY_OPTIONS, ...options };
    }
    /**
     * Resolve a TaskUpdate to its current snapshot.
     * UniqueId preferred, ExternalKey fallback.
     */
    resolveTask(update) {
        if (update.uniqueId != null) {
            const t = this.store.getTaskByUniqueId(update.uniqueId);
            if (t)
                return t;
        }
        if (update.externalKey) {
            return this.store.getTaskByExternalKey(update.externalKey, this.options.externalKeyFieldName);
        }
        return null;
    }
    /**
     * Compute diffs for a set of updates.
     */
    computeDiffs(updates) {
        return computeDiffs(updates, (u) => this.resolveTask(u), this.options.allowTaskCreation);
    }
    /**
     * Apply a set of diffs — three-phase process.
     *
     * Phase 1: Create hierarchy for new tasks
     * Phase 2: Apply updates to existing tasks
     * Phase 3: Record blocked new tasks as skipped
     */
    apply(diffs) {
        const allDetails = [];
        // Phase 1: Create new tasks with hierarchy
        const newTaskDiffs = diffs.filter((d) => d.isNewTask && !d.isBlocked);
        if (newTaskDiffs.length > 0) {
            const createResult = this.createTasksWithHierarchy(newTaskDiffs);
            allDetails.push(...createResult);
        }
        // Phase 2: Apply updates to existing (non-new) tasks
        const existingDiffs = diffs.filter((d) => !d.isNewTask);
        if (existingDiffs.length > 0) {
            const applyResult = this.store.applyUpdates(existingDiffs, this.options);
            allDetails.push(...applyResult.details);
        }
        // Phase 3: Record blocked new tasks
        const blockedNewDiffs = diffs.filter((d) => d.isNewTask && d.isBlocked);
        for (const diff of blockedNewDiffs) {
            allDetails.push({
                uniqueId: diff.uniqueId || null,
                externalKey: diff.update.externalKey,
                taskName: diff.taskName,
                status: TaskApplyStatus.Skipped,
                message: diff.warnings.map((w) => w.message).join('; '),
                changesApplied: ChangeFlags.None,
            });
        }
        // Aggregate
        const applied = allDetails.filter((d) => d.status === TaskApplyStatus.Applied).length;
        const skipped = allDetails.filter((d) => d.status === TaskApplyStatus.Skipped).length;
        const failed = allDetails.filter((d) => d.status === TaskApplyStatus.Failed ||
            d.status === TaskApplyStatus.NotMatched).length;
        return {
            totalProcessed: allDetails.length,
            applied,
            skipped,
            failed,
            details: allDetails,
        };
    }
    /**
     * Create new tasks, grouped by ProjectNumber → Location hierarchy.
     * Matches the C# CreateTasksWithHierarchy method.
     */
    createTasksWithHierarchy(diffs) {
        const details = [];
        const summaryCache = new Map();
        for (const diff of diffs) {
            try {
                const update = diff.update;
                const projectNumber = update.metadata?.['ProjectNumber'] ?? 'Unassigned';
                const location = update.metadata?.['Location'] ?? 'General';
                // Find or create project-level summary
                const projectKey = `project:${projectNumber}`;
                let projectSummary = summaryCache.get(projectKey);
                if (!projectSummary) {
                    projectSummary = this.store.findOrCreateSummaryTask(projectNumber, null);
                    summaryCache.set(projectKey, projectSummary);
                }
                // Find or create location-level summary under project
                const locationKey = `location:${projectNumber}:${location}`;
                let locationSummary = summaryCache.get(locationKey);
                if (!locationSummary) {
                    locationSummary = this.store.findOrCreateSummaryTask(location, projectSummary.uniqueId);
                    summaryCache.set(locationKey, locationSummary);
                }
                // Create the work task under the location summary
                const newTask = this.store.createTask(update, this.options, locationSummary.uniqueId);
                // Set deadline if provided
                if (update.newDeadline) {
                    this.store.setDeadline(newTask.uniqueId, update.newDeadline);
                }
                // Set external key
                if (update.externalKey) {
                    this.store.setCustomTextField(newTask.uniqueId, this.options.externalKeyFieldName, update.externalKey);
                }
                details.push({
                    uniqueId: newTask.uniqueId,
                    externalKey: update.externalKey,
                    taskName: update.name ?? '(new task)',
                    status: TaskApplyStatus.Applied,
                    message: 'Created new task.',
                    changesApplied: diff.changes,
                });
            }
            catch (e) {
                const msg = e instanceof Error ? e.message : String(e);
                details.push({
                    uniqueId: diff.update.uniqueId ?? null,
                    externalKey: diff.update.externalKey,
                    taskName: diff.update.name ?? '(new task)',
                    status: TaskApplyStatus.Failed,
                    message: `Failed to create task: ${msg}`,
                    changesApplied: ChangeFlags.None,
                });
            }
        }
        return details;
    }
    /**
     * Build a hierarchy preview without creating anything.
     * Groups updates by ProjectNumber → Location.
     */
    static buildHierarchyPlan(updates) {
        const projectMap = new Map();
        for (const update of updates) {
            if (!update.isNew)
                continue;
            const projectNumber = update.metadata?.['ProjectNumber'] ?? 'Unassigned';
            const location = update.metadata?.['Location'] ?? 'General';
            if (!projectMap.has(projectNumber)) {
                projectMap.set(projectNumber, new Map());
            }
            const locMap = projectMap.get(projectNumber);
            if (!locMap.has(location)) {
                locMap.set(location, []);
            }
            locMap.get(location).push(update);
        }
        const groups = [];
        for (const [projectName, locMap] of projectMap) {
            const children = [];
            let projectTaskCount = 0;
            for (const [locName, tasks] of locMap) {
                children.push({
                    name: locName,
                    level: 2,
                    taskCount: tasks.length,
                    children: [],
                });
                projectTaskCount += tasks.length;
            }
            groups.push({
                name: projectName,
                level: 1,
                taskCount: projectTaskCount,
                children,
            });
        }
        return groups;
    }
}
