// @schedulesync/engine/dist/types.js
/**
 * @schedulesync/engine — Type definitions
 *
 * Core data model for the scheduling engine, porting and extending
 * the C# ScheduleSync.Core models with full CPM scheduling support.
 */
// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------
/** Task type — determines rendering and scheduling behaviour. */
export var TaskType;
(function (TaskType) {
    TaskType["Task"] = "task";
    TaskType["Summary"] = "summary";
    TaskType["Milestone"] = "milestone";
})(TaskType || (TaskType = {}));
/** Dependency link type — matches MS Project's four relationship types. */
export var DependencyType;
(function (DependencyType) {
    /** Finish-to-Start (default) */
    DependencyType["FS"] = "FS";
    /** Start-to-Start */
    DependencyType["SS"] = "SS";
    /** Finish-to-Finish */
    DependencyType["FF"] = "FF";
    /** Start-to-Finish */
    DependencyType["SF"] = "SF";
})(DependencyType || (DependencyType = {}));
/**
 * Constraint type — mirrors the PjConstraint enum values from MS Project.
 * Integer values match the COM enum for MSPDI compatibility.
 */
export var ConstraintType;
(function (ConstraintType) {
    /** As Soon As Possible (default, forward scheduled) */
    ConstraintType[ConstraintType["ASAP"] = 0] = "ASAP";
    /** As Late As Possible */
    ConstraintType[ConstraintType["ALAP"] = 1] = "ALAP";
    /** Must Start On */
    ConstraintType[ConstraintType["MustStartOn"] = 2] = "MustStartOn";
    /** Must Finish On */
    ConstraintType[ConstraintType["MustFinishOn"] = 3] = "MustFinishOn";
    /** Start No Earlier Than */
    ConstraintType[ConstraintType["StartNoEarlierThan"] = 4] = "StartNoEarlierThan";
    /** Start No Later Than */
    ConstraintType[ConstraintType["StartNoLaterThan"] = 5] = "StartNoLaterThan";
    /** Finish No Earlier Than */
    ConstraintType[ConstraintType["FinishNoEarlierThan"] = 6] = "FinishNoEarlierThan";
    /** Finish No Later Than */
    ConstraintType[ConstraintType["FinishNoLaterThan"] = 7] = "FinishNoLaterThan";
})(ConstraintType || (ConstraintType = {}));
/** Resource type. */
export var ResourceType;
(function (ResourceType) {
    ResourceType["Work"] = "work";
    ResourceType["Material"] = "material";
    ResourceType["Cost"] = "cost";
})(ResourceType || (ResourceType = {}));
/** Task scheduling mode — controls which variable adjusts when assignments change. */
export var TaskMode;
(function (TaskMode) {
    TaskMode["FixedDuration"] = "fixedDuration";
    TaskMode["FixedWork"] = "fixedWork";
    TaskMode["FixedUnits"] = "fixedUnits";
})(TaskMode || (TaskMode = {}));
/** Cost accrual method. */
export var AccrueAt;
(function (AccrueAt) {
    AccrueAt["Start"] = "start";
    AccrueAt["End"] = "end";
    AccrueAt["Prorated"] = "prorated";
})(AccrueAt || (AccrueAt = {}));
/** Work contour profile for resource assignments. */
export var WorkContour;
(function (WorkContour) {
    WorkContour["Flat"] = "flat";
    WorkContour["BackLoaded"] = "backLoaded";
    WorkContour["FrontLoaded"] = "frontLoaded";
    WorkContour["Bell"] = "bell";
    WorkContour["Turtle"] = "turtle";
    WorkContour["EarlyPeak"] = "earlyPeak";
    WorkContour["LatePeak"] = "latePeak";
    WorkContour["DoublePeak"] = "doublePeak";
})(WorkContour || (WorkContour = {}));
/** Direction the project is scheduled from. */
export var ScheduleFrom;
(function (ScheduleFrom) {
    ScheduleFrom["Start"] = "start";
    ScheduleFrom["Finish"] = "finish";
})(ScheduleFrom || (ScheduleFrom = {}));
/**
 * Change flags — bitwise flags indicating which fields changed.
 * Ported from C# ScheduleSync.Core.Models.ChangeFlags.
 */
export var ChangeFlags;
(function (ChangeFlags) {
    ChangeFlags[ChangeFlags["None"] = 0] = "None";
    ChangeFlags[ChangeFlags["Start"] = 1] = "Start";
    ChangeFlags[ChangeFlags["Finish"] = 2] = "Finish";
    ChangeFlags[ChangeFlags["Duration"] = 4] = "Duration";
    ChangeFlags[ChangeFlags["PercentComplete"] = 8] = "PercentComplete";
    ChangeFlags[ChangeFlags["ConstraintType"] = 16] = "ConstraintType";
    ChangeFlags[ChangeFlags["ConstraintDate"] = 32] = "ConstraintDate";
    ChangeFlags[ChangeFlags["Notes"] = 64] = "Notes";
    ChangeFlags[ChangeFlags["Deadline"] = 128] = "Deadline";
    ChangeFlags[ChangeFlags["Name"] = 256] = "Name";
})(ChangeFlags || (ChangeFlags = {}));
/** Validation message severity. Ported from C# ValidationSeverity. */
export var ValidationSeverity;
(function (ValidationSeverity) {
    ValidationSeverity[ValidationSeverity["Info"] = 0] = "Info";
    ValidationSeverity[ValidationSeverity["Warning"] = 1] = "Warning";
    ValidationSeverity[ValidationSeverity["Error"] = 2] = "Error";
})(ValidationSeverity || (ValidationSeverity = {}));
/** Status of a single task apply operation. Ported from C# TaskApplyStatus. */
export var TaskApplyStatus;
(function (TaskApplyStatus) {
    TaskApplyStatus["Applied"] = "applied";
    TaskApplyStatus["Skipped"] = "skipped";
    TaskApplyStatus["Failed"] = "failed";
    TaskApplyStatus["NotMatched"] = "notMatched";
})(TaskApplyStatus || (TaskApplyStatus = {}));
/** Default apply options. */
export const DEFAULT_APPLY_OPTIONS = {
    externalKeyFieldName: 'Text30',
    blockOnWarnings: true,
    undoTransactionLabel: 'Apply Schedule Updates',
    allowTaskCreation: true,
};
/** Creates a ParseResult with computed success property. */
export function createParseResult(updates, errors) {
    return {
        updates,
        errors,
        get success() {
            return this.errors.length === 0;
        },
    };
}
