// @schedulesync/frontend/src/components/gantt/DependencyLines.tsx
/**
 * DependencyLines — SVG overlay that draws dependency arrows between bars.
 */

import React, { useMemo  } from 'react';
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';

import { type TaskRow, type DependencyRow } from '../../stores';

dayjs.extend(utc);

interface DependencyLinesProps {
  tasks: TaskRow[];
  dependencies: DependencyRow[];
  timelineStart: string;
  dayWidth: number;
  rowHeight: number;
  /** Map from task.id → visible row index (some rows may be collapsed) */
  taskIndexMap: Map<string, number>;
  totalHeight: number;
  totalWidth: number;
}

const DependencyLines: React.FC<DependencyLinesProps> = ({
  tasks,
  dependencies,
  timelineStart,
  dayWidth,
  rowHeight,
  taskIndexMap,
  totalHeight,
  totalWidth,
}) => {
  const tlStart = useMemo(() => dayjs.utc(timelineStart), [timelineStart]);
  const taskMap = useMemo(
    () => new Map(tasks.map((t) => [t.id, t])),
    [tasks],
  );

  const lines = useMemo(() => {
    return dependencies
      .map((dep) => {
        const from = taskMap.get(dep.fromTaskId);
        const to = taskMap.get(dep.toTaskId);
        if (!from || !to) return null;

        const fromIdx = taskIndexMap.get(from.id);
        const toIdx = taskIndexMap.get(to.id);
        if (fromIdx === undefined || toIdx === undefined) return null;

        const fromStart = dayjs.utc(from.start);
        const fromFinish = dayjs.utc(from.finish);
        const toStart = dayjs.utc(to.start);

        // Source point — depends on link type
        let sx: number;
        const sy = fromIdx * rowHeight + rowHeight / 2;

        // FS (1) or FF (0): source from finish side
        if (dep.type === 0 || dep.type === 1) {
          sx = fromFinish.diff(tlStart, 'day', true) * dayWidth;
        } else {
          sx = fromStart.diff(tlStart, 'day', true) * dayWidth;
        }

        // Target point
        let tx: number;
        const ty = toIdx * rowHeight + rowHeight / 2;

        // FS (1) or SS (3): target at start side
        if (dep.type === 1 || dep.type === 3) {
          tx = toStart.diff(tlStart, 'day', true) * dayWidth;
        } else {
          tx = dayjs.utc(to.finish).diff(tlStart, 'day', true) * dayWidth;
        }

        // Build elbow path  
        const midX = dep.type === 1 || dep.type === 0
          ? sx + 8
          : sx - 8;

        const path = `M ${sx} ${sy} L ${midX} ${sy} L ${midX} ${ty} L ${tx} ${ty}`;
        const arrowDir = (dep.type === 1 || dep.type === 3) ? 'right' : 'left';

        return { path, tx, ty, arrowDir, id: dep.id };
      })
      .filter(Boolean) as {
      path: string;
      tx: number;
      ty: number;
      arrowDir: string;
      id: string;
    }[];
  }, [dependencies, taskMap, taskIndexMap, tlStart, dayWidth, rowHeight]);

  return (
    <svg
      style={{
        position: 'absolute',
        top: 0,
        left: 0,
        width: totalWidth,
        height: totalHeight,
        pointerEvents: 'none',
      }}
    >
      <defs>
        <marker
          id="arrow-right"
          markerWidth="8"
          markerHeight="6"
          refX="8"
          refY="3"
          orient="auto"
        >
          <polygon points="0 0, 8 3, 0 6" fill="#666" />
        </marker>
        <marker
          id="arrow-left"
          markerWidth="8"
          markerHeight="6"
          refX="0"
          refY="3"
          orient="auto"
        >
          <polygon points="8 0, 0 3, 8 6" fill="#666" />
        </marker>
      </defs>
      {lines.map((l) => (
        <path
          key={l.id}
          d={l.path}
          fill="none"
          stroke="#666"
          strokeWidth={1.2}
          markerEnd={`url(#arrow-${l.arrowDir})`}
        />
      ))}
    </svg>
  );
};

export default DependencyLines;
