// @schedulesync/frontend/src/components/gantt/GanttChart.tsx
/**
 * GanttChart — the right-side Gantt panel: timeline header + bars + dependency lines.
 *
 * Scrolls horizontally. Row heights match the task grid so they align.
 */

import React, { useMemo, useRef } from 'react';
import { Box } from '@mui/material';
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import minMax from 'dayjs/plugin/minMax';

import TimelineHeader from './TimelineHeader';
import GanttBar from './GanttBar';
import DependencyLines from './DependencyLines';
import { useProjectStore, useUIStore, type GanttZoom } from '../../stores';

dayjs.extend(utc);
dayjs.extend(minMax);

const ROW_HEIGHT = 28; // matches grid row height

function dayWidthForZoom(zoom: GanttZoom): number {
  switch (zoom) {
    case 'day':
      return 30;
    case 'week':
      return 14;
    case 'month':
      return 5;
    case 'quarter':
      return 1.8;
    case 'year':
      return 0.8;
  }
}

const GanttChart: React.FC = () => {
  const tasks = useProjectStore((s) => s.tasks);
  const dependencies = useProjectStore((s) => s.dependencies);
  const selectedTaskIds = useProjectStore((s) => s.selectedTaskIds);
  const zoom = useUIStore((s) => s.ganttZoom);
  const containerRef = useRef<HTMLDivElement>(null);

  const dayWidth = dayWidthForZoom(zoom);

  // Compute timeline range — from earliest start to latest finish + buffer
  const { timelineStart, timelineEnd } = useMemo(() => {
    if (tasks.length === 0) {
      const now = dayjs.utc();
      return {
        timelineStart: now.startOf('month').toISOString(),
        timelineEnd: now.add(3, 'month').endOf('month').toISOString(),
      };
    }
    const starts = tasks.map((t) => dayjs.utc(t.start));
    const finishes = tasks.map((t) => dayjs.utc(t.finish));
    const earliest = dayjs.min(...starts)!.subtract(7, 'day').startOf('week');
    const latest = dayjs.max(...finishes)!.add(14, 'day').endOf('week');
    return {
      timelineStart: earliest.toISOString(),
      timelineEnd: latest.toISOString(),
    };
  }, [tasks]);

  const totalDays = dayjs.utc(timelineEnd).diff(dayjs.utc(timelineStart), 'day');
  const totalWidth = totalDays * dayWidth;
  const totalHeight = tasks.length * ROW_HEIGHT;

  const taskIndexMap = useMemo(
    () => new Map(tasks.map((t, i) => [t.id, i])),
    [tasks],
  );

  return (
    <Box
      ref={containerRef}
      sx={{
        height: '100%',
        overflow: 'auto',
        position: 'relative',
        bgcolor: '#FFFFFF',
      }}
    >
      {/* Timeline header (sticky top) */}
      <Box sx={{ position: 'sticky', top: 0, zIndex: 2, bgcolor: '#FFFFFF' }}>
        <TimelineHeader
          startDate={timelineStart}
          endDate={timelineEnd}
          dayWidth={dayWidth}
        />
      </Box>

      {/* Bar area */}
      <Box
        sx={{
          position: 'relative',
          width: totalWidth,
          height: totalHeight,
          minHeight: 200,
        }}
      >
        {/* Alternating row stripes */}
        {tasks.map((_, i) => (
          <Box
            key={i}
            sx={{
              position: 'absolute',
              top: i * ROW_HEIGHT,
              left: 0,
              right: 0,
              width: totalWidth,
              height: ROW_HEIGHT,
              bgcolor: i % 2 === 0 ? 'transparent' : 'rgba(0,0,0,0.02)',
              borderBottom: '1px solid #F0F0F0',
            }}
          />
        ))}

        {/* Today line */}
        {(() => {
          const todayOffset =
            dayjs.utc().diff(dayjs.utc(timelineStart), 'day', true) * dayWidth;
          if (todayOffset > 0 && todayOffset < totalWidth) {
            return (
              <Box
                sx={{
                  position: 'absolute',
                  top: 0,
                  left: todayOffset,
                  width: 1.5,
                  height: totalHeight,
                  bgcolor: '#ED6C02',
                  zIndex: 1,
                  opacity: 0.7,
                }}
              />
            );
          }
          return null;
        })()}

        {/* Bars */}
        {tasks.map((task, idx) => (
          <GanttBar
            key={task.id}
            task={task}
            timelineStart={timelineStart}
            dayWidth={dayWidth}
            rowHeight={ROW_HEIGHT}
            rowIndex={idx}
            isSelected={selectedTaskIds.has(task.id)}
          />
        ))}

        {/* Dependency lines */}
        <DependencyLines
          tasks={tasks}
          dependencies={dependencies}
          timelineStart={timelineStart}
          dayWidth={dayWidth}
          rowHeight={ROW_HEIGHT}
          taskIndexMap={taskIndexMap}
          totalHeight={totalHeight}
          totalWidth={totalWidth}
        />
      </Box>
    </Box>
  );
};

export default GanttChart;
