// @schedulesync/frontend/src/components/gantt/GanttBar.tsx
/**
 * GanttBar — renders a single task as a bar (or diamond for milestones,
 * bracket for summary tasks).
 *
 * Supports drag-to-move (shift bar left/right) and drag-to-resize
 * (right edge to change duration). Shows progress fill, deadline markers,
 * and right-click context menu.
 */

import React, { memo, useCallback, useRef, useState } from 'react';
import { Box, Tooltip, Menu, MenuItem, ListItemIcon, ListItemText } from '@mui/material';
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import InfoIcon from '@mui/icons-material/Info';
import LinkIcon from '@mui/icons-material/Link';
import DeleteIcon from '@mui/icons-material/Delete';
import CheckCircleOutlineIcon from '@mui/icons-material/CheckCircleOutline';

import { type TaskRow } from '../../stores';
import { useProjectStore, useUIStore } from '../../stores';
import { shortDate, durationDays, pctLabel } from '../../utils/format';

dayjs.extend(utc);

interface GanttBarProps {
  task: TaskRow;
  timelineStart: string;
  dayWidth: number;
  rowHeight: number;
  rowIndex: number;
  isSelected: boolean;
  baselineStart?: string | null;
  baselineFinish?: string | null;
}

const GanttBar: React.FC<GanttBarProps> = ({
  task,
  timelineStart,
  dayWidth,
  rowHeight,
  rowIndex,
  isSelected,
  baselineStart,
  baselineFinish,
}) => {
  const showCriticalPath = useUIStore((s) => s.showCriticalPath);
  const openDialogWith = useUIStore((s) => s.openDialogWith);
  const updateTask = useProjectStore((s) => s.updateTask);
  const selectTask = useProjectStore((s) => s.selectTask);
  const deleteTask = useProjectStore((s) => s.deleteTask);

  const tlStart = dayjs.utc(timelineStart);
  const taskStart = dayjs.utc(task.start);
  const taskFinish = dayjs.utc(task.finish);

  const offsetDays = taskStart.diff(tlStart, 'day', true);
  const durationDaysVal = taskFinish.diff(taskStart, 'day', true);
  const left = offsetDays * dayWidth;
  const width = Math.max(durationDaysVal * dayWidth, 4);
  const top = rowIndex * rowHeight;
  const barHeight = rowHeight * 0.65;
  const barTop = (rowHeight - barHeight) / 2;

  const isSummary = task.type === 'summary';
  const isMilestone = task.type === 'milestone' || task.durationMinutes === 0;

  const barColor =
    showCriticalPath && task.isCritical
      ? '#D32F2F'
      : isSummary
        ? '#333333'
        : isSelected
          ? '#1B6B3A'
          : '#4694E4';

  // Drag state
  const dragRef = useRef<{
    type: 'move' | 'resize';
    startX: number;
    origLeft: number;
    origWidth: number;
  } | null>(null);

  const handleDragStart = useCallback(
    (e: React.MouseEvent, type: 'move' | 'resize') => {
      if (isSummary) return; // Don't drag summary tasks
      e.preventDefault();
      e.stopPropagation();
      selectTask(task.id, false);

      dragRef.current = {
        type,
        startX: e.clientX,
        origLeft: left,
        origWidth: width,
      };

      const handleMouseMove = (ev: MouseEvent) => {
        // Visual feedback only during drag (via CSS transform or similar)
      };

      const handleMouseUp = (ev: MouseEvent) => {
        document.removeEventListener('mousemove', handleMouseMove);
        document.removeEventListener('mouseup', handleMouseUp);

        if (!dragRef.current) return;
        const dx = ev.clientX - dragRef.current.startX;
        const daysDelta = Math.round(dx / dayWidth);

        if (daysDelta === 0) return;

        if (dragRef.current.type === 'move') {
          const newStart = taskStart.add(daysDelta, 'day');
          const newFinish = taskFinish.add(daysDelta, 'day');
          updateTask(task.id, {
            start: newStart.toISOString(),
            finish: newFinish.toISOString(),
          });
        } else {
          // resize — change finish (and duration)
          const newFinish = taskFinish.add(daysDelta, 'day');
          if (newFinish.isAfter(taskStart)) {
            const newDuration = newFinish.diff(taskStart, 'minute');
            updateTask(task.id, {
              finish: newFinish.toISOString(),
              durationMinutes: newDuration,
            });
          }
        }
        dragRef.current = null;
      };

      document.addEventListener('mousemove', handleMouseMove);
      document.addEventListener('mouseup', handleMouseUp);
    },
    [task, taskStart, taskFinish, left, width, dayWidth, isSummary, selectTask, updateTask],
  );

  // Context menu
  const [contextMenu, setContextMenu] = useState<{ x: number; y: number } | null>(null);

  const handleContextMenu = useCallback(
    (e: React.MouseEvent) => {
      e.preventDefault();
      e.stopPropagation();
      selectTask(task.id, false);
      setContextMenu({ x: e.clientX, y: e.clientY });
    },
    [task.id, selectTask],
  );

  const closeContextMenu = () => setContextMenu(null);

  // Deadline marker
  const deadlineOffset = task.deadline
    ? dayjs.utc(task.deadline).diff(tlStart, 'day', true) * dayWidth
    : null;

  const tooltipText = `${task.name}\n${shortDate(task.start)} – ${shortDate(task.finish)}\n${durationDays(task.durationMinutes)} | ${pctLabel(task.percentComplete)}${task.deadline ? `\nDeadline: ${shortDate(task.deadline)}` : ''}`;

  return (
    <>
      <Tooltip title={<span style={{ whiteSpace: 'pre-line' }}>{tooltipText}</span>} arrow>
        <Box
          onContextMenu={handleContextMenu}
          onMouseDown={(e) => handleDragStart(e, 'move')}
          sx={{
            position: 'absolute',
            top: top + barTop,
            left,
            height: barHeight,
            userSelect: 'none',
            cursor: isSummary ? 'default' : 'grab',
          }}
        >
          {/* Baseline ghost bar */}
          {baselineStart && baselineFinish && (
            <Box
              sx={{
                position: 'absolute',
                top: barHeight * 0.8,
                left:
                  (dayjs.utc(baselineStart).diff(taskStart, 'day', true)) * dayWidth,
                width:
                  Math.max(
                    dayjs.utc(baselineFinish).diff(
                      dayjs.utc(baselineStart),
                      'day',
                      true,
                    ) * dayWidth,
                    4,
                  ),
                height: barHeight * 0.3,
                bgcolor: 'rgba(128,128,128,0.4)',
                borderRadius: 0.5,
              }}
            />
          )}

          {/* Milestone diamond */}
          {isMilestone && !isSummary ? (
            <Box
              sx={{
                width: barHeight * 0.7,
                height: barHeight * 0.7,
                bgcolor: barColor,
                transform: 'rotate(45deg)',
                position: 'absolute',
                top: barHeight * 0.15,
                left: -barHeight * 0.35,
              }}
            />
          ) : isSummary ? (
            /* Summary bracket */
            <Box
              sx={{
                width,
                height: barHeight * 0.35,
                bgcolor: barColor,
                borderRadius: 0,
                position: 'relative',
              }}
            >
              <Box
                sx={{
                  position: 'absolute',
                  left: 0,
                  bottom: -4,
                  width: 0,
                  height: 0,
                  borderLeft: '5px solid transparent',
                  borderRight: '5px solid transparent',
                  borderTop: `5px solid ${barColor}`,
                }}
              />
              <Box
                sx={{
                  position: 'absolute',
                  right: 0,
                  bottom: -4,
                  width: 0,
                  height: 0,
                  borderLeft: '5px solid transparent',
                  borderRight: '5px solid transparent',
                  borderTop: `5px solid ${barColor}`,
                }}
              />
            </Box>
          ) : (
            /* Normal bar with progress */
            <Box sx={{ width, height: barHeight, borderRadius: 0.5, overflow: 'hidden', position: 'relative' }}>
              {/* Background */}
              <Box
                sx={{
                  width: '100%',
                  height: '100%',
                  bgcolor: barColor,
                  opacity: 0.35,
                  position: 'absolute',
                  borderRadius: 0.5,
                }}
              />
              {/* Progress fill */}
              <Box
                sx={{
                  width: `${Math.min(task.percentComplete, 100)}%`,
                  height: '100%',
                  bgcolor: barColor,
                  position: 'absolute',
                  borderRadius: 0.5,
                }}
              />
              {/* Progress black line inside bar (MS Project style) */}
              {task.percentComplete > 0 && task.percentComplete < 100 && (
                <Box
                  sx={{
                    position: 'absolute',
                    top: barHeight * 0.3,
                    left: 0,
                    width: `${Math.min(task.percentComplete, 100)}%`,
                    height: barHeight * 0.4,
                    borderRight: '2px solid #000',
                  }}
                />
              )}
              {/* Task name label to right of bar */}
              <Box
                sx={{
                  position: 'absolute',
                  left: width + 4,
                  top: 0,
                  height: '100%',
                  display: 'flex',
                  alignItems: 'center',
                  whiteSpace: 'nowrap',
                  pointerEvents: 'none',
                }}
              >
                <span
                  style={{
                    fontSize: '0.65rem',
                    color: '#333',
                  }}
                >
                  {task.name}
                </span>
              </Box>

              {/* Right resize handle */}
              <Box
                onMouseDown={(e) => {
                  e.stopPropagation();
                  handleDragStart(e, 'resize');
                }}
                sx={{
                  position: 'absolute',
                  right: 0,
                  top: 0,
                  width: 6,
                  height: '100%',
                  cursor: 'ew-resize',
                  '&:hover': { bgcolor: 'rgba(0,0,0,0.2)' },
                }}
              />
            </Box>
          )}

          {/* Deadline marker (green down-arrow) */}
          {deadlineOffset !== null && !isSummary && (
            <Box
              sx={{
                position: 'absolute',
                left: deadlineOffset - left,
                top: -4,
                width: 0,
                height: 0,
                borderLeft: '4px solid transparent',
                borderRight: '4px solid transparent',
                borderTop: '6px solid #4CAF50',
              }}
            />
          )}
        </Box>
      </Tooltip>

      {/* Context menu */}
      <Menu
        open={contextMenu !== null}
        onClose={closeContextMenu}
        anchorReference="anchorPosition"
        anchorPosition={
          contextMenu ? { top: contextMenu.y, left: contextMenu.x } : undefined
        }
      >
        <MenuItem
          onClick={() => {
            closeContextMenu();
            openDialogWith('taskInfo', task);
          }}
        >
          <ListItemIcon><InfoIcon fontSize="small" /></ListItemIcon>
          <ListItemText>Task Information</ListItemText>
        </MenuItem>
        <MenuItem
          onClick={() => {
            closeContextMenu();
            updateTask(task.id, { percentComplete: 100 });
          }}
        >
          <ListItemIcon><CheckCircleOutlineIcon fontSize="small" /></ListItemIcon>
          <ListItemText>Mark 100% Complete</ListItemText>
        </MenuItem>
        <MenuItem
          onClick={() => {
            closeContextMenu();
            deleteTask(task.id);
          }}
        >
          <ListItemIcon><DeleteIcon fontSize="small" color="error" /></ListItemIcon>
          <ListItemText>Delete Task</ListItemText>
        </MenuItem>
      </Menu>
    </>
  );
};

export default memo(GanttBar);
