// @schedulesync/frontend/src/components/gantt/GanttBar.tsx
/**
 * GanttBar — renders a single task as a bar (or diamond for milestones,
 * bracket for summary tasks).
 */

import React, { memo } from 'react';
import { Box, Tooltip } from '@mui/material';
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';

import { type TaskRow } from '../../stores';
import { useUIStore } from '../../stores';
import { shortDate, durationDays, pctLabel } from '../../utils/format';

dayjs.extend(utc);

interface GanttBarProps {
  task: TaskRow;
  timelineStart: string;
  dayWidth: number;
  rowHeight: number;
  rowIndex: number;
  isSelected: boolean;
  baselineStart?: string | null;
  baselineFinish?: string | null;
}

const GanttBar: React.FC<GanttBarProps> = ({
  task,
  timelineStart,
  dayWidth,
  rowHeight,
  rowIndex,
  isSelected,
  baselineStart,
  baselineFinish,
}) => {
  const showCriticalPath = useUIStore((s) => s.showCriticalPath);
  const tlStart = dayjs.utc(timelineStart);
  const taskStart = dayjs.utc(task.start);
  const taskFinish = dayjs.utc(task.finish);

  const offsetDays = taskStart.diff(tlStart, 'day', true);
  const durationDaysVal = taskFinish.diff(taskStart, 'day', true);
  const left = offsetDays * dayWidth;
  const width = Math.max(durationDaysVal * dayWidth, 4);
  const top = rowIndex * rowHeight;
  const barHeight = rowHeight * 0.65;
  const barTop = (rowHeight - barHeight) / 2;

  const isSummary = task.type === 2;
  const isMilestone = task.type === 1 || task.durationMinutes === 0;

  const barColor =
    showCriticalPath && task.isCritical
      ? '#D32F2F'
      : isSummary
        ? '#333333'
        : isSelected
          ? '#1B6B3A'
          : '#4694E4';

  const tooltipText = `${task.name}\n${shortDate(task.start)} – ${shortDate(task.finish)}\n${durationDays(task.durationMinutes)} | ${pctLabel(task.percentComplete)}`;

  return (
    <Tooltip title={<span style={{ whiteSpace: 'pre-line' }}>{tooltipText}</span>} arrow>
      <Box
        sx={{
          position: 'absolute',
          top: top + barTop,
          left,
          height: barHeight,
          userSelect: 'none',
        }}
      >
        {/* Baseline ghost bar */}
        {baselineStart && baselineFinish && (
          <Box
            sx={{
              position: 'absolute',
              top: barHeight * 0.8,
              left:
                (dayjs.utc(baselineStart).diff(taskStart, 'day', true)) * dayWidth,
              width:
                Math.max(
                  dayjs.utc(baselineFinish).diff(
                    dayjs.utc(baselineStart),
                    'day',
                    true,
                  ) * dayWidth,
                  4,
                ),
              height: barHeight * 0.3,
              bgcolor: 'rgba(128,128,128,0.4)',
              borderRadius: 0.5,
            }}
          />
        )}

        {/* Milestone diamond */}
        {isMilestone && !isSummary ? (
          <Box
            sx={{
              width: barHeight * 0.7,
              height: barHeight * 0.7,
              bgcolor: barColor,
              transform: 'rotate(45deg)',
              position: 'absolute',
              top: barHeight * 0.15,
              left: -barHeight * 0.35,
            }}
          />
        ) : isSummary ? (
          /* Summary bracket */
          <Box
            sx={{
              width,
              height: barHeight * 0.35,
              bgcolor: barColor,
              borderRadius: 0,
              position: 'relative',
            }}
          >
            {/* Down-pointing ends */}
            <Box
              sx={{
                position: 'absolute',
                left: 0,
                bottom: -4,
                width: 0,
                height: 0,
                borderLeft: '5px solid transparent',
                borderRight: '5px solid transparent',
                borderTop: `5px solid ${barColor}`,
              }}
            />
            <Box
              sx={{
                position: 'absolute',
                right: 0,
                bottom: -4,
                width: 0,
                height: 0,
                borderLeft: '5px solid transparent',
                borderRight: '5px solid transparent',
                borderTop: `5px solid ${barColor}`,
              }}
            />
          </Box>
        ) : (
          /* Normal bar */
          <Box sx={{ width, height: barHeight, borderRadius: 0.5, overflow: 'hidden' }}>
            {/* Background */}
            <Box
              sx={{
                width: '100%',
                height: '100%',
                bgcolor: barColor,
                opacity: 0.8,
                position: 'absolute',
                borderRadius: 0.5,
              }}
            />
            {/* Progress fill */}
            {task.percentComplete > 0 && (
              <Box
                sx={{
                  width: `${Math.min(task.percentComplete, 100)}%`,
                  height: '100%',
                  bgcolor: barColor,
                  position: 'absolute',
                  borderRadius: 0.5,
                }}
              />
            )}
          </Box>
        )}
      </Box>
    </Tooltip>
  );
};

export default memo(GanttBar);
