// @schedulesync/backend/dist/routes/ai.js
/**
 * /api/ai routes — AI chat, suggestions, patterns, and feedback.
 */
import { z } from 'zod';
import { chat, suggest, recordFeedback, checkOllamaHealth, listModels, } from '../services/aiService.js';
import { prisma } from '../db.js';
const chatSchema = z.object({
    projectId: z.string().uuid(),
    message: z.string().min(1).max(4000),
    conversationId: z.string().uuid().optional(),
});
const suggestSchema = z.object({
    projectId: z.string().uuid(),
    type: z.enum(['duration', 'dependency', 'name', 'resource', 'general']),
    context: z.record(z.unknown()),
});
const feedbackSchema = z.object({
    projectId: z.string().uuid(),
    suggestionType: z.string(),
    suggestion: z.record(z.unknown()),
    accepted: z.boolean(),
    correctedValue: z.string().optional(),
});
export default async function aiRoutes(app) {
    // Health / status check
    app.get('/health', async () => {
        const ollamaUp = await checkOllamaHealth();
        let models = [];
        if (ollamaUp) {
            try {
                models = await listModels();
            }
            catch {
                // ignore
            }
        }
        return { ollamaAvailable: ollamaUp, models };
    });
    // Chat with AI
    app.post('/chat', async (req, reply) => {
        const body = chatSchema.parse(req.body);
        try {
            const result = await chat(body.projectId, body.message, body.conversationId);
            return result;
        }
        catch (err) {
            const message = err instanceof Error ? err.message : 'AI service unavailable';
            return reply.code(502).send({ error: message });
        }
    });
    // Quick suggestion
    app.post('/suggest', async (req, reply) => {
        const body = suggestSchema.parse(req.body);
        try {
            const response = await suggest(body.projectId, body.type, body.context);
            return { suggestion: response };
        }
        catch (err) {
            const message = err instanceof Error ? err.message : 'AI service unavailable';
            return reply.code(502).send({ error: message });
        }
    });
    // Record feedback
    app.post('/feedback', async (req, reply) => {
        const body = feedbackSchema.parse(req.body);
        await recordFeedback(body.projectId, body.suggestionType, body.suggestion, body.accepted, body.correctedValue);
        return reply.code(201).send({ ok: true });
    });
    // List conversations for a project
    app.get('/conversations', async (req) => {
        const projectId = req.query.projectId;
        return prisma.aiConversation.findMany({
            where: { projectId },
            orderBy: { updatedAt: 'desc' },
            select: { id: true, createdAt: true, updatedAt: true },
        });
    });
    // Get conversation messages
    app.get('/conversations/:id', async (req) => {
        const conv = await prisma.aiConversation.findUniqueOrThrow({
            where: { id: req.params.id },
        });
        return {
            id: conv.id,
            projectId: conv.projectId,
            messages: JSON.parse(conv.messages),
            createdAt: conv.createdAt,
            updatedAt: conv.updatedAt,
        };
    });
    // Delete conversation
    app.delete('/conversations/:id', async (req, reply) => {
        await prisma.aiConversation.delete({ where: { id: req.params.id } });
        return reply.code(204).send();
    });
    // Get learned patterns
    app.get('/patterns', async () => {
        return prisma.aiPattern.findMany({
            orderBy: { confidence: 'desc' },
        });
    });
    // Get feedback stats
    app.get('/feedback/stats', async () => {
        const [total, accepted] = await Promise.all([
            prisma.aiFeedback.count(),
            prisma.aiFeedback.count({ where: { accepted: true } }),
        ]);
        return {
            total,
            accepted,
            acceptanceRate: total > 0 ? accepted / total : 0,
        };
    });
}
