// @schedulesync/backend/dist/routes/assignments.js
/**
 * /api/projects/:projectId/assignments routes
 */
import { z } from 'zod';
import { prisma } from '../db.js';
const createSchema = z.object({
    taskId: z.string(),
    resourceId: z.string(),
    units: z.number().min(0).default(1),
    workMinutes: z.number().min(0).default(0),
});
const updateSchema = z.object({
    units: z.number().min(0).optional(),
    workMinutes: z.number().min(0).optional(),
});
export default async function assignmentRoutes(app) {
    // List assignments for a project
    app.get('/', async (req) => {
        const { projectId } = req.params;
        return prisma.assignment.findMany({
            where: { task: { projectId } },
            include: {
                task: { select: { id: true, name: true } },
                resource: { select: { id: true, name: true } },
            },
        });
    });
    // Create assignment
    app.post('/', async (req, reply) => {
        const { projectId } = req.params;
        const body = createSchema.parse(req.body);
        // Validate task belongs to project
        const task = await prisma.task.findFirst({
            where: { id: body.taskId, projectId },
        });
        if (!task) {
            return reply.code(400).send({ error: 'Task not found in project' });
        }
        // Validate resource belongs to project
        const resource = await prisma.resource.findFirst({
            where: { id: body.resourceId, projectId },
        });
        if (!resource) {
            return reply.code(400).send({ error: 'Resource not found in project' });
        }
        // Prevent duplicate
        const existing = await prisma.assignment.findFirst({
            where: { taskId: body.taskId, resourceId: body.resourceId },
        });
        if (existing) {
            return reply.code(409).send({ error: 'Assignment already exists' });
        }
        const assignment = await prisma.assignment.create({
            data: {
                taskId: body.taskId,
                resourceId: body.resourceId,
                units: body.units,
                workMinutes: body.workMinutes,
            },
        });
        return reply.code(201).send(assignment);
    });
    // Update assignment
    app.patch('/:assignId', async (req) => {
        const { assignId } = req.params;
        const body = updateSchema.parse(req.body);
        const data = {};
        if (body.units !== undefined)
            data.units = body.units;
        if (body.workMinutes !== undefined)
            data.workMinutes = body.workMinutes;
        return prisma.assignment.update({ where: { id: assignId }, data });
    });
    // Delete assignment
    app.delete('/:assignId', async (req, reply) => {
        const { assignId } = req.params;
        await prisma.assignment.delete({ where: { id: assignId } });
        return reply.code(204).send();
    });
}
