// @schedulesync/backend/dist/routes/importExport.js
/**
 * /api/projects/:projectId/import-export routes
 *
 * - MSPDI XML import / export
 * - CSV / JSON update import + preview + apply
 */
import { z } from 'zod';
import { prisma } from '../db.js';
import { recalculateProject } from '../services/schedulingService.js';
import { captureUndo } from '../services/undoService.js';
import { CsvUpdateSource, JsonUpdateSource, StratusCsvUpdateSource, ImportOrchestrator, DiffEngine, TaskValidator, DEFAULT_APPLY_OPTIONS, ConstraintType, } from '@schedulesync/engine';
import { importMspdi, exportMspdi } from '@schedulesync/mspdi';
// ---------- MSPDI Import ----------
export default async function importExportRoutes(app) {
    app.post('/import/mspdi', async (req, reply) => {
        const { projectId } = req.params;
        const data = await req.file();
        if (!data)
            return reply.code(400).send({ error: 'No file uploaded' });
        const xml = (await data.toBuffer()).toString('utf-8');
        const projectData = importMspdi(xml);
        await captureUndo(projectId, 'Import MSPDI');
        // Clear existing data
        await prisma.$transaction([
            prisma.assignment.deleteMany({ where: { task: { projectId } } }),
            prisma.baseline.deleteMany({ where: { task: { projectId } } }),
            prisma.dependency.deleteMany({ where: { projectId } }),
            prisma.task.deleteMany({ where: { projectId } }),
            prisma.calendarException.deleteMany({
                where: { calendar: { projectId } },
            }),
            prisma.calendar.deleteMany({ where: { projectId } }),
            prisma.resource.deleteMany({ where: { projectId } }),
        ]);
        // Update project metadata
        await prisma.project.update({
            where: { id: projectId },
            data: {
                name: projectData.settings.name,
                startDate: new Date(projectData.settings.startDate),
                finishDate: projectData.settings.finishDate
                    ? new Date(projectData.settings.finishDate)
                    : null,
                scheduleFrom: projectData.settings.scheduleFrom,
            },
        });
        // Insert calendars
        for (const cal of projectData.calendars) {
            await prisma.calendar.create({
                data: {
                    id: cal.id,
                    projectId,
                    name: cal.name,
                    workingDaysOfWeek: JSON.stringify(cal.workingDaysOfWeek),
                    defaultWorkingHours: JSON.stringify(cal.defaultWorkingHours),
                    exceptions: {
                        create: cal.exceptions.map((e) => ({
                            startDate: e.startDate,
                            endDate: e.endDate,
                            isWorking: e.isWorking,
                            workingHours: e.workingHours
                                ? JSON.stringify(e.workingHours)
                                : null,
                        })),
                    },
                },
            });
        }
        // Set default calendar
        if (projectData.settings.defaultCalendarId) {
            await prisma.project.update({
                where: { id: projectId },
                data: { defaultCalendarId: projectData.settings.defaultCalendarId },
            });
        }
        // Insert tasks
        for (const t of projectData.tasks) {
            await prisma.task.create({
                data: {
                    id: t.id,
                    projectId,
                    wbsCode: t.wbsCode ?? '',
                    outlineLevel: t.outlineLevel,
                    parentId: t.parentId,
                    name: t.name,
                    type: t.type,
                    durationMinutes: t.durationMinutes,
                    start: new Date(t.start),
                    finish: new Date(t.finish),
                    constraintType: t.constraintType,
                    constraintDate: t.constraintDate
                        ? new Date(t.constraintDate)
                        : null,
                    calendarId: t.calendarId,
                    percentComplete: t.percentComplete,
                    isManuallyScheduled: t.isManuallyScheduled,
                    isCritical: t.isCritical,
                    totalSlackMinutes: t.totalSlackMinutes,
                    freeSlackMinutes: t.freeSlackMinutes,
                    deadline: t.deadline ? new Date(t.deadline) : null,
                    notes: t.notes,
                    externalKey: t.externalKey,
                    sortOrder: t.sortOrder,
                },
            });
        }
        // Insert resources
        for (const r of projectData.resources) {
            await prisma.resource.create({
                data: {
                    id: r.id,
                    projectId,
                    name: r.name,
                    type: r.type,
                    maxUnits: r.maxUnits,
                    calendarId: r.calendarId,
                },
            });
        }
        // Insert dependencies
        for (const d of projectData.dependencies) {
            await prisma.dependency.create({
                data: {
                    id: d.id,
                    projectId,
                    fromTaskId: d.fromTaskId,
                    toTaskId: d.toTaskId,
                    type: d.type,
                    lagMinutes: d.lagMinutes,
                },
            });
        }
        // Insert assignments
        for (const a of projectData.assignments) {
            await prisma.assignment.create({
                data: {
                    id: a.id,
                    taskId: a.taskId,
                    resourceId: a.resourceId,
                    units: a.units,
                    workMinutes: a.workMinutes,
                },
            });
        }
        // Insert baselines
        for (const b of projectData.baselines) {
            await prisma.baseline.create({
                data: {
                    taskId: b.taskId,
                    baselineIndex: b.baselineIndex,
                    baselineStart: new Date(b.baselineStart),
                    baselineFinish: new Date(b.baselineFinish),
                    baselineDurationMinutes: b.baselineDurationMinutes,
                },
            });
        }
        await recalculateProject(projectId);
        return {
            imported: {
                tasks: projectData.tasks.length,
                dependencies: projectData.dependencies.length,
                calendars: projectData.calendars.length,
                resources: projectData.resources.length,
                assignments: projectData.assignments.length,
            },
        };
    });
    // ---------- MSPDI Export ----------
    app.get('/export/mspdi', async (req, reply) => {
        const { projectId } = req.params;
        // Load everything
        const project = await prisma.project.findUniqueOrThrow({
            where: { id: projectId },
        });
        const [tasks, deps, cals, resources, assignments, baselines] = await Promise.all([
            prisma.task.findMany({ where: { projectId }, orderBy: { sortOrder: 'asc' } }),
            prisma.dependency.findMany({ where: { projectId } }),
            prisma.calendar.findMany({
                where: { projectId },
                include: { exceptions: true },
            }),
            prisma.resource.findMany({ where: { projectId } }),
            prisma.assignment.findMany({ where: { task: { projectId } } }),
            prisma.baseline.findMany({ where: { task: { projectId } } }),
        ]);
        const projectData = {
            settings: {
                id: project.id,
                name: project.name,
                startDate: project.startDate.toISOString(),
                finishDate: project.finishDate?.toISOString() ?? null,
                defaultCalendarId: project.defaultCalendarId,
                scheduleFrom: project.scheduleFrom,
                statusDate: project.statusDate?.toISOString() ?? null,
            },
            tasks: tasks.map((t) => ({
                id: t.id,
                wbsCode: t.wbsCode,
                outlineLevel: t.outlineLevel,
                parentId: t.parentId,
                name: t.name,
                type: t.type,
                durationMinutes: t.durationMinutes,
                start: t.start.toISOString(),
                finish: t.finish.toISOString(),
                constraintType: t.constraintType,
                constraintDate: t.constraintDate?.toISOString() ?? null,
                calendarId: t.calendarId,
                percentComplete: t.percentComplete,
                isManuallyScheduled: t.isManuallyScheduled,
                isCritical: t.isCritical,
                totalSlackMinutes: t.totalSlackMinutes,
                freeSlackMinutes: t.freeSlackMinutes,
                earlyStart: t.earlyStart?.toISOString() ?? null,
                earlyFinish: t.earlyFinish?.toISOString() ?? null,
                lateStart: t.lateStart?.toISOString() ?? null,
                lateFinish: t.lateFinish?.toISOString() ?? null,
                deadline: t.deadline?.toISOString() ?? null,
                notes: t.notes,
                externalKey: t.externalKey,
                sortOrder: t.sortOrder,
            })),
            dependencies: deps.map((d) => ({
                id: d.id,
                fromTaskId: d.fromTaskId,
                toTaskId: d.toTaskId,
                type: d.type,
                lagMinutes: d.lagMinutes,
            })),
            calendars: cals.map((c) => ({
                id: c.id,
                name: c.name,
                workingDaysOfWeek: JSON.parse(c.workingDaysOfWeek),
                defaultWorkingHours: JSON.parse(c.defaultWorkingHours),
                exceptions: c.exceptions.map((e) => ({
                    startDate: e.startDate,
                    endDate: e.endDate,
                    isWorking: e.isWorking,
                    workingHours: e.workingHours ? JSON.parse(e.workingHours) : null,
                })),
            })),
            resources: resources.map((r) => ({
                id: r.id,
                name: r.name,
                type: r.type,
                maxUnits: r.maxUnits,
                calendarId: r.calendarId,
            })),
            assignments: assignments.map((a) => ({
                id: a.id,
                taskId: a.taskId,
                resourceId: a.resourceId,
                units: a.units,
                workMinutes: a.workMinutes,
            })),
            baselines: baselines.map((b) => ({
                taskId: b.taskId,
                baselineIndex: b.baselineIndex,
                baselineStart: b.baselineStart.toISOString(),
                baselineFinish: b.baselineFinish.toISOString(),
                baselineDurationMinutes: b.baselineDurationMinutes,
            })),
        };
        const xml = exportMspdi(projectData);
        reply
            .header('Content-Type', 'application/xml')
            .header('Content-Disposition', `attachment; filename="${project.name.replace(/[^a-zA-Z0-9._-]/g, '_')}.xml"`)
            .send(xml);
    });
    // ---------- CSV / JSON Update Preview ----------
    app.post('/import/preview', async (req, reply) => {
        const { projectId } = req.params;
        const data = await req.file();
        if (!data)
            return reply.code(400).send({ error: 'No file uploaded' });
        const content = (await data.toBuffer()).toString('utf-8');
        const filename = data.filename ?? 'update.csv';
        // Parse updates
        const updates = parseUpdateFile(content, filename);
        if ('error' in updates) {
            return reply.code(400).send(updates);
        }
        // Load current tasks to build snapshots
        const tasks = await prisma.task.findMany({
            where: { projectId },
            orderBy: { sortOrder: 'asc' },
        });
        const resolver = (uniqueId) => {
            const task = tasks.find((t) => parseInt(t.id, 10) === uniqueId || t.externalKey === String(uniqueId));
            if (!task)
                return null;
            return {
                uniqueId,
                name: task.name,
                durationMinutes: task.durationMinutes,
                start: task.start.toISOString(),
                finish: task.finish.toISOString(),
                percentComplete: task.percentComplete,
                constraintType: task.constraintType,
                isSummary: task.type === 2,
                isManuallyScheduled: task.isManuallyScheduled,
                hasConstraint: task.constraintType !== ConstraintType.Asap,
            };
        };
        const diffs = DiffEngine.computeDiffs(updates, resolver);
        // Validate diffs
        const validated = diffs.map((diff) => ({
            ...diff,
            warnings: TaskValidator.validate(diff),
        }));
        return { diffs: validated, totalUpdates: updates.length };
    });
    // ---------- Apply Updates ----------
    const applySchema = z.object({
        format: z.enum(['csv', 'json', 'stratus']).optional(),
        options: z
            .object({
            allowManualOverride: z.boolean().optional(),
            allowConstraintOverride: z.boolean().optional(),
            enableHierarchyCreation: z.boolean().optional(),
            keyFieldName: z.string().nullable().optional(),
        })
            .optional(),
    });
    app.post('/import/apply', async (req, reply) => {
        const { projectId } = req.params;
        const data = await req.file();
        if (!data)
            return reply.code(400).send({ error: 'No file uploaded' });
        const content = (await data.toBuffer()).toString('utf-8');
        const filename = data.filename ?? 'update.csv';
        // Parse the body fields from multipart
        const bodyFields = data.fields;
        let options = { ...DEFAULT_APPLY_OPTIONS };
        if (bodyFields && typeof bodyFields === 'object') {
            const optField = bodyFields.options;
            if (optField?.value) {
                try {
                    const parsed = JSON.parse(optField.value);
                    options = { ...DEFAULT_APPLY_OPTIONS, ...parsed };
                }
                catch {
                    // ignore
                }
            }
        }
        const updates = parseUpdateFile(content, filename);
        if ('error' in updates) {
            return reply.code(400).send(updates);
        }
        await captureUndo(projectId, `Import ${filename}`);
        // Build a simple IProjectStore adapter from DB
        const store = await buildProjectStore(projectId);
        const orchestrator = new ImportOrchestrator(store);
        const result = orchestrator.apply(updates, options);
        // Apply changes to DB
        for (const detail of result.details) {
            if (detail.status === 'Updated' && detail.diff) {
                const data = {};
                const d = detail.diff;
                if (d.newStart)
                    data.start = new Date(d.newStart);
                if (d.newFinish)
                    data.finish = new Date(d.newFinish);
                if (d.newDurationMinutes !== undefined)
                    data.durationMinutes = d.newDurationMinutes;
                if (d.newPercentComplete !== undefined)
                    data.percentComplete = d.newPercentComplete;
                // Find task by uniqueId
                const task = await prisma.task.findFirst({
                    where: {
                        projectId,
                        OR: [
                            { externalKey: String(d.uniqueId) },
                        ],
                    },
                });
                if (task) {
                    await prisma.task.update({ where: { id: task.id }, data });
                }
            }
        }
        await recalculateProject(projectId);
        return result;
    });
    // ---------- Undo / Redo ----------
    app.post('/undo', async (req) => {
        const { projectId } = req.params;
        const { undo } = await import('../services/undoService.js');
        const success = await undo(projectId);
        return { success };
    });
    app.post('/redo', async (req) => {
        const { projectId } = req.params;
        const { redo } = await import('../services/undoService.js');
        const success = await redo(projectId);
        return { success };
    });
}
// ---------- Helpers ----------
function parseUpdateFile(content, filename) {
    const lowerFilename = filename.toLowerCase();
    try {
        if (lowerFilename.endsWith('.json')) {
            const source = new JsonUpdateSource();
            const result = source.parse(content);
            if (result.errors.length > 0 && result.updates.length === 0) {
                return { error: result.errors.map((e) => e.message).join('; ') };
            }
            return result.updates;
        }
        // Check if it's a STRATUS CSV
        if (StratusCsvUpdateSource.isStratusFormat(content)) {
            const source = new StratusCsvUpdateSource();
            const result = source.parse(content);
            if (result.errors.length > 0 && result.updates.length === 0) {
                return { error: result.errors.map((e) => e.message).join('; ') };
            }
            return result.updates;
        }
        // Default CSV
        const source = new CsvUpdateSource();
        const result = source.parse(content);
        if (result.errors.length > 0 && result.updates.length === 0) {
            return { error: result.errors.map((e) => e.message).join('; ') };
        }
        return result.updates;
    }
    catch (err) {
        return { error: err.message ?? 'Failed to parse file' };
    }
}
async function buildProjectStore(projectId) {
    const tasks = await prisma.task.findMany({
        where: { projectId },
        orderBy: { sortOrder: 'asc' },
    });
    const taskMap = new Map(tasks.map((t) => [t.id, t]));
    return {
        getSnapshot(uniqueId) {
            const task = tasks.find((t) => parseInt(t.id, 10) === uniqueId ||
                t.externalKey === String(uniqueId));
            if (!task)
                return null;
            return {
                uniqueId,
                name: task.name,
                durationMinutes: task.durationMinutes,
                start: task.start.toISOString(),
                finish: task.finish.toISOString(),
                percentComplete: task.percentComplete,
                constraintType: task.constraintType,
                isSummary: task.type === 2,
                isManuallyScheduled: task.isManuallyScheduled,
                hasConstraint: task.constraintType !== ConstraintType.Asap,
            };
        },
        applyUpdate(uniqueId, update) {
            // In-memory updates—actual persistence happens in the route handler
            return true;
        },
        createTask(update) {
            // Won't be called in preview, but return a placeholder
            return -1;
        },
        setSummaryValue(uniqueId, fieldName, value) {
            // no-op for preview
        },
        addToSelection(uniqueId) {
            // no-op
        },
        openUndoTransaction(label) {
            // handled at route level
        },
        closeUndoTransaction() {
            // handled at route level
        },
        getTaskByField(fieldName, value) {
            const task = tasks.find((t) => (fieldName === 'Text30' && t.externalKey === value) ||
                (fieldName === 'Name' && t.name === value));
            return task ? parseInt(task.id, 10) || 0 : null;
        },
    };
}
