// @schedulesync/backend/dist/routes/projects.js
/**
 * /api/projects routes
 */
import { z } from 'zod';
import { prisma } from '../db.js';
const createSchema = z.object({
    name: z.string().min(1),
    startDate: z.string().datetime(),
    scheduleFrom: z.string().default('start'), // "start" | "finish"
});
const updateSchema = z.object({
    name: z.string().min(1).optional(),
    startDate: z.string().datetime().optional(),
    finishDate: z.string().datetime().nullable().optional(),
    defaultCalendarId: z.string().nullable().optional(),
    scheduleFrom: z.string().optional(),
    statusDate: z.string().datetime().nullable().optional(),
});
export default async function projectRoutes(app) {
    // List all projects
    app.get('/', async () => {
        return prisma.project.findMany({
            orderBy: { updatedAt: 'desc' },
            select: {
                id: true,
                name: true,
                startDate: true,
                finishDate: true,
                createdAt: true,
                updatedAt: true,
            },
        });
    });
    // Get single project with counts
    app.get('/:id', async (req, reply) => {
        const { id } = req.params;
        const project = await prisma.project.findUnique({
            where: { id },
            include: {
                _count: {
                    select: { tasks: true, calendars: true, resources: true },
                },
            },
        });
        if (!project)
            return reply.code(404).send({ error: 'Project not found' });
        return project;
    });
    // Create project
    app.post('/', async (req, reply) => {
        const body = createSchema.parse(req.body);
        const project = await prisma.project.create({
            data: {
                name: body.name,
                startDate: new Date(body.startDate),
                scheduleFrom: body.scheduleFrom,
            },
        });
        return reply.code(201).send(project);
    });
    // Update project
    app.patch('/:id', async (req, reply) => {
        const { id } = req.params;
        const body = updateSchema.parse(req.body);
        const data = {};
        if (body.name !== undefined)
            data.name = body.name;
        if (body.startDate !== undefined)
            data.startDate = new Date(body.startDate);
        if (body.finishDate !== undefined)
            data.finishDate = body.finishDate ? new Date(body.finishDate) : null;
        if (body.defaultCalendarId !== undefined)
            data.defaultCalendarId = body.defaultCalendarId;
        if (body.scheduleFrom !== undefined)
            data.scheduleFrom = body.scheduleFrom;
        if (body.statusDate !== undefined)
            data.statusDate = body.statusDate ? new Date(body.statusDate) : null;
        const project = await prisma.project.update({ where: { id }, data });
        return project;
    });
    // Delete project (cascades to tasks/deps via Prisma)
    app.delete('/:id', async (req, reply) => {
        const { id } = req.params;
        await prisma.project.delete({ where: { id } });
        return reply.code(204).send();
    });
}
