// @schedulesync/backend/dist/routes/resources.js
/**
 * /api/projects/:projectId/resources routes
 */
import { z } from 'zod';
import { prisma } from '../db.js';
const createSchema = z.object({
    name: z.string().min(1),
    type: z.string().default('work'), // "work" | "material" | "cost"
    maxUnits: z.number().min(0).default(1),
    calendarId: z.string().nullable().optional(),
});
const updateSchema = z.object({
    name: z.string().min(1).optional(),
    type: z.string().optional(),
    maxUnits: z.number().min(0).optional(),
    calendarId: z.string().nullable().optional(),
});
export default async function resourceRoutes(app) {
    // List resources
    app.get('/', async (req) => {
        const { projectId } = req.params;
        return prisma.resource.findMany({ where: { projectId } });
    });
    // Get single resource
    app.get('/:resId', async (req, reply) => {
        const { resId } = req.params;
        const res = await prisma.resource.findUnique({ where: { id: resId } });
        if (!res)
            return reply.code(404).send({ error: 'Resource not found' });
        return res;
    });
    // Create resource
    app.post('/', async (req, reply) => {
        const { projectId } = req.params;
        const body = createSchema.parse(req.body);
        const resource = await prisma.resource.create({
            data: {
                projectId,
                name: body.name,
                type: body.type,
                maxUnits: body.maxUnits,
                calendarId: body.calendarId ?? null,
            },
        });
        return reply.code(201).send(resource);
    });
    // Update resource
    app.patch('/:resId', async (req) => {
        const { resId } = req.params;
        const body = updateSchema.parse(req.body);
        const data = {};
        if (body.name !== undefined)
            data.name = body.name;
        if (body.type !== undefined)
            data.type = body.type;
        if (body.maxUnits !== undefined)
            data.maxUnits = body.maxUnits;
        if (body.calendarId !== undefined)
            data.calendarId = body.calendarId;
        return prisma.resource.update({ where: { id: resId }, data });
    });
    // Delete resource (and assignments)
    app.delete('/:resId', async (req, reply) => {
        const { resId } = req.params;
        await prisma.assignment.deleteMany({ where: { resourceId: resId } });
        await prisma.resource.delete({ where: { id: resId } });
        return reply.code(204).send();
    });
}
