// @schedulesync/backend/dist/services/aiService.js
/**
 * AI Service — integrates with a local Ollama instance for LLM-powered
 * scheduling advice, task suggestions, and pattern-based learning.
 */
import { prisma } from '../db.js';
import { buildProjectContext } from './contextBuilder.js';
import { extractAndStorePatterns } from './patternExtractor.js';
const OLLAMA_BASE = process.env.OLLAMA_URL ?? 'http://localhost:11434';
const DEFAULT_MODEL = process.env.OLLAMA_MODEL ?? 'llama3.1:8b';
/** Check if Ollama is reachable. */
export async function checkOllamaHealth() {
    try {
        const res = await fetch(`${OLLAMA_BASE}/api/tags`, {
            signal: AbortSignal.timeout(3000),
        });
        return res.ok;
    }
    catch {
        return false;
    }
}
/** List available models on the local Ollama instance. */
export async function listModels() {
    const res = await fetch(`${OLLAMA_BASE}/api/tags`);
    if (!res.ok)
        throw new Error('Cannot reach Ollama');
    const data = (await res.json());
    return data.models.map((m) => m.name);
}
/**
 * Send a chat completion request to Ollama and return the full response text.
 */
async function ollamaChat(messages, model = DEFAULT_MODEL) {
    const res = await fetch(`${OLLAMA_BASE}/api/chat`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ model, messages, stream: false }),
    });
    if (!res.ok) {
        const text = await res.text();
        throw new Error(`Ollama error ${res.status}: ${text}`);
    }
    const data = (await res.json());
    return data.message.content;
}
/** System prompt that establishes the AI as a construction scheduling expert. */
function buildSystemPrompt(learnedPatterns) {
    return `You are a construction project scheduling expert assistant embedded in a scheduling application similar to Microsoft Project.
Your role is to help users build, optimize, and maintain construction schedules.

Key capabilities:
- Suggest task names, durations, and dependencies based on construction best practices
- Identify scheduling issues (critical path problems, resource conflicts, unrealistic durations)
- Recommend task sequences for common construction activities
- Explain scheduling concepts in simple terms
- Suggest improvements to existing schedules

Construction domain knowledge:
- You understand CSI MasterFormat divisions and typical construction sequences
- You know typical duration ranges for common construction activities
- You understand predecessor/successor relationships in construction
- You are aware of weather, crew size, and productivity factors

${learnedPatterns ? `Learned patterns from this user's projects:\n${learnedPatterns}` : ''}

Guidelines:
- Be concise and actionable in your responses
- When suggesting changes, specify exact field values (duration in days, dependency types)
- Format task suggestions as structured data when possible
- If you're unsure about a specific duration or sequence, say so and provide a range
- Always consider the critical path impact of suggested changes`;
}
/**
 * Chat with the AI about a project. Maintains conversation history in the database.
 */
export async function chat(projectId, userMessage, conversationId) {
    // Load or create conversation
    let conversation = conversationId
        ? await prisma.aiConversation.findUnique({ where: { id: conversationId } })
        : null;
    if (!conversation) {
        conversation = await prisma.aiConversation.create({
            data: { projectId, messages: '[]' },
        });
    }
    const history = JSON.parse(conversation.messages);
    // Build context from project data + learned patterns
    const [projectContext, patterns] = await Promise.all([
        buildProjectContext(projectId),
        getLearnedPatternsText(projectId),
    ]);
    const systemMsg = {
        role: 'system',
        content: buildSystemPrompt(patterns),
    };
    // Include project context in the first user message or refresh periodically
    const contextPrefix = history.length === 0
        ? `[Current project context]\n${projectContext}\n\n`
        : '';
    const newUserMsg = {
        role: 'user',
        content: `${contextPrefix}${userMessage}`,
    };
    history.push(newUserMsg);
    // Call Ollama
    const responseText = await ollamaChat([systemMsg, ...history]);
    const assistantMsg = {
        role: 'assistant',
        content: responseText,
    };
    history.push(assistantMsg);
    // Persist conversation
    await prisma.aiConversation.update({
        where: { id: conversation.id },
        data: {
            messages: JSON.stringify(history),
            updatedAt: new Date(),
        },
    });
    // Async pattern extraction — don't block the response
    extractAndStorePatterns(projectId).catch(() => { });
    return { conversationId: conversation.id, response: responseText };
}
/**
 * Get a quick AI suggestion for a specific action (non-conversational).
 */
export async function suggest(projectId, suggestionType, context) {
    const [projectContext, patterns] = await Promise.all([
        buildProjectContext(projectId),
        getLearnedPatternsText(projectId),
    ]);
    const prompts = {
        duration: `Given this task context, suggest an appropriate duration in working days. Task: ${JSON.stringify(context)}`,
        dependency: `Given these tasks, suggest logical predecessor/successor relationships. Tasks: ${JSON.stringify(context)}`,
        name: `Suggest a proper construction task name for this activity. Context: ${JSON.stringify(context)}`,
        resource: `Suggest appropriate resource assignments for this task. Task: ${JSON.stringify(context)}`,
        general: `Analyze this project data and suggest improvements. Context: ${JSON.stringify(context)}`,
    };
    const messages = [
        { role: 'system', content: buildSystemPrompt(patterns) },
        {
            role: 'user',
            content: `[Project summary]\n${projectContext}\n\n${prompts[suggestionType]}\n\nRespond with a concise, actionable suggestion.`,
        },
    ];
    return ollamaChat(messages);
}
/**
 * Record user feedback on an AI suggestion (accepted/rejected/corrected).
 */
export async function recordFeedback(projectId, suggestionType, suggestion, accepted, correctedValue) {
    await prisma.aiFeedback.create({
        data: {
            projectId,
            suggestionType,
            suggestion: JSON.stringify(suggestion),
            accepted,
            correctedValue: correctedValue ?? null,
        },
    });
}
/** Compile learned patterns into a text summary for the system prompt. */
async function getLearnedPatternsText(projectId) {
    const patterns = await prisma.aiPattern.findMany({
        where: { confidence: { gte: 0.5 } },
        orderBy: { confidence: 'desc' },
        take: 20,
    });
    if (patterns.length === 0)
        return '';
    return patterns
        .map((p) => {
        const data = JSON.parse(p.pattern);
        return `- ${p.patternType}: ${JSON.stringify(data)} (confidence: ${(p.confidence * 100).toFixed(0)}%, seen ${p.occurrenceCount}x)`;
    })
        .join('\n');
}
