// @schedulesync/backend/dist/services/patternExtractor.js
/**
 * Pattern extractor — analyzes project data to discover recurring patterns
 * (naming conventions, typical durations, common dependency chains, etc.)
 * and stores them for AI learning.
 */
import { prisma } from '../db.js';
/**
 * Analyze project data and extract/update patterns.
 * Called after each AI conversation turn (fire-and-forget).
 */
export async function extractAndStorePatterns(projectId) {
    const tasks = await prisma.task.findMany({
        where: { projectId },
        orderBy: { sortOrder: 'asc' },
    });
    if (tasks.length < 3)
        return; // not enough data to learn from
    await Promise.all([
        extractNamingPatterns(tasks),
        extractDurationPatterns(tasks),
        extractSequencePatterns(projectId, tasks),
    ]);
}
/**
 * Extract naming conventions — common prefixes, suffixes, word patterns.
 */
async function extractNamingPatterns(tasks) {
    // Count word frequency in task names (excluding summary tasks)
    const wordFreq = new Map();
    const leafTasks = tasks.filter((t) => t.type !== 'summary');
    for (const t of leafTasks) {
        const words = t.name
            .toLowerCase()
            .split(/[\s\-_/]+/)
            .filter((w) => w.length > 2);
        for (const w of words) {
            wordFreq.set(w, (wordFreq.get(w) ?? 0) + 1);
        }
    }
    // Find words that appear in >20% of tasks — these are naming conventions
    const threshold = Math.max(2, leafTasks.length * 0.2);
    const commonWords = [...wordFreq.entries()]
        .filter(([, count]) => count >= threshold)
        .sort((a, b) => b[1] - a[1])
        .slice(0, 20)
        .map(([word, count]) => ({ word, count }));
    if (commonWords.length > 0) {
        await upsertPattern('naming', { commonWords }, commonWords.length / 20);
    }
}
/**
 * Extract typical duration ranges by task name keywords.
 */
async function extractDurationPatterns(tasks) {
    const leafTasks = tasks.filter((t) => t.type !== 'summary' && t.durationMinutes > 0);
    if (leafTasks.length < 3)
        return;
    // Group by significant keywords
    const keywordDurations = new Map();
    const keywords = [
        'concrete',
        'pour',
        'form',
        'rebar',
        'steel',
        'erect',
        'install',
        'drywall',
        'paint',
        'electrical',
        'plumbing',
        'hvac',
        'roofing',
        'foundation',
        'excavat',
        'grading',
        'framing',
        'insulation',
        'waterproof',
        'masonry',
        'glazing',
        'flooring',
        'ceiling',
        'landscape',
        'paving',
        'demolition',
        'mobilization',
        'inspect',
        'cure',
        'backfill',
        'rough-in',
        'finish',
        'trim',
        'elevator',
        'fire',
        'sprinkler',
    ];
    for (const t of leafTasks) {
        const nameLower = t.name.toLowerCase();
        for (const kw of keywords) {
            if (nameLower.includes(kw)) {
                const list = keywordDurations.get(kw) ?? [];
                list.push(t.durationMinutes / 480); // convert to days
                keywordDurations.set(kw, list);
            }
        }
    }
    const durationPatterns = [];
    for (const [keyword, durations] of keywordDurations) {
        if (durations.length < 2)
            continue;
        durations.sort((a, b) => a - b);
        const avg = durations.reduce((s, d) => s + d, 0) / durations.length;
        durationPatterns.push({
            keyword,
            avgDays: Math.round(avg * 10) / 10,
            minDays: durations[0],
            maxDays: durations[durations.length - 1],
            sampleSize: durations.length,
        });
    }
    if (durationPatterns.length > 0) {
        await upsertPattern('duration', { durationPatterns }, Math.min(1, durationPatterns.length / 10));
    }
}
/**
 * Extract common task sequences (A always follows B).
 */
async function extractSequencePatterns(projectId, tasks) {
    const deps = await prisma.dependency.findMany({ where: { projectId } });
    if (deps.length < 3)
        return;
    const taskMap = new Map(tasks.map((t) => [t.id, t.name]));
    const sequenceCount = new Map();
    for (const d of deps) {
        const fromName = taskMap.get(d.fromTaskId);
        const toName = taskMap.get(d.toTaskId);
        if (!fromName || !toName)
            continue;
        // Normalize names to keywords for generalization
        const fromKey = normalizeTaskName(fromName);
        const toKey = normalizeTaskName(toName);
        if (fromKey && toKey) {
            const seq = `${fromKey} -> ${toKey}`;
            sequenceCount.set(seq, (sequenceCount.get(seq) ?? 0) + 1);
        }
    }
    const sequences = [...sequenceCount.entries()]
        .filter(([, count]) => count >= 2)
        .sort((a, b) => b[1] - a[1])
        .slice(0, 20)
        .map(([sequence, count]) => ({ sequence, count }));
    if (sequences.length > 0) {
        await upsertPattern('sequence', { sequences }, Math.min(1, sequences.length / 10));
    }
}
/** Reduce a task name to its most significant keyword(s). */
function normalizeTaskName(name) {
    const stopWords = new Set([
        'the',
        'a',
        'an',
        'and',
        'or',
        'for',
        'of',
        'to',
        'in',
        'on',
        'at',
        'by',
        'with',
        'from',
        'up',
        'out',
        'phase',
        'area',
        'level',
        'floor',
        'zone',
        'section',
        'building',
        'bldg',
    ]);
    const words = name
        .toLowerCase()
        .replace(/[^a-z\s]/g, '')
        .split(/\s+/)
        .filter((w) => w.length > 2 && !stopWords.has(w));
    return words.slice(0, 2).join(' ') || null;
}
/**
 * Upsert a pattern — if the same patternType exists, update it
 * with increased occurrence count and merged data.
 */
async function upsertPattern(patternType, data, confidence) {
    const existing = await prisma.aiPattern.findFirst({
        where: { patternType },
        orderBy: { lastSeen: 'desc' },
    });
    if (existing) {
        await prisma.aiPattern.update({
            where: { id: existing.id },
            data: {
                pattern: JSON.stringify(data),
                confidence: Math.min(1, confidence),
                occurrenceCount: existing.occurrenceCount + 1,
                lastSeen: new Date(),
            },
        });
    }
    else {
        await prisma.aiPattern.create({
            data: {
                patternType,
                pattern: JSON.stringify(data),
                confidence: Math.min(1, confidence),
                occurrenceCount: 1,
            },
        });
    }
}
