// @schedulesync/backend/dist/services/schedulingService.js
/**
 * Scheduling service — bridges the database and the CPM engine.
 *
 * Loads tasks + dependencies + calendars from Prisma,
 * calls the engine's recalculate(), writes results back.
 */
import { prisma } from '../db.js';
import { recalculate, } from '@schedulesync/engine';
/**
 * Recalculate the schedule for a project and persist the results.
 */
export async function recalculateProject(projectId) {
    // Load project
    const project = await prisma.project.findUniqueOrThrow({
        where: { id: projectId },
    });
    // Load all data
    const [dbTasks, dbDeps, dbCalendars, dbResources, dbAssignments, dbBaselines] = await Promise.all([
        prisma.task.findMany({
            where: { projectId },
            orderBy: { sortOrder: 'asc' },
        }),
        prisma.dependency.findMany({ where: { projectId } }),
        prisma.calendar.findMany({
            where: { OR: [{ projectId }, { projectId: null }] },
            include: { exceptions: true },
        }),
        prisma.resource.findMany({ where: { projectId } }),
        prisma.assignment.findMany({
            where: { task: { projectId } },
        }),
        prisma.baseline.findMany({
            where: { task: { projectId } },
        }),
    ]);
    // Map to engine types
    const settings = {
        id: project.id,
        name: project.name,
        startDate: project.startDate.toISOString(),
        finishDate: project.finishDate?.toISOString() ?? null,
        defaultCalendarId: project.defaultCalendarId,
        scheduleFrom: project.scheduleFrom,
        statusDate: project.statusDate?.toISOString() ?? null,
    };
    const tasks = dbTasks.map((t) => ({
        id: t.id,
        wbsCode: t.wbsCode,
        outlineLevel: t.outlineLevel,
        parentId: t.parentId,
        name: t.name,
        type: t.type,
        durationMinutes: t.durationMinutes,
        start: t.start.toISOString(),
        finish: t.finish.toISOString(),
        constraintType: t.constraintType,
        constraintDate: t.constraintDate?.toISOString() ?? null,
        calendarId: t.calendarId,
        percentComplete: t.percentComplete,
        isManuallyScheduled: t.isManuallyScheduled,
        isCritical: t.isCritical,
        totalSlackMinutes: t.totalSlackMinutes,
        freeSlackMinutes: t.freeSlackMinutes,
        earlyStart: t.earlyStart?.toISOString() ?? null,
        earlyFinish: t.earlyFinish?.toISOString() ?? null,
        lateStart: t.lateStart?.toISOString() ?? null,
        lateFinish: t.lateFinish?.toISOString() ?? null,
        deadline: t.deadline?.toISOString() ?? null,
        notes: t.notes,
        externalKey: t.externalKey,
        sortOrder: t.sortOrder,
    }));
    const dependencies = dbDeps.map((d) => ({
        id: d.id,
        fromTaskId: d.fromTaskId,
        toTaskId: d.toTaskId,
        type: d.type,
        lagMinutes: d.lagMinutes,
    }));
    const calendars = dbCalendars.map((c) => ({
        id: c.id,
        name: c.name,
        workingDaysOfWeek: JSON.parse(c.workingDaysOfWeek),
        defaultWorkingHours: JSON.parse(c.defaultWorkingHours),
        exceptions: c.exceptions.map((e) => ({
            startDate: e.startDate,
            endDate: e.endDate,
            isWorking: e.isWorking,
            workingHours: e.workingHours
                ? JSON.parse(e.workingHours)
                : null,
        })),
    }));
    const projectData = {
        settings,
        tasks,
        dependencies,
        calendars,
        resources: dbResources.map((r) => ({
            id: r.id,
            name: r.name,
            type: r.type,
            maxUnits: r.maxUnits,
            calendarId: r.calendarId,
        })),
        assignments: dbAssignments.map((a) => ({
            id: a.id,
            taskId: a.taskId,
            resourceId: a.resourceId,
            units: a.units,
            workMinutes: a.workMinutes,
        })),
        baselines: dbBaselines.map((b) => ({
            taskId: b.taskId,
            baselineIndex: b.baselineIndex,
            baselineStart: b.baselineStart.toISOString(),
            baselineFinish: b.baselineFinish.toISOString(),
            baselineDurationMinutes: b.baselineDurationMinutes,
        })),
    };
    // Recalculate
    const result = recalculate(projectData);
    // Write updated tasks back to DB
    const updates = result.tasks.map((t) => prisma.task.update({
        where: { id: t.id },
        data: {
            start: new Date(t.start),
            finish: new Date(t.finish),
            durationMinutes: t.durationMinutes,
            percentComplete: t.percentComplete,
            isCritical: t.isCritical,
            totalSlackMinutes: t.totalSlackMinutes,
            freeSlackMinutes: t.freeSlackMinutes,
            earlyStart: t.earlyStart ? new Date(t.earlyStart) : null,
            earlyFinish: t.earlyFinish ? new Date(t.earlyFinish) : null,
            lateStart: t.lateStart ? new Date(t.lateStart) : null,
            lateFinish: t.lateFinish ? new Date(t.lateFinish) : null,
        },
    }));
    // Update project finish date
    const projectUpdate = prisma.project.update({
        where: { id: projectId },
        data: {
            finishDate: projectData.settings.finishDate
                ? new Date(projectData.settings.finishDate)
                : null,
        },
    });
    await prisma.$transaction([...updates, projectUpdate]);
}
