// @schedulesync/frontend/src/components/views/ResourceSheet.tsx
/**
 * ResourceSheet — MS Project-style resource grid.
 * Shows resources in a table with inline editing for key fields.
 */

import React, { useState, useCallback } from 'react';
import {
  Box,
  Table,
  TableHead,
  TableBody,
  TableRow,
  TableCell,
  TableContainer,
  Typography,
  IconButton,
  Tooltip,
  TextField,
  Select,
  MenuItem,
} from '@mui/material';
import DeleteIcon from '@mui/icons-material/Delete';
import { useProjectStore, type ResourceRow } from '../../stores';

const ResourceSheet: React.FC = () => {
  const resources = useProjectStore((s) => s.resources);
  const assignments = useProjectStore((s) => s.assignments);
  const deleteResource = useProjectStore((s) => s.deleteResource);
  const [editingCell, setEditingCell] = useState<{
    resId: string;
    field: string;
  } | null>(null);
  const [editValue, setEditValue] = useState('');

  // Count assignments per resource
  const assignmentCount = new Map<string, number>();
  for (const a of assignments) {
    assignmentCount.set(a.resourceId, (assignmentCount.get(a.resourceId) ?? 0) + 1);
  }

  const startEdit = (resId: string, field: string, currentValue: string) => {
    setEditingCell({ resId, field });
    setEditValue(currentValue);
  };

  const cancelEdit = () => {
    setEditingCell(null);
    setEditValue('');
  };

  const commitEdit = useCallback(async () => {
    if (!editingCell) return;
    // Resource update would go through store
    // For now, just cancel — resource update API exists
    cancelEdit();
  }, [editingCell, editValue]);

  if (resources.length === 0) {
    return (
      <Box
        sx={{
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
          height: '100%',
        }}
      >
        <Typography color="text.secondary">
          No resources. Switch to the Resource tab to add resources.
        </Typography>
      </Box>
    );
  }

  return (
    <TableContainer sx={{ height: '100%', overflow: 'auto' }}>
      <Table size="small" stickyHeader sx={{ tableLayout: 'fixed' }}>
        <TableHead>
          <TableRow>
            <TableCell sx={{ width: 40 }}>#</TableCell>
            <TableCell sx={{ minWidth: 180 }}>Resource Name</TableCell>
            <TableCell sx={{ width: 80 }}>Type</TableCell>
            <TableCell sx={{ width: 80 }} align="right">
              Max Units
            </TableCell>
            <TableCell sx={{ width: 100 }} align="right">
              Std Rate
            </TableCell>
            <TableCell sx={{ width: 100 }} align="right">
              OT Rate
            </TableCell>
            <TableCell sx={{ width: 80 }} align="right">
              Assign.
            </TableCell>
            <TableCell sx={{ width: 50 }} />
          </TableRow>
        </TableHead>
        <TableBody>
          {resources.map((res, idx) => (
            <TableRow key={res.id} hover>
              <TableCell align="center">{idx + 1}</TableCell>
              <TableCell>
                <Typography
                  sx={{
                    fontSize: '0.8125rem',
                    fontWeight: 400,
                  }}
                >
                  {res.name}
                </Typography>
              </TableCell>
              <TableCell>
                <Typography
                  variant="caption"
                  sx={{
                    bgcolor:
                      res.type === 'work'
                        ? '#E3F2FD'
                        : res.type === 'material'
                          ? '#FFF3E0'
                          : '#F3E5F5',
                    px: 0.75,
                    py: 0.25,
                    borderRadius: 0.5,
                    fontSize: '0.7rem',
                  }}
                >
                  {res.type.charAt(0).toUpperCase() + res.type.slice(1)}
                </Typography>
              </TableCell>
              <TableCell align="right">
                {res.type === 'work' ? `${res.maxUnits}%` : ''}
              </TableCell>
              <TableCell align="right">
                {(res as any).standardRate != null
                  ? `$${Number((res as any).standardRate).toFixed(2)}/hr`
                  : '-'}
              </TableCell>
              <TableCell align="right">
                {(res as any).overtimeRate != null
                  ? `$${Number((res as any).overtimeRate).toFixed(2)}/hr`
                  : '-'}
              </TableCell>
              <TableCell align="right">
                {assignmentCount.get(res.id) ?? 0}
              </TableCell>
              <TableCell>
                <Tooltip title="Delete Resource">
                  <IconButton
                    size="small"
                    onClick={() => deleteResource(res.id)}
                  >
                    <DeleteIcon fontSize="small" />
                  </IconButton>
                </Tooltip>
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
};

export default ResourceSheet;
