// @schedulesync/frontend/src/components/views/TeamPlanner.tsx
/**
 * TeamPlanner — Gantt-like view organized by resource, showing task assignments
 * as horizontal bars on each resource's timeline row. Unassigned tasks shown separately.
 */

import React, { useMemo } from 'react';
import { Box, Typography, Tooltip } from '@mui/material';
import dayjs from 'dayjs';
import { useProjectStore, useUIStore } from '../../stores';

const ROW_HEIGHT = 32;
const DAY_WIDTH = 24;

const TeamPlanner: React.FC = () => {
  const tasks = useProjectStore((s) => s.tasks);
  const resources = useProjectStore((s) => s.resources);
  const assignments = useProjectStore((s) => s.assignments);
  const ganttZoom = useUIStore((s) => s.ganttZoom);

  const { rows, days, projectStart } = useMemo(() => {
    if (tasks.length === 0) { return { rows: [], days: [], projectStart: dayjs() }; }

    const starts = tasks.map((t) => dayjs(t.start));
    const ends = tasks.map((t) => dayjs(t.finish));
    const pStart = starts.reduce((a, b) => (a.isBefore(b) ? a : b)).startOf('day');
    const pEnd = ends.reduce((a, b) => (a.isAfter(b) ? a : b)).endOf('day');
    const totalDays = pEnd.diff(pStart, 'day') + 1;

    const dayList: dayjs.Dayjs[] = [];
    for (let i = 0; i < totalDays; i++) dayList.push(pStart.add(i, 'day'));

    // Assigned tasks by resource
    const resourceRows = resources
      .filter((r) => r.type === 'work')
      .map((r) => {
        const resAssignments = assignments.filter((a) => a.resourceId === r.id);
        const resTasks = resAssignments
          .map((a) => {
            const task = tasks.find((t) => t.id === a.taskId);
            if (!task || task.type === 'summary') return null;
            return { ...task, units: a.units };
          })
          .filter(Boolean) as (typeof tasks[0] & { units: number })[];
        return { resource: r, tasks: resTasks };
      });

    // Unassigned tasks
    const assignedTaskIds = new Set(assignments.map((a) => a.taskId));
    const unassignedTasks = tasks.filter(
      (t) => t.type !== 'summary' && !assignedTaskIds.has(t.id),
    );

    return {
      rows: [
        ...resourceRows,
        ...(unassignedTasks.length > 0
          ? [{ resource: { id: '__unassigned__', name: 'Unassigned', type: 'work', maxUnits: 0 } as any, tasks: unassignedTasks.map((t) => ({ ...t, units: 0 })) }]
          : []),
      ],
      days: dayList,
      projectStart: pStart,
    };
  }, [tasks, resources, assignments, ganttZoom]);

  if (resources.length === 0 && tasks.length === 0) {
    return (
      <Box sx={{ p: 4, textAlign: 'center' }}>
        <Typography color="text.secondary">No resources or tasks to display.</Typography>
      </Box>
    );
  }

  const dayWidth = ganttZoom === 'day' ? DAY_WIDTH * 2 : ganttZoom === 'week' ? DAY_WIDTH : DAY_WIDTH / 2;

  return (
    <Box sx={{ display: 'flex', flexDirection: 'column', height: '100%', overflow: 'hidden' }}>
      <Box sx={{ bgcolor: '#F5F5F5', borderBottom: '1px solid #E0E0E0', px: 2, py: 0.5 }}>
        <Typography variant="subtitle2" sx={{ fontWeight: 600, fontSize: '0.75rem' }}>
          Team Planner
        </Typography>
      </Box>

      <Box sx={{ flex: 1, overflow: 'auto', display: 'flex' }}>
        {/* Resource names */}
        <Box sx={{ width: 160, flexShrink: 0, borderRight: '2px solid #E0E0E0', bgcolor: '#FAFAFA' }}>
          <Box sx={{ height: 24, borderBottom: '1px solid #E0E0E0' }} />
          {rows.map((row) => (
            <Box
              key={row.resource.id}
              sx={{
                height: Math.max(ROW_HEIGHT * Math.max(row.tasks.length, 1), ROW_HEIGHT),
                display: 'flex',
                alignItems: 'flex-start',
                px: 1,
                pt: 0.5,
                borderBottom: '1px solid #E0E0E0',
              }}
            >
              <Typography
                variant="body2"
                sx={{
                  fontSize: '0.75rem',
                  fontWeight: row.resource.id === '__unassigned__' ? 400 : 500,
                  fontStyle: row.resource.id === '__unassigned__' ? 'italic' : 'normal',
                }}
                noWrap
              >
                {row.resource.name}
              </Typography>
            </Box>
          ))}
        </Box>

        {/* Timeline area */}
        <Box sx={{ flex: 1, overflow: 'auto' }}>
          {/* Day headers */}
          <Box sx={{ display: 'flex', borderBottom: '1px solid #E0E0E0', height: 24 }}>
            {days.map((d, i) => (
              <Box
                key={i}
                sx={{
                  width: dayWidth,
                  minWidth: dayWidth,
                  textAlign: 'center',
                  borderRight: '1px solid #F0F0F0',
                  fontSize: '0.55rem',
                  lineHeight: '24px',
                  color: 'text.secondary',
                  bgcolor: d.day() === 0 || d.day() === 6 ? '#F5F5F5' : 'transparent',
                }}
              >
                {d.format('D')}
              </Box>
            ))}
          </Box>

          {/* Resource rows with task bars */}
          {rows.map((row) => (
            <Box
              key={row.resource.id}
              sx={{
                height: Math.max(ROW_HEIGHT * Math.max(row.tasks.length, 1), ROW_HEIGHT),
                position: 'relative',
                borderBottom: '1px solid #E0E0E0',
                minWidth: days.length * dayWidth,
              }}
            >
              {/* Weekend stripes */}
              {days.map((d, i) =>
                d.day() === 0 || d.day() === 6 ? (
                  <Box
                    key={i}
                    sx={{
                      position: 'absolute',
                      left: i * dayWidth,
                      top: 0,
                      bottom: 0,
                      width: dayWidth,
                      bgcolor: '#F9F9F9',
                    }}
                  />
                ) : null,
              )}
              {/* Task bars */}
              {row.tasks.map((task, i) => {
                const tStart = dayjs(task.start);
                const tEnd = dayjs(task.finish);
                const left = tStart.diff(projectStart, 'day') * dayWidth;
                const width = Math.max(tEnd.diff(tStart, 'day'), 1) * dayWidth;
                const pctComplete = task.percentComplete ?? 0;

                return (
                  <Tooltip key={task.id} title={`${task.name} (${pctComplete}%)`}>
                    <Box
                      sx={{
                        position: 'absolute',
                        left,
                        top: i * ROW_HEIGHT + 4,
                        width,
                        height: ROW_HEIGHT - 8,
                        bgcolor: task.isCritical ? '#D32F2F' : '#0078D4',
                        borderRadius: '3px',
                        opacity: 0.85,
                        display: 'flex',
                        alignItems: 'center',
                        px: 0.5,
                        overflow: 'hidden',
                        cursor: 'pointer',
                      }}
                    >
                      {/* Progress fill */}
                      <Box
                        sx={{
                          position: 'absolute',
                          left: 0,
                          top: 0,
                          bottom: 0,
                          width: `${pctComplete}%`,
                          bgcolor: 'rgba(0,0,0,0.15)',
                        }}
                      />
                      <Typography
                        variant="caption"
                        sx={{ color: '#FFF', fontSize: '0.6rem', zIndex: 1, whiteSpace: 'nowrap' }}
                      >
                        {task.name}
                      </Typography>
                    </Box>
                  </Tooltip>
                );
              })}
            </Box>
          ))}
        </Box>
      </Box>
    </Box>
  );
};

export default TeamPlanner;
