// @schedulesync/frontend/src/components/views/Timeline.tsx
/**
 * Timeline — executive summary view showing key tasks and milestones
 * on a horizontal timeline with swimlanes. Similar to MS Project's
 * Timeline view at the top of the Gantt chart.
 */

import React, { useMemo } from 'react';
import { Box, Typography, Tooltip, Chip } from '@mui/material';
import dayjs from 'dayjs';
import { useProjectStore } from '../../stores';

const TIMELINE_HEIGHT = 100;
const CALLOUT_HEIGHT = 60;
const MONTH_WIDTH = 80;

const Timeline: React.FC = () => {
  const tasks = useProjectStore((s) => s.tasks);

  const { months, displayTasks, projectStart, projectEnd, totalWidth } = useMemo(() => {
    if (tasks.length === 0) {
      return { months: [], displayTasks: [], projectStart: dayjs(), projectEnd: dayjs(), totalWidth: 0 };
    }

    const starts = tasks.map((t) => dayjs(t.start));
    const ends = tasks.map((t) => dayjs(t.finish));
    const pStart = starts.reduce((a, b) => (a.isBefore(b) ? a : b)).startOf('month');
    const pEnd = ends.reduce((a, b) => (a.isAfter(b) ? a : b)).endOf('month');

    const monthList: dayjs.Dayjs[] = [];
    let cur = pStart;
    while (cur.isBefore(pEnd)) {
      monthList.push(cur);
      cur = cur.add(1, 'month');
    }

    // Show milestones and summary tasks, plus critical tasks
    const display = tasks.filter(
      (t) =>
        t.type === 'milestone' ||
        t.durationMinutes === 0 ||
        t.type === 'summary' ||
        t.isCritical,
    );

    return {
      months: monthList,
      displayTasks: display,
      projectStart: pStart,
      projectEnd: pEnd,
      totalWidth: monthList.length * MONTH_WIDTH,
    };
  }, [tasks]);

  if (tasks.length === 0) {
    return (
      <Box sx={{ p: 4, textAlign: 'center' }}>
        <Typography color="text.secondary">No tasks to display on the timeline.</Typography>
      </Box>
    );
  }

  const getLeft = (date: string) => {
    const d = dayjs(date);
    const monthsDiff = d.diff(projectStart, 'month', true);
    return monthsDiff * MONTH_WIDTH;
  };

  // Separate milestones and bar tasks
  const milestones = displayTasks.filter((t) => t.type === 'milestone' || t.durationMinutes === 0);
  const barTasks = displayTasks.filter((t) => t.type !== 'milestone' && t.durationMinutes > 0);

  return (
    <Box sx={{ display: 'flex', flexDirection: 'column', height: '100%', overflow: 'hidden' }}>
      <Box sx={{ bgcolor: '#F5F5F5', borderBottom: '1px solid #E0E0E0', px: 2, py: 0.5 }}>
        <Typography variant="subtitle2" sx={{ fontWeight: 600, fontSize: '0.75rem' }}>
          Timeline
        </Typography>
      </Box>

      <Box sx={{ flex: 1, overflow: 'auto' }}>
        <Box sx={{ minWidth: totalWidth, position: 'relative' }}>
          {/* Month headers */}
          <Box sx={{ display: 'flex', borderBottom: '2px solid #E0E0E0', height: 28 }}>
            {months.map((m, i) => (
              <Box
                key={i}
                sx={{
                  width: MONTH_WIDTH,
                  minWidth: MONTH_WIDTH,
                  textAlign: 'center',
                  fontSize: '0.7rem',
                  fontWeight: 600,
                  lineHeight: '28px',
                  borderRight: '1px solid #E0E0E0',
                  bgcolor: '#F8F8F8',
                }}
              >
                {m.format('MMM YYYY')}
              </Box>
            ))}
          </Box>

          {/* Callout area for milestones */}
          <Box sx={{ height: CALLOUT_HEIGHT, position: 'relative', borderBottom: '1px solid #E0E0E0' }}>
            {milestones.map((task, i) => {
              const left = getLeft(task.start);
              return (
                <Tooltip key={task.id} title={`${task.name} — ${dayjs(task.start).format('MMM D, YYYY')}`}>
                  <Box
                    sx={{
                      position: 'absolute',
                      left: left - 4,
                      top: 10 + (i % 2) * 24,
                      display: 'flex',
                      flexDirection: 'column',
                      alignItems: 'center',
                    }}
                  >
                    <Box
                      sx={{
                        width: 8,
                        height: 8,
                        bgcolor: task.percentComplete === 100 ? '#1B6B3A' : '#0078D4',
                        transform: 'rotate(45deg)',
                      }}
                    />
                    <Typography
                      variant="caption"
                      sx={{
                        fontSize: '0.55rem',
                        whiteSpace: 'nowrap',
                        mt: 0.25,
                        maxWidth: 80,
                        overflow: 'hidden',
                        textOverflow: 'ellipsis',
                      }}
                    >
                      {task.name}
                    </Typography>
                  </Box>
                </Tooltip>
              );
            })}
          </Box>

          {/* Main timeline band */}
          <Box sx={{ height: TIMELINE_HEIGHT, position: 'relative', bgcolor: '#FAFAFA' }}>
            {/* Month grid lines */}
            {months.map((_, i) => (
              <Box
                key={i}
                sx={{
                  position: 'absolute',
                  left: i * MONTH_WIDTH,
                  top: 0,
                  bottom: 0,
                  borderLeft: '1px solid #E8E8E8',
                }}
              />
            ))}

            {/* Task bars */}
            {barTasks.map((task, i) => {
              const left = getLeft(task.start);
              const right = getLeft(task.finish);
              const width = Math.max(right - left, 4);
              const isSummary = task.type === 'summary';
              const top = 8 + (i % 3) * 28;

              return (
                <Tooltip
                  key={task.id}
                  title={`${task.name} (${task.percentComplete}% complete)`}
                >
                  <Box
                    sx={{
                      position: 'absolute',
                      left,
                      top,
                      width,
                      height: isSummary ? 10 : 22,
                      bgcolor: isSummary
                        ? '#666'
                        : task.isCritical
                          ? '#D32F2F'
                          : '#0078D4',
                      borderRadius: isSummary ? 0 : '3px',
                      opacity: 0.8,
                      display: 'flex',
                      alignItems: 'center',
                      px: 0.5,
                      overflow: 'hidden',
                      cursor: 'pointer',
                    }}
                  >
                    {!isSummary && (
                      <>
                        <Box
                          sx={{
                            position: 'absolute',
                            left: 0,
                            top: 0,
                            bottom: 0,
                            width: `${task.percentComplete}%`,
                            bgcolor: 'rgba(0,0,0,0.15)',
                          }}
                        />
                        <Typography
                          variant="caption"
                          sx={{ color: '#FFF', fontSize: '0.55rem', zIndex: 1, whiteSpace: 'nowrap' }}
                        >
                          {task.name}
                        </Typography>
                      </>
                    )}
                  </Box>
                </Tooltip>
              );
            })}

            {/* Today line */}
            {(() => {
              const todayLeft = getLeft(dayjs().toISOString());
              if (todayLeft >= 0 && todayLeft <= totalWidth) {
                return (
                  <Box
                    sx={{
                      position: 'absolute',
                      left: todayLeft,
                      top: 0,
                      bottom: 0,
                      borderLeft: '2px solid #1B6B3A',
                      opacity: 0.6,
                    }}
                  />
                );
              }
              return null;
            })()}
          </Box>
        </Box>
      </Box>
    </Box>
  );
};

export default Timeline;
