// @schedulesync/backend/src/routes/dependencies.ts
/**
 * /api/projects/:projectId/dependencies routes
 */

import { FastifyInstance } from 'fastify';
import { z } from 'zod';
import { prisma } from '../db.js';
import { recalculateProject } from '../services/schedulingService.js';
import { captureUndo } from '../services/undoService.js';

const createSchema = z.object({
  fromTaskId: z.string(),
  toTaskId: z.string(),
  type: z.number().int().min(0).max(3).default(1), // FS default
  lagMinutes: z.number().default(0),
});

const updateSchema = z.object({
  type: z.number().int().min(0).max(3).optional(),
  lagMinutes: z.number().optional(),
});

export default async function dependencyRoutes(app: FastifyInstance) {
  // List dependencies for a project
  app.get<{ Params: { projectId: string } }>('/', async (req) => {
    const { projectId } = req.params;
    return prisma.dependency.findMany({ where: { projectId } });
  });

  // Create dependency
  app.post<{ Params: { projectId: string } }>('/', async (req, reply) => {
    const { projectId } = req.params;
    const body = createSchema.parse(req.body);

    // Validate both tasks belong to this project
    const [from, to] = await Promise.all([
      prisma.task.findFirst({ where: { id: body.fromTaskId, projectId } }),
      prisma.task.findFirst({ where: { id: body.toTaskId, projectId } }),
    ]);
    if (!from || !to) {
      return reply.code(400).send({ error: 'Both tasks must belong to the project' });
    }

    // Prevent self-link
    if (body.fromTaskId === body.toTaskId) {
      return reply.code(400).send({ error: 'Cannot link a task to itself' });
    }

    // Prevent duplicates
    const existing = await prisma.dependency.findFirst({
      where: { projectId, fromTaskId: body.fromTaskId, toTaskId: body.toTaskId },
    });
    if (existing) {
      return reply.code(409).send({ error: 'Dependency already exists' });
    }

    await captureUndo(projectId, 'Add dependency');

    const dep = await prisma.dependency.create({
      data: {
        projectId,
        fromTaskId: body.fromTaskId,
        toTaskId: body.toTaskId,
        type: body.type,
        lagMinutes: body.lagMinutes,
      },
    });

    await recalculateProject(projectId);
    return reply.code(201).send(dep);
  });

  // Update dependency
  app.patch<{ Params: { projectId: string; depId: string } }>(
    '/:depId',
    async (req) => {
      const { projectId, depId } = req.params;
      const body = updateSchema.parse(req.body);

      await captureUndo(projectId, 'Update dependency');

      const data: Record<string, unknown> = {};
      if (body.type !== undefined) data.type = body.type;
      if (body.lagMinutes !== undefined) data.lagMinutes = body.lagMinutes;

      const dep = await prisma.dependency.update({ where: { id: depId }, data });
      await recalculateProject(projectId);
      return dep;
    },
  );

  // Delete dependency
  app.delete<{ Params: { projectId: string; depId: string } }>(
    '/:depId',
    async (req, reply) => {
      const { projectId, depId } = req.params;

      await captureUndo(projectId, 'Remove dependency');
      await prisma.dependency.delete({ where: { id: depId } });
      await recalculateProject(projectId);
      return reply.code(204).send();
    },
  );
}
