// @schedulesync/backend/src/routes/tasks.ts
/**
 * /api/projects/:projectId/tasks routes
 */

import { FastifyInstance } from 'fastify';
import { z } from 'zod';
import { prisma } from '../db.js';
import { recalculateProject } from '../services/schedulingService.js';
import { captureUndo } from '../services/undoService.js';

const createSchema = z.object({
  name: z.string().min(1),
  durationMinutes: z.number().min(0).default(480),
  start: z.string().datetime().optional(),
  finish: z.string().datetime().optional(),
  parentId: z.string().nullable().optional(),
  type: z.string().default('task'), // "task" | "summary" | "milestone"
  constraintType: z.number().int().default(0),
  constraintDate: z.string().datetime().nullable().optional(),
  isManuallyScheduled: z.boolean().default(false),
  percentComplete: z.number().min(0).max(100).default(0),
  calendarId: z.string().nullable().optional(),
  notes: z.string().optional(),
  externalKey: z.string().optional(),
  sortOrder: z.number().int().optional(),
});

const updateSchema = z.object({
  name: z.string().min(1).optional(),
  durationMinutes: z.number().min(0).optional(),
  start: z.string().datetime().optional(),
  finish: z.string().datetime().optional(),
  parentId: z.string().nullable().optional(),
  type: z.string().optional(),
  constraintType: z.number().int().optional(),
  constraintDate: z.string().datetime().nullable().optional(),
  isManuallyScheduled: z.boolean().optional(),
  percentComplete: z.number().min(0).max(100).optional(),
  calendarId: z.string().nullable().optional(),
  notes: z.string().optional(),
  externalKey: z.string().optional(),
  sortOrder: z.number().int().optional(),
  deadline: z.string().datetime().nullable().optional(),
  actualStart: z.string().datetime().nullable().optional(),
  actualFinish: z.string().datetime().nullable().optional(),
  actualWork: z.number().min(0).optional(),
  actualCost: z.number().min(0).optional(),
  remainingWork: z.number().min(0).optional(),
  remainingCost: z.number().min(0).optional(),
  fixedCost: z.number().min(0).optional(),
});

const batchUpdateSchema = z.object({
  updates: z.array(
    z.object({
      id: z.string(),
      data: updateSchema,
    }),
  ),
  recalculate: z.boolean().default(true),
});

export default async function taskRoutes(app: FastifyInstance) {
  // List tasks for a project (flat, sorted)
  app.get<{ Params: { projectId: string } }>(
    '/',
    async (req) => {
      const { projectId } = req.params;
      return prisma.task.findMany({
        where: { projectId },
        orderBy: { sortOrder: 'asc' },
      });
    },
  );

  // Get single task
  app.get<{ Params: { projectId: string; taskId: string } }>(
    '/:taskId',
    async (req, reply) => {
      const { taskId } = req.params;
      const task = await prisma.task.findUnique({ where: { id: taskId } });
      if (!task) return reply.code(404).send({ error: 'Task not found' });
      return task;
    },
  );

  // Create task
  app.post<{ Params: { projectId: string } }>(
    '/',
    async (req, reply) => {
      const { projectId } = req.params;
      const body = createSchema.parse(req.body);

      await captureUndo(projectId, `Create task "${body.name}"`);

      // Determine project start date as fallback
      const project = await prisma.project.findUniqueOrThrow({
        where: { id: projectId },
      });

      const start = body.start ? new Date(body.start) : project.startDate;
      const finish = body.finish
        ? new Date(body.finish)
        : new Date(start.getTime() + body.durationMinutes * 60_000);

      // Auto sort-order: max + 1
      const maxTask = await prisma.task.findFirst({
        where: { projectId },
        orderBy: { sortOrder: 'desc' },
        select: { sortOrder: true },
      });
      const sortOrder = body.sortOrder ?? (maxTask?.sortOrder ?? -1) + 1;

      const task = await prisma.task.create({
        data: {
          projectId,
          name: body.name,
          wbsCode: '',
          outlineLevel: 0,
          type: body.type,
          durationMinutes: body.durationMinutes,
          start,
          finish,
          constraintType: body.constraintType,
          constraintDate: body.constraintDate ? new Date(body.constraintDate) : null,
          isManuallyScheduled: body.isManuallyScheduled,
          percentComplete: body.percentComplete,
          calendarId: body.calendarId ?? null,
          parentId: body.parentId ?? null,
          sortOrder,
          notes: body.notes ?? '',
          externalKey: body.externalKey,
        },
      });

      await recalculateProject(projectId);
      return reply.code(201).send(task);
    },
  );

  // Update single task
  app.patch<{ Params: { projectId: string; taskId: string } }>(
    '/:taskId',
    async (req) => {
      const { projectId, taskId } = req.params;
      const body = updateSchema.parse(req.body);

      await captureUndo(projectId, `Update task`);

      const data: Record<string, unknown> = {};
      if (body.name !== undefined) data.name = body.name;
      if (body.durationMinutes !== undefined) data.durationMinutes = body.durationMinutes;
      if (body.start !== undefined) data.start = new Date(body.start);
      if (body.finish !== undefined) data.finish = new Date(body.finish);
      if (body.parentId !== undefined) data.parentId = body.parentId;
      if (body.type !== undefined) data.type = body.type;
      if (body.constraintType !== undefined) data.constraintType = body.constraintType;
      if (body.constraintDate !== undefined)
        data.constraintDate = body.constraintDate ? new Date(body.constraintDate) : null;
      if (body.isManuallyScheduled !== undefined)
        data.isManuallyScheduled = body.isManuallyScheduled;
      if (body.percentComplete !== undefined) data.percentComplete = body.percentComplete;
      if (body.calendarId !== undefined) data.calendarId = body.calendarId;
      if (body.notes !== undefined) data.notes = body.notes;
      if (body.externalKey !== undefined) data.externalKey = body.externalKey;
      if (body.sortOrder !== undefined) data.sortOrder = body.sortOrder;
      if (body.deadline !== undefined)
        data.deadline = body.deadline ? new Date(body.deadline) : null;
      if (body.actualStart !== undefined)
        data.actualStart = body.actualStart ? new Date(body.actualStart) : null;
      if (body.actualFinish !== undefined)
        data.actualFinish = body.actualFinish ? new Date(body.actualFinish) : null;
      if (body.actualWork !== undefined) data.actualWork = body.actualWork;
      if (body.actualCost !== undefined) data.actualCost = body.actualCost;
      if (body.remainingWork !== undefined) data.remainingWork = body.remainingWork;
      if (body.remainingCost !== undefined) data.remainingCost = body.remainingCost;
      if (body.fixedCost !== undefined) data.fixedCost = body.fixedCost;

      const task = await prisma.task.update({ where: { id: taskId }, data });

      await recalculateProject(projectId);
      return task;
    },
  );

  // Batch update (for drag, indent/outdent, reorder)
  app.post<{ Params: { projectId: string } }>(
    '/batch',
    async (req) => {
      const { projectId } = req.params;
      const body = batchUpdateSchema.parse(req.body);

      await captureUndo(projectId, `Batch update ${body.updates.length} tasks`);

      const results = await prisma.$transaction(
        body.updates.map((u) => {
          const data: Record<string, unknown> = {};
          const d = u.data;
          if (d.name !== undefined) data.name = d.name;
          if (d.durationMinutes !== undefined) data.durationMinutes = d.durationMinutes;
          if (d.start !== undefined) data.start = new Date(d.start);
          if (d.finish !== undefined) data.finish = new Date(d.finish);
          if (d.parentId !== undefined) data.parentId = d.parentId;
          if (d.type !== undefined) data.type = d.type;
          if (d.constraintType !== undefined) data.constraintType = d.constraintType;
          if (d.constraintDate !== undefined)
            data.constraintDate = d.constraintDate ? new Date(d.constraintDate) : null;
          if (d.isManuallyScheduled !== undefined)
            data.isManuallyScheduled = d.isManuallyScheduled;
          if (d.percentComplete !== undefined) data.percentComplete = d.percentComplete;
          if (d.sortOrder !== undefined) data.sortOrder = d.sortOrder;

          return prisma.task.update({ where: { id: u.id }, data });
        }),
      );

      if (body.recalculate) {
        await recalculateProject(projectId);
      }

      return { updated: results.length };
    },
  );

  // Delete task
  app.delete<{ Params: { projectId: string; taskId: string } }>(
    '/:taskId',
    async (req, reply) => {
      const { projectId, taskId } = req.params;

      await captureUndo(projectId, `Delete task`);

      // Remove dependencies, assignments, baselines, and the task atomically
      await prisma.$transaction([
        prisma.dependency.deleteMany({
          where: {
            OR: [{ fromTaskId: taskId }, { toTaskId: taskId }],
          },
        }),
        prisma.assignment.deleteMany({ where: { taskId } }),
        prisma.baseline.deleteMany({ where: { taskId } }),
        prisma.task.delete({ where: { id: taskId } }),
      ]);

      await recalculateProject(projectId);
      return reply.code(204).send();
    },
  );

  // Recalculate on demand
  app.post<{ Params: { projectId: string } }>(
    '/recalculate',
    async (req) => {
      const { projectId } = req.params;
      await recalculateProject(projectId);
      return { ok: true };
    },
  );
}
