// @schedulesync/mspdi/dist/fieldMapping.js
/**
 * MSPDI field mapping tables.
 *
 * Maps between MSPDI XML element names and internal model properties.
 */
import { ConstraintType, DependencyType, ResourceType, TaskType } from '@schedulesync/engine';
/**
 * MSPDI dependency type values:
 * 0 = FF, 1 = FS, 2 = SF, 3 = SS
 */
export function mspdiDependencyType(mspdiType) {
    switch (mspdiType) {
        case 0: return DependencyType.FF;
        case 1: return DependencyType.FS;
        case 2: return DependencyType.SF;
        case 3: return DependencyType.SS;
        default: return DependencyType.FS;
    }
}
export function dependencyTypeToMspdi(type) {
    switch (type) {
        case DependencyType.FF: return 0;
        case DependencyType.FS: return 1;
        case DependencyType.SF: return 2;
        case DependencyType.SS: return 3;
    }
}
/** MSPDI constraint type int → our ConstraintType enum. Values match directly. */
export function mspdiConstraintType(mspdiType) {
    if (mspdiType >= 0 && mspdiType <= 7)
        return mspdiType;
    return ConstraintType.ASAP;
}
export function constraintTypeToMspdi(type) {
    return type;
}
/** MSPDI resource type: 0 = Material, 1 = Work */
export function mspdiResourceType(mspdiType) {
    return mspdiType === 0 ? ResourceType.Material : ResourceType.Work;
}
export function resourceTypeToMspdi(type) {
    return type === ResourceType.Material ? 0 : 1;
}
/**
 * Determine task type from MSPDI task element attributes.
 */
export function determineTaskType(summary, milestone) {
    if (summary)
        return TaskType.Summary;
    if (milestone)
        return TaskType.Milestone;
    return TaskType.Task;
}
/**
 * Map MSPDI weekday DayType to day-of-week index (0=Sun..6=Sat).
 * MSPDI: 1=Sunday, 2=Monday, ... 7=Saturday
 */
export function mspdiDayOfWeek(dayType) {
    return (dayType - 1) % 7;
}
export function dayOfWeekToMspdi(dayIndex) {
    return dayIndex + 1;
}
